

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>hicberg package &mdash; hicberg 0.1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=01f34227"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="hicberg" href="modules.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            hicberg
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">hicberg</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">hicberg package</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-hicberg.align">hicberg.align module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.align.hic_align"><code class="docutils literal notranslate"><span class="pre">hic_align()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.align.hic_build_index"><code class="docutils literal notranslate"><span class="pre">hic_build_index()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.align.hic_index"><code class="docutils literal notranslate"><span class="pre">hic_index()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.align.hic_sort"><code class="docutils literal notranslate"><span class="pre">hic_sort()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.align.hic_view"><code class="docutils literal notranslate"><span class="pre">hic_view()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-hicberg.benchmark">hicberg.benchmark module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.benchmark.benchmark"><code class="docutils literal notranslate"><span class="pre">benchmark()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-hicberg.eval">hicberg.eval module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.arange_multi"><code class="docutils literal notranslate"><span class="pre">arange_multi()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.check_emptiness"><code class="docutils literal notranslate"><span class="pre">check_emptiness()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.choice"><code class="docutils literal notranslate"><span class="pre">choice()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.chromosight_cmd_generator"><code class="docutils literal notranslate"><span class="pre">chromosight_cmd_generator()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.draw_intervals"><code class="docutils literal notranslate"><span class="pre">draw_intervals()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.generate_dict_coordinates"><code class="docutils literal notranslate"><span class="pre">generate_dict_coordinates()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.get_FN_table"><code class="docutils literal notranslate"><span class="pre">get_FN_table()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.get_FP_table"><code class="docutils literal notranslate"><span class="pre">get_FP_table()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.get_TP_table"><code class="docutils literal notranslate"><span class="pre">get_TP_table()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.get_bin_indexes"><code class="docutils literal notranslate"><span class="pre">get_bin_indexes()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.get_boundaries"><code class="docutils literal notranslate"><span class="pre">get_boundaries()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.get_chromosomes_intervals"><code class="docutils literal notranslate"><span class="pre">get_chromosomes_intervals()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.get_f1_score"><code class="docutils literal notranslate"><span class="pre">get_f1_score()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.get_interval_index"><code class="docutils literal notranslate"><span class="pre">get_interval_index()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.get_intervals_proportions"><code class="docutils literal notranslate"><span class="pre">get_intervals_proportions()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.get_precision"><code class="docutils literal notranslate"><span class="pre">get_precision()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.get_recall"><code class="docutils literal notranslate"><span class="pre">get_recall()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.get_top_pattern"><code class="docutils literal notranslate"><span class="pre">get_top_pattern()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.hicberg_benchmark_cmd_generator"><code class="docutils literal notranslate"><span class="pre">hicberg_benchmark_cmd_generator()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.intersect2D"><code class="docutils literal notranslate"><span class="pre">intersect2D()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.overlap_intervals"><code class="docutils literal notranslate"><span class="pre">overlap_intervals()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.overlap_intervals_outer"><code class="docutils literal notranslate"><span class="pre">overlap_intervals_outer()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.select_reads"><code class="docutils literal notranslate"><span class="pre">select_reads()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.select_reads_multithreads"><code class="docutils literal notranslate"><span class="pre">select_reads_multithreads()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.eval.select_reads_multithreads_bckp"><code class="docutils literal notranslate"><span class="pre">select_reads_multithreads_bckp()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-hicberg.filter">hicberg.filter module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.filter.filter_library"><code class="docutils literal notranslate"><span class="pre">filter_library()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.filter.is_N_rich"><code class="docutils literal notranslate"><span class="pre">is_N_rich()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.filter.is_low_complexity"><code class="docutils literal notranslate"><span class="pre">is_low_complexity()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.filter.is_low_quality"><code class="docutils literal notranslate"><span class="pre">is_low_quality()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-hicberg.io">hicberg.io module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.io.build_matrix"><code class="docutils literal notranslate"><span class="pre">build_matrix()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.io.build_pairs"><code class="docutils literal notranslate"><span class="pre">build_pairs()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.io.create_folder"><code class="docutils literal notranslate"><span class="pre">create_folder()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.io.load_cooler"><code class="docutils literal notranslate"><span class="pre">load_cooler()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.io.load_dictionary"><code class="docutils literal notranslate"><span class="pre">load_dictionary()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.io.merge_predictions"><code class="docutils literal notranslate"><span class="pre">merge_predictions()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.io.tidy_folder"><code class="docutils literal notranslate"><span class="pre">tidy_folder()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-hicberg.pipeline">hicberg.pipeline module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.pipeline.check_tool"><code class="docutils literal notranslate"><span class="pre">check_tool()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.pipeline.pipeline"><code class="docutils literal notranslate"><span class="pre">pipeline()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-hicberg.plot">hicberg.plot module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.plot.plot_benchmark"><code class="docutils literal notranslate"><span class="pre">plot_benchmark()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.plot.plot_couple_repartition"><code class="docutils literal notranslate"><span class="pre">plot_couple_repartition()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.plot.plot_coverages"><code class="docutils literal notranslate"><span class="pre">plot_coverages()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.plot.plot_d1d2"><code class="docutils literal notranslate"><span class="pre">plot_d1d2()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.plot.plot_density"><code class="docutils literal notranslate"><span class="pre">plot_density()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.plot.plot_laws"><code class="docutils literal notranslate"><span class="pre">plot_laws()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.plot.plot_matrix"><code class="docutils literal notranslate"><span class="pre">plot_matrix()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.plot.plot_pattern_reconstruction"><code class="docutils literal notranslate"><span class="pre">plot_pattern_reconstruction()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.plot.plot_trans_ps"><code class="docutils literal notranslate"><span class="pre">plot_trans_ps()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-hicberg.statistics">hicberg.statistics module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.attribute_xs"><code class="docutils literal notranslate"><span class="pre">attribute_xs()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.choice"><code class="docutils literal notranslate"><span class="pre">choice()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.compute_density"><code class="docutils literal notranslate"><span class="pre">compute_density()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.compute_propensity"><code class="docutils literal notranslate"><span class="pre">compute_propensity()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.draw_read_couple"><code class="docutils literal notranslate"><span class="pre">draw_read_couple()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.generate_coverages"><code class="docutils literal notranslate"><span class="pre">generate_coverages()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.generate_d1d2"><code class="docutils literal notranslate"><span class="pre">generate_d1d2()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.generate_density_map"><code class="docutils literal notranslate"><span class="pre">generate_density_map()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.generate_density_map_backup"><code class="docutils literal notranslate"><span class="pre">generate_density_map_backup()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.generate_trans_ps"><code class="docutils literal notranslate"><span class="pre">generate_trans_ps()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.generate_xs"><code class="docutils literal notranslate"><span class="pre">generate_xs()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.get_d1d2"><code class="docutils literal notranslate"><span class="pre">get_d1d2()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.get_density"><code class="docutils literal notranslate"><span class="pre">get_density()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.get_dist_frags"><code class="docutils literal notranslate"><span class="pre">get_dist_frags()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.get_pair_cover"><code class="docutils literal notranslate"><span class="pre">get_pair_cover()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.get_pair_ps"><code class="docutils literal notranslate"><span class="pre">get_pair_ps()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.get_patterns"><code class="docutils literal notranslate"><span class="pre">get_patterns()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.get_restriction_map"><code class="docutils literal notranslate"><span class="pre">get_restriction_map()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.get_top_pattern"><code class="docutils literal notranslate"><span class="pre">get_top_pattern()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.get_trans_ps"><code class="docutils literal notranslate"><span class="pre">get_trans_ps()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.log_bin_genome"><code class="docutils literal notranslate"><span class="pre">log_bin_genome()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.pearson_score"><code class="docutils literal notranslate"><span class="pre">pearson_score()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.statistics.reattribute_reads"><code class="docutils literal notranslate"><span class="pre">reattribute_reads()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-hicberg.utils">hicberg.utils module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.bam_iterator"><code class="docutils literal notranslate"><span class="pre">bam_iterator()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.block_counter"><code class="docutils literal notranslate"><span class="pre">block_counter()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.choice"><code class="docutils literal notranslate"><span class="pre">choice()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.chunk_bam"><code class="docutils literal notranslate"><span class="pre">chunk_bam()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.classify_reads"><code class="docutils literal notranslate"><span class="pre">classify_reads()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.detrend_matrix"><code class="docutils literal notranslate"><span class="pre">detrend_matrix()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.format_blacklist"><code class="docutils literal notranslate"><span class="pre">format_blacklist()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.generate_gaussian_kernel"><code class="docutils literal notranslate"><span class="pre">generate_gaussian_kernel()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.get_bad_bins"><code class="docutils literal notranslate"><span class="pre">get_bad_bins()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.get_bin_table"><code class="docutils literal notranslate"><span class="pre">get_bin_table()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.get_chromosomes_sizes"><code class="docutils literal notranslate"><span class="pre">get_chromosomes_sizes()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.get_chunks"><code class="docutils literal notranslate"><span class="pre">get_chunks()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.get_cis_distance"><code class="docutils literal notranslate"><span class="pre">get_cis_distance()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.get_local_density"><code class="docutils literal notranslate"><span class="pre">get_local_density()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.get_ordered_reads"><code class="docutils literal notranslate"><span class="pre">get_ordered_reads()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.is_blacklisted"><code class="docutils literal notranslate"><span class="pre">is_blacklisted()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.is_circle"><code class="docutils literal notranslate"><span class="pre">is_circle()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.is_duplicated"><code class="docutils literal notranslate"><span class="pre">is_duplicated()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.is_empty_alignment"><code class="docutils literal notranslate"><span class="pre">is_empty_alignment()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.is_intra_chromosome"><code class="docutils literal notranslate"><span class="pre">is_intra_chromosome()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.is_poor_quality"><code class="docutils literal notranslate"><span class="pre">is_poor_quality()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.is_reverse"><code class="docutils literal notranslate"><span class="pre">is_reverse()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.is_uncut"><code class="docutils literal notranslate"><span class="pre">is_uncut()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.is_unmapped"><code class="docutils literal notranslate"><span class="pre">is_unmapped()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.is_unqualitative"><code class="docutils literal notranslate"><span class="pre">is_unqualitative()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.is_weird"><code class="docutils literal notranslate"><span class="pre">is_weird()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.mad_smoothing"><code class="docutils literal notranslate"><span class="pre">mad_smoothing()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.max_consecutive_nans"><code class="docutils literal notranslate"><span class="pre">max_consecutive_nans()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.nan_conv"><code class="docutils literal notranslate"><span class="pre">nan_conv()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.replace_consecutive_zeros_with_mean"><code class="docutils literal notranslate"><span class="pre">replace_consecutive_zeros_with_mean()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.subsample_restriction_map"><code class="docutils literal notranslate"><span class="pre">subsample_restriction_map()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#hicberg.utils.sum_mat_bins"><code class="docutils literal notranslate"><span class="pre">sum_mat_bins()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-hicberg.version">hicberg.version module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-hicberg">Module contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">hicberg package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-hicberg.align">hicberg.align module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.align.hic_align"><code class="docutils literal notranslate"><span class="pre">hic_align()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.align.hic_build_index"><code class="docutils literal notranslate"><span class="pre">hic_build_index()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.align.hic_index"><code class="docutils literal notranslate"><span class="pre">hic_index()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.align.hic_sort"><code class="docutils literal notranslate"><span class="pre">hic_sort()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.align.hic_view"><code class="docutils literal notranslate"><span class="pre">hic_view()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-hicberg.benchmark">hicberg.benchmark module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.benchmark.benchmark"><code class="docutils literal notranslate"><span class="pre">benchmark()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-hicberg.eval">hicberg.eval module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.arange_multi"><code class="docutils literal notranslate"><span class="pre">arange_multi()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.check_emptiness"><code class="docutils literal notranslate"><span class="pre">check_emptiness()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.choice"><code class="docutils literal notranslate"><span class="pre">choice()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.chromosight_cmd_generator"><code class="docutils literal notranslate"><span class="pre">chromosight_cmd_generator()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.draw_intervals"><code class="docutils literal notranslate"><span class="pre">draw_intervals()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.generate_dict_coordinates"><code class="docutils literal notranslate"><span class="pre">generate_dict_coordinates()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.get_FN_table"><code class="docutils literal notranslate"><span class="pre">get_FN_table()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.get_FP_table"><code class="docutils literal notranslate"><span class="pre">get_FP_table()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.get_TP_table"><code class="docutils literal notranslate"><span class="pre">get_TP_table()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.get_bin_indexes"><code class="docutils literal notranslate"><span class="pre">get_bin_indexes()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.get_boundaries"><code class="docutils literal notranslate"><span class="pre">get_boundaries()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.get_chromosomes_intervals"><code class="docutils literal notranslate"><span class="pre">get_chromosomes_intervals()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.get_f1_score"><code class="docutils literal notranslate"><span class="pre">get_f1_score()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.get_interval_index"><code class="docutils literal notranslate"><span class="pre">get_interval_index()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.get_intervals_proportions"><code class="docutils literal notranslate"><span class="pre">get_intervals_proportions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.get_precision"><code class="docutils literal notranslate"><span class="pre">get_precision()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.get_recall"><code class="docutils literal notranslate"><span class="pre">get_recall()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.get_top_pattern"><code class="docutils literal notranslate"><span class="pre">get_top_pattern()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.hicberg_benchmark_cmd_generator"><code class="docutils literal notranslate"><span class="pre">hicberg_benchmark_cmd_generator()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.intersect2D"><code class="docutils literal notranslate"><span class="pre">intersect2D()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.overlap_intervals"><code class="docutils literal notranslate"><span class="pre">overlap_intervals()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.overlap_intervals_outer"><code class="docutils literal notranslate"><span class="pre">overlap_intervals_outer()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.select_reads"><code class="docutils literal notranslate"><span class="pre">select_reads()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.select_reads_multithreads"><code class="docutils literal notranslate"><span class="pre">select_reads_multithreads()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.eval.select_reads_multithreads_bckp"><code class="docutils literal notranslate"><span class="pre">select_reads_multithreads_bckp()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-hicberg.filter">hicberg.filter module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.filter.filter_library"><code class="docutils literal notranslate"><span class="pre">filter_library()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.filter.is_N_rich"><code class="docutils literal notranslate"><span class="pre">is_N_rich()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.filter.is_low_complexity"><code class="docutils literal notranslate"><span class="pre">is_low_complexity()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.filter.is_low_quality"><code class="docutils literal notranslate"><span class="pre">is_low_quality()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-hicberg.io">hicberg.io module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.io.build_matrix"><code class="docutils literal notranslate"><span class="pre">build_matrix()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.io.build_pairs"><code class="docutils literal notranslate"><span class="pre">build_pairs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.io.create_folder"><code class="docutils literal notranslate"><span class="pre">create_folder()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.io.load_cooler"><code class="docutils literal notranslate"><span class="pre">load_cooler()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.io.load_dictionary"><code class="docutils literal notranslate"><span class="pre">load_dictionary()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.io.merge_predictions"><code class="docutils literal notranslate"><span class="pre">merge_predictions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.io.tidy_folder"><code class="docutils literal notranslate"><span class="pre">tidy_folder()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-hicberg.pipeline">hicberg.pipeline module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.pipeline.check_tool"><code class="docutils literal notranslate"><span class="pre">check_tool()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.pipeline.pipeline"><code class="docutils literal notranslate"><span class="pre">pipeline()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-hicberg.plot">hicberg.plot module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.plot.plot_benchmark"><code class="docutils literal notranslate"><span class="pre">plot_benchmark()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.plot.plot_couple_repartition"><code class="docutils literal notranslate"><span class="pre">plot_couple_repartition()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.plot.plot_coverages"><code class="docutils literal notranslate"><span class="pre">plot_coverages()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.plot.plot_d1d2"><code class="docutils literal notranslate"><span class="pre">plot_d1d2()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.plot.plot_density"><code class="docutils literal notranslate"><span class="pre">plot_density()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.plot.plot_laws"><code class="docutils literal notranslate"><span class="pre">plot_laws()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.plot.plot_matrix"><code class="docutils literal notranslate"><span class="pre">plot_matrix()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.plot.plot_pattern_reconstruction"><code class="docutils literal notranslate"><span class="pre">plot_pattern_reconstruction()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.plot.plot_trans_ps"><code class="docutils literal notranslate"><span class="pre">plot_trans_ps()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-hicberg.statistics">hicberg.statistics module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.attribute_xs"><code class="docutils literal notranslate"><span class="pre">attribute_xs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.choice"><code class="docutils literal notranslate"><span class="pre">choice()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.compute_density"><code class="docutils literal notranslate"><span class="pre">compute_density()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.compute_propensity"><code class="docutils literal notranslate"><span class="pre">compute_propensity()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.draw_read_couple"><code class="docutils literal notranslate"><span class="pre">draw_read_couple()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.generate_coverages"><code class="docutils literal notranslate"><span class="pre">generate_coverages()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.generate_d1d2"><code class="docutils literal notranslate"><span class="pre">generate_d1d2()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.generate_density_map"><code class="docutils literal notranslate"><span class="pre">generate_density_map()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.generate_density_map_backup"><code class="docutils literal notranslate"><span class="pre">generate_density_map_backup()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.generate_trans_ps"><code class="docutils literal notranslate"><span class="pre">generate_trans_ps()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.generate_xs"><code class="docutils literal notranslate"><span class="pre">generate_xs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.get_d1d2"><code class="docutils literal notranslate"><span class="pre">get_d1d2()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.get_density"><code class="docutils literal notranslate"><span class="pre">get_density()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.get_dist_frags"><code class="docutils literal notranslate"><span class="pre">get_dist_frags()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.get_pair_cover"><code class="docutils literal notranslate"><span class="pre">get_pair_cover()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.get_pair_ps"><code class="docutils literal notranslate"><span class="pre">get_pair_ps()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.get_patterns"><code class="docutils literal notranslate"><span class="pre">get_patterns()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.get_restriction_map"><code class="docutils literal notranslate"><span class="pre">get_restriction_map()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.get_top_pattern"><code class="docutils literal notranslate"><span class="pre">get_top_pattern()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.get_trans_ps"><code class="docutils literal notranslate"><span class="pre">get_trans_ps()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.log_bin_genome"><code class="docutils literal notranslate"><span class="pre">log_bin_genome()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.pearson_score"><code class="docutils literal notranslate"><span class="pre">pearson_score()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.statistics.reattribute_reads"><code class="docutils literal notranslate"><span class="pre">reattribute_reads()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-hicberg.utils">hicberg.utils module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.bam_iterator"><code class="docutils literal notranslate"><span class="pre">bam_iterator()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.block_counter"><code class="docutils literal notranslate"><span class="pre">block_counter()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.choice"><code class="docutils literal notranslate"><span class="pre">choice()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.chunk_bam"><code class="docutils literal notranslate"><span class="pre">chunk_bam()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.classify_reads"><code class="docutils literal notranslate"><span class="pre">classify_reads()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.detrend_matrix"><code class="docutils literal notranslate"><span class="pre">detrend_matrix()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.format_blacklist"><code class="docutils literal notranslate"><span class="pre">format_blacklist()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.generate_gaussian_kernel"><code class="docutils literal notranslate"><span class="pre">generate_gaussian_kernel()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.get_bad_bins"><code class="docutils literal notranslate"><span class="pre">get_bad_bins()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.get_bin_table"><code class="docutils literal notranslate"><span class="pre">get_bin_table()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.get_chromosomes_sizes"><code class="docutils literal notranslate"><span class="pre">get_chromosomes_sizes()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.get_chunks"><code class="docutils literal notranslate"><span class="pre">get_chunks()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.get_cis_distance"><code class="docutils literal notranslate"><span class="pre">get_cis_distance()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.get_local_density"><code class="docutils literal notranslate"><span class="pre">get_local_density()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.get_ordered_reads"><code class="docutils literal notranslate"><span class="pre">get_ordered_reads()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.is_blacklisted"><code class="docutils literal notranslate"><span class="pre">is_blacklisted()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.is_circle"><code class="docutils literal notranslate"><span class="pre">is_circle()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.is_duplicated"><code class="docutils literal notranslate"><span class="pre">is_duplicated()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.is_empty_alignment"><code class="docutils literal notranslate"><span class="pre">is_empty_alignment()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.is_intra_chromosome"><code class="docutils literal notranslate"><span class="pre">is_intra_chromosome()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.is_poor_quality"><code class="docutils literal notranslate"><span class="pre">is_poor_quality()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.is_reverse"><code class="docutils literal notranslate"><span class="pre">is_reverse()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.is_uncut"><code class="docutils literal notranslate"><span class="pre">is_uncut()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.is_unmapped"><code class="docutils literal notranslate"><span class="pre">is_unmapped()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.is_unqualitative"><code class="docutils literal notranslate"><span class="pre">is_unqualitative()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.is_weird"><code class="docutils literal notranslate"><span class="pre">is_weird()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.mad_smoothing"><code class="docutils literal notranslate"><span class="pre">mad_smoothing()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.max_consecutive_nans"><code class="docutils literal notranslate"><span class="pre">max_consecutive_nans()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.nan_conv"><code class="docutils literal notranslate"><span class="pre">nan_conv()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.replace_consecutive_zeros_with_mean"><code class="docutils literal notranslate"><span class="pre">replace_consecutive_zeros_with_mean()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.subsample_restriction_map"><code class="docutils literal notranslate"><span class="pre">subsample_restriction_map()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#hicberg.utils.sum_mat_bins"><code class="docutils literal notranslate"><span class="pre">sum_mat_bins()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-hicberg.version">hicberg.version module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-hicberg">Module contents</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">hicberg</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="modules.html">hicberg</a></li>
      <li class="breadcrumb-item active">hicberg package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/hicberg.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="hicberg-package">
<h1>hicberg package<a class="headerlink" href="#hicberg-package" title="Link to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="module-hicberg.align">
<span id="hicberg-align-module"></span><h2>hicberg.align module<a class="headerlink" href="#module-hicberg.align" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="hicberg.align.hic_align">
<span class="sig-prename descclassname"><span class="pre">hicberg.align.</span></span><span class="sig-name descname"><span class="pre">hic_align</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fq_for</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fq_rev</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensitivity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'very-sensitive'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_alignment</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cpus</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/align.html#hic_align"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.align.hic_align" title="Link to this definition"></a></dt>
<dd><p>Alignment of reads from HiC experiments along an indexed genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>index</strong> (<em>str</em>) – Path to the index of the genome along which reads are going to be aligned (path to .bt2l files). Default to None, index files are searched to sample_name/data/index/sample_name.</p></li>
<li><p><strong>fq_for</strong> (<em>str</em>) – Path to .fasta containing set of reads to align (forward mate).</p></li>
<li><p><strong>fq_rev</strong> (<em>str</em>) – Path to .fasta containing set of reads to align (forward mate).</p></li>
<li><p><strong>sensitivity</strong> (<em>str</em><em>, </em><em>optional</em>) – Sensitivity of the alignment., by default ‘very_sensitive’</p></li>
<li><p><strong>max_alignment</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum number of alignments to be returned, by default None</p></li>
<li><p><strong>cpus</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of threads allocated for the alignment, by default 1</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path where the alignment files (.sam) should be stored, by default None</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – Set wether or not the shell command should be printed, by default False</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.align.hic_build_index">
<span class="sig-prename descclassname"><span class="pre">hicberg.align.</span></span><span class="sig-name descname"><span class="pre">hic_build_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cpus</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/align.html#hic_build_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.align.hic_build_index" title="Link to this definition"></a></dt>
<dd><p>Building of bowtie2 index (.bt2l files) for read alignment.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome</strong> (<em>str</em>) – Path to the genome file along which reads are going to be aligned.</p></li>
<li><p><strong>cpus</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of threads allocated for the alignment, by default 1</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path where the Bowtie2 index files should be stored, by default None</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – Set wether or not the shell command should be printed, by default False</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.align.hic_index">
<span class="sig-prename descclassname"><span class="pre">hicberg.align.</span></span><span class="sig-name descname"><span class="pre">hic_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam_for</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'1.sorted.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bam_rev</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'2.sorted.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cpus</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/align.html#hic_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.align.hic_index" title="Link to this definition"></a></dt>
<dd><p>Index a coordinate-sorted BGZIP-compressed SAM, BAM or CRAM file for fast random access.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bam_for</strong> (<em>str</em><em>, </em><em>optional</em>) – Forward alignment file to be indexed, by default “1.sorted.bam”</p></li>
<li><p><strong>bam_rev</strong> (<em>str</em><em>, </em><em>optional</em>) – Reverse alignment file to be indexed,, by default “2.sorted.bam”</p></li>
<li><p><strong>cpus</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of threads allocated for the alignment, by default 1</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path where the alignment files (.bam) should be stored, by default None</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – Set wether or not the shell command should be printed, by default False</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.align.hic_sort">
<span class="sig-prename descclassname"><span class="pre">hicberg.align.</span></span><span class="sig-name descname"><span class="pre">hic_sort</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam_for</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'1.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bam_rev</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'2.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cpus</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/align.html#hic_sort"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.align.hic_sort" title="Link to this definition"></a></dt>
<dd><p>Sort .bam alignment files by read_name  (using samtools).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bam_for</strong> (<em>str</em><em>, </em><em>optional</em>) – Forward alignment file to be sorted, by default “1.bam”</p></li>
<li><p><strong>bam_rev</strong> (<em>str</em><em>, </em><em>optional</em>) – Reverse alignment file to be sorted, by default “2.bam”</p></li>
<li><p><strong>cpus</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of threads allocated for the alignment, by default 1</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path where the alignment files (.bam) should be stored, by default None</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – Set wether or not the shell command should be printed, by default False</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.align.hic_view">
<span class="sig-prename descclassname"><span class="pre">hicberg.align.</span></span><span class="sig-name descname"><span class="pre">hic_view</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sam_for</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'1.sam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sam_rev</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'2.sam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cpus</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/align.html#hic_view"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.align.hic_view" title="Link to this definition"></a></dt>
<dd><p>Conversion of .sam alignment files to .bam alignment format (using samtools).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sam_for</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to forward .sam alignment file, by default “1.sam”</p></li>
<li><p><strong>sam_rev</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to reverse .sam alignment file, by default “2.sam”</p></li>
<li><p><strong>cpus</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of threads allocated for the alignment, by default 1</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path where the alignment files (.bam) should be stored, by default None</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – Set wether or not the shell command should be printed, by default False</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-hicberg.benchmark">
<span id="hicberg-benchmark-module"></span><h2>hicberg.benchmark module<a class="headerlink" href="#module-hicberg.benchmark" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="hicberg.benchmark.benchmark">
<span class="sig-prename descclassname"><span class="pre">hicberg.benchmark.</span></span><span class="sig-name descname"><span class="pre">benchmark</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trans_chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trans_position</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strides</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'full'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">auto</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kernel_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">11</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">deviation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pattern</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jitter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trend</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cpus</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">8</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hicberg/benchmark.html#benchmark"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.benchmark.benchmark" title="Link to this definition"></a></dt>
<dd><p>AI is creating summary for benchmark</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – [description], by default None</p></li>
<li><p><strong>chromosome</strong> (<em>str</em><em>, </em><em>optional</em>) – [description], by default “”</p></li>
<li><p><strong>position</strong> (<em>int</em><em>, </em><em>optional</em>) – [description], by default 0</p></li>
<li><p><strong>trans_chromosome</strong> (<em>str</em><em>, </em><em>optional</em>) – [description], by default None</p></li>
<li><p><strong>trans_position</strong> (<em>int</em><em>, </em><em>optional</em>) – [description], by default None</p></li>
<li><p><strong>strides</strong> (<em>list</em><em>[</em><em>int</em><em>]</em><em>, </em><em>optional</em>) – [description], by default []</p></li>
<li><p><strong>mode</strong> (<em>str</em><em>, </em><em>optional</em>) – [description], by default “full”</p></li>
<li><p><strong>auto</strong> (<em>int</em><em>, </em><em>optional</em>) – [description], by default None</p></li>
<li><p><strong>kernel_size</strong> (<em>int</em><em>, </em><em>optional</em>) – [description], by default 11</p></li>
<li><p><strong>deviation</strong> (<em>float</em><em>, </em><em>optional</em>) – [description], by default 0.5</p></li>
<li><p><strong>bins</strong> (<em>int</em><em>, </em><em>optional</em>) – [description], by default None</p></li>
<li><p><strong>circular</strong> (<em>str</em><em>, </em><em>optional</em>) – [description], by default “”</p></li>
<li><p><strong>genome</strong> (<em>str</em><em>, </em><em>optional</em>) – [description], by default None</p></li>
<li><p><strong>pattern</strong> (<em>str</em><em>, </em><em>optional</em>) – [description], by default None</p></li>
<li><p><strong>threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – [description], by default 0.0</p></li>
<li><p><strong>jitter</strong> (<em>int</em><em>, </em><em>optional</em>) – [description], by default 0</p></li>
<li><p><strong>trend</strong> (<em>bool</em><em>, </em><em>optional</em>) – [description], by default True</p></li>
<li><p><strong>top</strong> (<em>int</em><em>, </em><em>optional</em>) – [description], by default 100</p></li>
<li><p><strong>force</strong> (<em>bool</em><em>, </em><em>optional</em>) – [description], by default False</p></li>
<li><p><strong>iterations</strong> (<em>int</em><em>, </em><em>optional</em>) – [description], by default 3</p></li>
<li><p><strong>cpus</strong> (<em>int</em><em>, </em><em>optional</em>) – [description], by default 8</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – [description]</p></li>
<li><p><strong>ValueError</strong> – [description]</p></li>
<li><p><strong>ValueError</strong> – [description]</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-hicberg.eval">
<span id="hicberg-eval-module"></span><h2>hicberg.eval module<a class="headerlink" href="#module-hicberg.eval" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.arange_multi">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">arange_multi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">starts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stops</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hicberg/eval.html#arange_multi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.arange_multi" title="Link to this definition"></a></dt>
<dd><p>Create concatenated ranges of integers for multiple start/length.
:param starts: Starts for each range
:type starts: numpy.ndarray
:param stops: Stops for each range
:type stops: numpy.ndarray
:param lengths: Lengths for each range. Either stops or lengths must be provided.
:type lengths: numpy.ndarray</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>concat_ranges</strong> – Concatenated ranges.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>numpy.ndarray</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>See the following illustrative example:
starts = np.array([1, 3, 4, 6])
stops = np.array([1, 5, 7, 6])
print arange_multi(starts, lengths)
&gt;&gt;&gt; [3 4 4 5 6]
From: <a class="reference external" href="https://codereview.stackexchange.com/questions/83018/vectorized-numpy-version-of-arange-with-multiple-start-stop">https://codereview.stackexchange.com/questions/83018/vectorized-numpy-version-of-arange-with-multiple-start-stop</a></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.check_emptiness">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">check_emptiness</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Cooler</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#check_emptiness"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.check_emptiness" title="Link to this definition"></a></dt>
<dd><p>Check if intervals are empty in a given matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>list</em><em>[</em><em>(</em><em>int</em><em>, </em><em>int</em><em>)</em><em>]</em>) – Lists of intervals (sets) where the value is not in between as element 0 and where the value is  in between as element 1.
One list per chromosome in a dictionary.</p></li>
<li><p><strong>matrix</strong> (<em>cooler.Cooler</em><em>, </em><em>optional</em>) – Cooler Hi-C matrix to be checked for emptiness, by default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Returns True if one of the intervals are empty in the matrix, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.choice">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">choice</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hicberg.eval.choice" title="Link to this definition"></a></dt>
<dd><p>Generates a random sample from a given 1-D array</p>
<div class="versionadded">
<p><span class="versionmodified added">Added in version 1.7.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>New code should use the <code class="docutils literal notranslate"><span class="pre">choice</span></code> method of a <code class="docutils literal notranslate"><span class="pre">default_rng()</span></code>
instance instead; please see the <span class="xref std std-ref">random-quick-start</span>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a</strong> (<em>1-D array-like</em><em> or </em><em>int</em>) – If an ndarray, a random sample is generated from its elements.
If an int, the random sample is generated as if it were <code class="docutils literal notranslate"><span class="pre">np.arange(a)</span></code></p></li>
<li><p><strong>size</strong> (<em>int</em><em> or </em><em>tuple</em><em> of </em><em>ints</em><em>, </em><em>optional</em>) – Output shape.  If the given shape is, e.g., <code class="docutils literal notranslate"><span class="pre">(m,</span> <span class="pre">n,</span> <span class="pre">k)</span></code>, then
<code class="docutils literal notranslate"><span class="pre">m</span> <span class="pre">*</span> <span class="pre">n</span> <span class="pre">*</span> <span class="pre">k</span></code> samples are drawn.  Default is None, in which case a
single value is returned.</p></li>
<li><p><strong>replace</strong> (<em>boolean</em><em>, </em><em>optional</em>) – Whether the sample is with or without replacement. Default is True,
meaning that a value of <code class="docutils literal notranslate"><span class="pre">a</span></code> can be selected multiple times.</p></li>
<li><p><strong>p</strong> (<em>1-D array-like</em><em>, </em><em>optional</em>) – The probabilities associated with each entry in a.
If not given, the sample assumes a uniform distribution over all
entries in <code class="docutils literal notranslate"><span class="pre">a</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>samples</strong> – The generated random samples</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>single item or ndarray</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If a is an int and less than zero, if a or p are not 1-dimensional,
    if a is an array-like of size 0, if p is not a vector of
    probabilities, if a and p have different lengths, or if
    replace=False and the sample size is greater than the population
    size</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">randint</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">shuffle</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">permutation</span></code></p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">random.Generator.choice</span></code></dt><dd><p>which should be used in new code</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>Setting user-specified probabilities through <code class="docutils literal notranslate"><span class="pre">p</span></code> uses a more general but less
efficient sampler than the default. The general sampler produces a different sample
than the optimized sampler even if each element of <code class="docutils literal notranslate"><span class="pre">p</span></code> is 1 / len(a).</p>
<p>Sampling random rows from a 2-D array is not possible with this function,
but is possible with <cite>Generator.choice</cite> through its <code class="docutils literal notranslate"><span class="pre">axis</span></code> keyword.</p>
<p class="rubric">Examples</p>
<p>Generate a uniform random sample from np.arange(5) of size 3:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="go">array([0, 3, 4]) # random</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#This is equivalent to np.random.randint(0,5,3)</span>
</pre></div>
</div>
<p>Generate a non-uniform random sample from np.arange(5) of size 3:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">array([3, 3, 0]) # random</span>
</pre></div>
</div>
<p>Generate a uniform random sample from np.arange(5) of size 3 without
replacement:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">array([3,1,0]) # random</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#This is equivalent to np.random.permutation(np.arange(5))[:3]</span>
</pre></div>
</div>
<p>Generate a non-uniform random sample from np.arange(5) of size
3 without replacement:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">array([2, 3, 0]) # random</span>
</pre></div>
</div>
<p>Any of the above can be repeated with an arbitrary array-like
instead of just integers. For instance:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aa_milne_arr</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;pooh&#39;</span><span class="p">,</span> <span class="s1">&#39;rabbit&#39;</span><span class="p">,</span> <span class="s1">&#39;piglet&#39;</span><span class="p">,</span> <span class="s1">&#39;Christopher&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">aa_milne_arr</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">])</span>
<span class="go">array([&#39;pooh&#39;, &#39;pooh&#39;, &#39;pooh&#39;, &#39;Christopher&#39;, &#39;piglet&#39;], # random</span>
<span class="go">      dtype=&#39;&lt;U11&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.chromosight_cmd_generator">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">chromosight_cmd_generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pattern</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'loops'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">untrend</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#chromosight_cmd_generator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.chromosight_cmd_generator" title="Link to this definition"></a></dt>
<dd><p>Generate chromosight command line to run pattern detection.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to Hi-C balanced contact matrix in .cool format , by default None</p></li>
<li><p><strong>pattern</strong> (<em>str</em><em>, </em><em>optional</em>) – Pattern to detect, by default “loops”</p></li>
<li><p><strong>untrend</strong> (<em>bool</em><em>, </em><em>optional</em>) – Set if map has to be detrended, by default True</p></li>
<li><p><strong>mode</strong> (<em>bool</em><em>, </em><em>optional</em>) – Set either the detection has to be performed before or after map reconstruction, by default False</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save alignments, by default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Chromosight command line to run to run pattern detection</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.draw_intervals">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">draw_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nb_intervals</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_sizes_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'chromosome_sizes.npy'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#draw_intervals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.draw_intervals" title="Link to this definition"></a></dt>
<dd><p>Draw intervals from a given chromosome sizes dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nb_intervals</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of intervals to draw, by default 1</p></li>
<li><p><strong>bins</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of the desired bin, by default 2000., by default 2000</p></li>
<li><p><strong>chrom_sizes_dict</strong> (<em>str</em>) – Path to a dictionary containing chromosome sizes as {chromosome : size} saved in .npy format. By default chromosome_sizes.npy</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary containing the intervals as {chromosome : [(start, end), …]}.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, list[tuple[int, int]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.generate_dict_coordinates">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">generate_dict_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'unrescued_map.cool'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_sizes_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'chromosome_sizes.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strides</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[0]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trans_chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trans_position</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">auto</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#generate_dict_coordinates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.generate_dict_coordinates" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.get_FN_table">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">get_FN_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_pattern</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_pattern_recall</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jitter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#get_FN_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.get_FN_table" title="Link to this definition"></a></dt>
<dd><p>Take dataframe of pattern call (Chromosight) before and after reconstruction
and return table of retieved patterns</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_pattern</strong> (<em>[</em><em>dataframe</em><em>]</em>) – Pandas DataFrame given by chromosight before HiC map reconstruction.</p></li>
<li><p><strong>df_pattern_recall</strong> (<em>[</em><em>dataframe</em><em>]</em>) – Pandas DataFrame given by chromosight after HiC map reconstruction.</p></li>
<li><p><strong>chromosome</strong> (<em>[</em><em>str</em><em>]</em>) – chromosome to select position within.</p></li>
<li><p><strong>bin_size</strong> (<em>[</em><em>int</em><em>]</em>) – bin size used to construct the HiC map.</p></li>
<li><p><strong>jitter</strong> (<em>[</em><em>int</em><em>]</em>) – jitter to apply to the pattern recall table to allow overlapping with the pattern table. By default 0.</p></li>
<li><p><strong>threshold</strong> (<em>[</em><em>float</em><em>]</em>) – threshold to apply to the pattern table to select patterns to consider. By default None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>dataframe</strong> – Table containing components before and after pattern recall (True positives)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>[pandas DataFrame]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.get_FP_table">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">get_FP_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_pattern</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_pattern_recall</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jitter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hicberg/eval.html#get_FP_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.get_FP_table" title="Link to this definition"></a></dt>
<dd><p>Take dataframe of pattern call (Chromosight) before and after reconstruction
and return table of retieved patterns</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_pattern</strong> (<em>[</em><em>dataframe</em><em>]</em>) – Pandas DataFrame given by chromosight before HiC map reconstruction.</p></li>
<li><p><strong>df_pattern_recall</strong> (<em>[</em><em>dataframe</em><em>]</em>) – Pandas DataFrame given by chromosight after HiC map reconstruction.</p></li>
<li><p><strong>chromosome</strong> (<em>[</em><em>str</em><em>]</em>) – chromosome to select position within.</p></li>
<li><p><strong>bin_size</strong> (<em>[</em><em>int</em><em>]</em>) – bin size used to construct the HiC map.</p></li>
<li><p><strong>jitter</strong> (<em>[</em><em>int</em><em>]</em>) – jitter to apply to the pattern recall table to allow overlapping with the pattern table. By default 0.</p></li>
<li><p><strong>threshold</strong> (<em>[</em><em>float</em><em>]</em>) – threshold to apply to the pattern table to select patterns to consider. By default None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>dataframe</strong> – Table containing components before and after pattern recall (True positives)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>[pandas DataFrame]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.get_TP_table">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">get_TP_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_pattern</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_pattern_recall</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jitter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#get_TP_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.get_TP_table" title="Link to this definition"></a></dt>
<dd><p>Take dataframe of pattern call (Chromosight) before and after reconstruction
and return table of retieved patterns</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_pattern</strong> (<em>[</em><em>dataframe</em><em>]</em>) – Pandas DataFrame given by chromosight before HiC map reconstruction.</p></li>
<li><p><strong>df_pattern_recall</strong> (<em>[</em><em>dataframe</em><em>]</em>) – Pandas DataFrame given by chromosight after HiC map reconstruction.</p></li>
<li><p><strong>chromosome</strong> (<em>[</em><em>str</em><em>]</em>) – chromosome to select position within.</p></li>
<li><p><strong>bin_size</strong> (<em>[</em><em>int</em><em>]</em>) – bin size used to construct the HiC map.</p></li>
<li><p><strong>jitter</strong> (<em>[</em><em>int</em><em>]</em>) – jitter to apply to the pattern recall table to allow overlapping with the pattern table. By default 0.</p></li>
<li><p><strong>threshold</strong> (<em>[</em><em>float</em><em>]</em>) – threshold to apply to the pattern table to select patterns to consider. By default None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>dataframe</strong> – Table containing components before and after pattern recall (True positives)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>[pandas DataFrame]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.get_bin_indexes">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">get_bin_indexes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dictionary</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hicberg/eval.html#get_bin_indexes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.get_bin_indexes" title="Link to this definition"></a></dt>
<dd><p>Return the bins indexes (in the cooler matrix) corresponding to the genomic coordinates in the dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>matrix</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to a cooler matrix file, by default None</p></li>
<li><p><strong>dictionary</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary of genomic coordinates where reads are going to be selected for duplication, by default None</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.get_boundaries">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">get_boundaries</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">position</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_sizes_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'chromosome_sizes.npy'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#get_boundaries"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.get_boundaries" title="Link to this definition"></a></dt>
<dd><p>Return boundaries surrounding position considering regular binning of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>position</strong> (<em>int</em><em>, </em><em>optional</em>) – Position to get surrounding boundaries from, by default None</p></li>
<li><p><strong>bins</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of the desired bin, by default 2000., by default 2000</p></li>
<li><p><strong>chromosome</strong> (<em>str</em><em> or </em><em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Chromosomes associated to positions to get surrounding boundaries from, by default None</p></li>
<li><p><strong>chrom_sizes_dict</strong> (<em>str</em>) – Path to a dictionary containing chromosome sizes as {chromosome : size} saved in .npy format. By default chromosome_sizes.npy</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Tuple containing the boundaries of the position : (lower_bound, upper_bound).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[int, int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.get_chromosomes_intervals">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">get_chromosomes_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_sizes_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'chromosome_sizes.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#get_chromosomes_intervals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.get_chromosomes_intervals" title="Link to this definition"></a></dt>
<dd><p>Get all possible intervals from a given chromosome considering a chromosome sizes dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bins</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of the desired bin, by default 2000., by default 2000</p></li>
<li><p><strong>chrom_sizes_dict</strong> (<em>str</em>) – Path to a dictionary containing chromosome sizes as {chromosome : size} saved in .npy format. By default chromosome_sizes.npy</p></li>
<li><p><strong>chromosome</strong> (<em>str</em><em>, </em><em>optional</em>) – Chromosome to get intervals from, by default “”</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List containing the intervals as [(start, end), …]</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[(int, int)]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.get_f1_score">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">get_f1_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_pattern</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_pattern_recall</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jitter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#get_f1_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.get_f1_score" title="Link to this definition"></a></dt>
<dd><p>Take dataframe of pattern call (Chromosight) before and after reconstruction
and return f1 score i.e. 2 * ((precision * recall) / (precision + recall)).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_pattern</strong> (<em>[</em><em>dataframe</em><em>]</em>) – Pandas DataFrame given by chromosight before HiC map reconstruction.</p></li>
<li><p><strong>df_pattern_recall</strong> (<em>[</em><em>dataframe</em><em>]</em>) – Pandas DataFrame given by chromosight after HiC map reconstruction.</p></li>
<li><p><strong>chromosome</strong> (<em>[</em><em>str</em><em>]</em>) – chromosome to select position within.</p></li>
<li><p><strong>bin_size</strong> (<em>[</em><em>int</em><em>]</em>) – bin size used to construct the HiC map.</p></li>
<li><p><strong>jitter</strong> (<em>[</em><em>int</em><em>]</em>) – jitter to apply to the pattern recall table to allow overlapping with the pattern table. By default 0.</p></li>
<li><p><strong>threshold</strong> (<em>[</em><em>float</em><em>]</em>) – threshold to apply to the pattern table to select patterns to consider. By default None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>f1_score</strong> – 2 * ((precision * recall) / (precision + recall))</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.get_interval_index">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">get_interval_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intervals_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_sizes_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#get_interval_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.get_interval_index" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chromosome</strong> (<em>[</em><em>str</em><em>]</em>) – chromosome linked to value and intervals to search in.</p></li>
<li><p><strong>value</strong> (<em>[</em><em>int</em><em>]</em>) – Value to search in intervals.</p></li>
<li><p><strong>intervals_dict</strong> (<em>[</em><em>dict</em><em>]</em>) – Dictionary of intervals as key : chromosome, value : [(low_limit_0, up_limit_0), (low_limit_1, up_limit_1), …, (low_limit_n, up_limit_n)].</p></li>
<li><p><strong>chrom_sizes_dict</strong> (<em>str</em>) – Path to a dictionary containing chromosome sizes as {chromosome : size} saved in .npy format. By default chromosome_sizes.npy</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Lists of intervals (sets) where the value is not in between as element 0 and where the value is  in between as element 1.
One list per chromosome in a dictionary.
If the value in not in any interval, return None.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>[dict]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.get_intervals_proportions">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">get_intervals_proportions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chrom_sizes_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'chromosome_sizes.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_intervals</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#get_intervals_proportions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.get_intervals_proportions" title="Link to this definition"></a></dt>
<dd><p>Extract a dictionary containing the number of intervals to draw considering the size of each chromosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom_sizes_dict</strong> (<em>str</em>) – Path to a dictionary containing chromosome sizes as {chromosome : size} saved in .npy format. By default chromosome_sizes.npy</p></li>
<li><p><strong>nb_intervals</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of intervals to draw, by default 1</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary containing proportion by intervals as {chromosome : proportion}.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.get_precision">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">get_precision</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_pattern</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_pattern_recall</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jitter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#get_precision"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.get_precision" title="Link to this definition"></a></dt>
<dd><p>Take dataframe of pattern call (Chromosight) before and after reconstruction
and return precision score corresponding to the number of true positives over the number of true positives and false positives.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_pattern</strong> (<em>[</em><em>dataframe</em><em>]</em>) – Pandas DataFrame given by chromosight before HiC map reconstruction.</p></li>
<li><p><strong>df_pattern_recall</strong> (<em>[</em><em>dataframe</em><em>]</em>) – Pandas DataFrame given by chromosight after HiC map reconstruction.</p></li>
<li><p><strong>chromosome</strong> (<em>[</em><em>str</em><em>]</em>) – chromosome to select position within.</p></li>
<li><p><strong>bin_size</strong> (<em>[</em><em>int</em><em>]</em>) – bin size used to construct the HiC map.</p></li>
<li><p><strong>jitter</strong> (<em>[</em><em>int</em><em>]</em>) – jitter to apply to the pattern recall table to allow overlapping with the pattern table. By default 0.</p></li>
<li><p><strong>threshold</strong> (<em>[</em><em>float</em><em>]</em>) – threshold to apply to the pattern table to select patterns to consider. By default None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>precision</strong> – number of true positives over the number of true positives and false positives</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.get_recall">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">get_recall</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_pattern</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_pattern_recall</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jitter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#get_recall"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.get_recall" title="Link to this definition"></a></dt>
<dd><p>Take dataframe of pattern call (Chromosight) before and after reconstruction
and return recall score corresponding to the number of true positives over the number of true positives and false negatives.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_pattern</strong> (<em>[</em><em>dataframe</em><em>]</em>) – Pandas DataFrame given by chromosight before HiC map reconstruction.</p></li>
<li><p><strong>df_pattern_recall</strong> (<em>[</em><em>dataframe</em><em>]</em>) – Pandas DataFrame given by chromosight after HiC map reconstruction.</p></li>
<li><p><strong>chromosome</strong> (<em>[</em><em>str</em><em>]</em>) – chromosome to select position within.</p></li>
<li><p><strong>bin_size</strong> (<em>[</em><em>int</em><em>]</em>) – bin size used to construct the HiC map.</p></li>
<li><p><strong>jitter</strong> (<em>[</em><em>int</em><em>]</em>) – jitter to apply to the pattern recall table to allow overlapping with the pattern table. By default 0.</p></li>
<li><p><strong>threshold</strong> (<em>[</em><em>float</em><em>]</em>) – threshold to apply to the pattern table to select patterns to consider. By default None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>recall</strong> – number of true positives over the number of true positives and false negatives</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.get_top_pattern">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">get_top_pattern</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#get_top_pattern"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.get_top_pattern" title="Link to this definition"></a></dt>
<dd><p>Get top patterns from a dataframe</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em><em>, </em><em>optional</em>) – Dataframe containing patterns given by Chromosight, by default None</p></li>
<li><p><strong>top</strong> (<em>int</em><em>, </em><em>optional</em>) – Percentage of top patterns to get, by default 10</p></li>
<li><p><strong>threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Pattern Pearson score to consider to select pattern, by default 0.0</p></li>
<li><p><strong>chromosome</strong> (<em>str</em><em>, </em><em>optional</em>) – Chromosome to consider, by default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe containing top percentage patterns.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.hicberg_benchmark_cmd_generator">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">hicberg_benchmark_cmd_generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'full'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#hicberg_benchmark_cmd_generator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.hicberg_benchmark_cmd_generator" title="Link to this definition"></a></dt>
<dd><p>Get top patterns from a dataframe</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em><em>, </em><em>optional</em>) – Dataframe containing patterns given by Chromosight, by default None</p></li>
<li><p><strong>top</strong> (<em>int</em><em>, </em><em>optional</em>) – Percentage of top patterns to get, by default 10</p></li>
<li><p><strong>threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Pattern Pearson score to consider to select pattern, by default 0.0</p></li>
<li><p><strong>chromosome</strong> (<em>str</em><em>, </em><em>optional</em>) – Chromosome to consider, by default None</p></li>
<li><p><strong>mode</strong> (<em>str</em><em>, </em><em>optional</em>) – Mode to consider for hicberg benchmark, by default “full”</p></li>
<li><p><strong>genome</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the genome to consider for hicberg benchmark, by default None</p></li>
<li><p><strong>bins</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of bins to consider for hicberg benchmark, by default 0</p></li>
<li><p><strong>output</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the output folder to consider for hicberg benchmark, by default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>HiC-BERG command line to run to evaluate reconstruction after pattern discarding</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.intersect2D">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">intersect2D</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hicberg/eval.html#intersect2D"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.intersect2D" title="Link to this definition"></a></dt>
<dd><p>Find row intersection between 2D numpy arrays, a and b.
Returns another numpy array with shared rows</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.overlap_intervals">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">overlap_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">starts1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ends1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">starts2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ends2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">closed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sort</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hicberg/eval.html#overlap_intervals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.overlap_intervals" title="Link to this definition"></a></dt>
<dd><p>Take two sets of intervals and return the indices of pairs of overlapping intervals.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>starts1</strong> (<em>numpy.ndarray</em>) – Interval coordinates. Warning: if provided as pandas.Series, indices
will be ignored.</p></li>
<li><p><strong>ends1</strong> (<em>numpy.ndarray</em>) – Interval coordinates. Warning: if provided as pandas.Series, indices
will be ignored.</p></li>
<li><p><strong>starts2</strong> (<em>numpy.ndarray</em>) – Interval coordinates. Warning: if provided as pandas.Series, indices
will be ignored.</p></li>
<li><p><strong>ends2</strong> (<em>numpy.ndarray</em>) – Interval coordinates. Warning: if provided as pandas.Series, indices
will be ignored.</p></li>
<li><p><strong>closed</strong> (<em>bool</em>) – If True, then treat intervals as closed and report single-point overlaps.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>overlap_ids</strong> – An Nx2 array containing the indices of pairs of overlapping intervals.
The 1st column contains ids from the 1st set, the 2nd column has ids
from the 2nd set.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.overlap_intervals_outer">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">overlap_intervals_outer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">starts1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ends1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">starts2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ends2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">closed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hicberg/eval.html#overlap_intervals_outer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.overlap_intervals_outer" title="Link to this definition"></a></dt>
<dd><p>Take two sets of intervals and return the indices of pairs of overlapping intervals,
as well as the indices of the intervals that do not overlap any other interval.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>starts1</strong> (<em>numpy.ndarray</em>) – Interval coordinates. Warning: if provided as pandas.Series, indices
will be ignored.</p></li>
<li><p><strong>ends1</strong> (<em>numpy.ndarray</em>) – Interval coordinates. Warning: if provided as pandas.Series, indices
will be ignored.</p></li>
<li><p><strong>starts2</strong> (<em>numpy.ndarray</em>) – Interval coordinates. Warning: if provided as pandas.Series, indices
will be ignored.</p></li>
<li><p><strong>ends2</strong> (<em>numpy.ndarray</em>) – Interval coordinates. Warning: if provided as pandas.Series, indices
will be ignored.</p></li>
<li><p><strong>closed</strong> (<em>bool</em>) – If True, then treat intervals as closed and report single-point overlaps.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>overlap_ids</strong> (<em>numpy.ndarray</em>) – An Nx2 array containing the indices of pairs of overlapping intervals.
The 1st column contains ids from the 1st set, the 2nd column has ids
from the 2nd set.</p></li>
<li><p><strong>no_overlap_ids1, no_overlap_ids2</strong> (<em>numpy.ndarray</em>) – Two 1D arrays containing the indices of intervals in sets 1 and 2
respectively that do not overlap with any interval in the other set.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.select_reads">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">select_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam_for</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group1.1.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bam_rev</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group1.2.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matrix_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'unrescued_map.cool'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_sizes_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'chromosome_sizes.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strides</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[0]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trans_chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trans_position</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">auto</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#select_reads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.select_reads" title="Link to this definition"></a></dt>
<dd><p>Select reads from a given matrix and alignment files. Two groups of alignment files are produced (.bam):
- group1.1.in.bam and group1.2.in.bam : Forward and reverse reads from the source and target genomic interval. These reads are going to be duplicated un each intervals.
- group1.1.out.bam and group1.2.out.bam : Forward and reverse reads not included the source and target genomic interval. These reads are going to be conserved and not duplicated.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bam_for</strong> (<em>str</em><em>, </em><em>optional</em>) – Forward alignment file to select reads from, by default “group1.1.bam”</p></li>
<li><p><strong>bam_rev</strong> (<em>str</em><em>, </em><em>optional</em>) – Reverse alignment file to select reads from, by default “group1.2.bam”</p></li>
<li><p><strong>matrix_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the matrix from which the selection will be based (.cool format), by default “unrescued_map.cool”</p></li>
<li><p><strong>position</strong> (<em>int</em><em>, </em><em>optional</em>) – Genomic coordinates (0-based) used to defined the source genomic interval, by default 0</p></li>
<li><p><strong>chromosome</strong> (<em>str</em><em> or </em><em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Chromosome associated to the 0-based coordinates given by “position’, by default “”</p></li>
<li><p><strong>bin_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the given matrix, by default 2000</p></li>
<li><p><strong>chrom_sizes_dict</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to a dictionary containing chromosome sizes as {chromosome : size} saved in .npy format, by default “chromosome_sizes.npy”</p></li>
<li><p><strong>strides</strong> (<em>list</em><em>[</em><em>int</em><em>]</em><em>, </em><em>optional</em>) – List of strides to apply from “position” to define target genomics intervals (in bp.), by default [0]</p></li>
<li><p><strong>trans_chromosome</strong> (<em>str</em><em>, </em><em>optional</em>) – Chromosomes to consider as trans-chromosomes to define genomic target intervals, by default None</p></li>
<li><p><strong>trans_position</strong> (<em>list</em><em>[</em><em>int</em><em>]</em><em>, </em><em>optional</em>) – Genomic coordinates (0-based) used to defined the trans-chromosomal target genomic interval, by default None</p></li>
<li><p><strong>nb_bins</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of bins to consider on each side of the defined source and target intervals, by default 1</p></li>
<li><p><strong>random</strong> (<em>bool</em><em>, </em><em>optional</em>) – Set wether or not the source and target intervals are defined at random, by default False</p></li>
<li><p><strong>auto</strong> (<em>int</em><em>, </em><em>optional</em>) – Set the number of intervals to get while picking at random, by default None</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save alignments, by default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary of intervals as key : chromosome, value : [(low_limit_0, up_limit_0), (low_limit_1, up_limit_1), …, (low_limit_n, up_limit_n)].</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, list[(int, int)]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.select_reads_multithreads">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">select_reads_multithreads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">bam_couple:</span> <span class="pre">tuple[str,</span> <span class="pre">str],</span> <span class="pre">interval_dictionary:</span> <span class="pre">dict[str,</span> <span class="pre">(&lt;class</span> <span class="pre">'int'&gt;,</span> <span class="pre">&lt;class</span> <span class="pre">'int'&gt;)],</span> <span class="pre">chrom_sizes_dict:</span> <span class="pre">str</span> <span class="pre">=</span> <span class="pre">'chromosome_sizes.npy',</span> <span class="pre">output_dir:</span> <span class="pre">str</span> <span class="pre">|</span> <span class="pre">None</span> <span class="pre">=</span> <span class="pre">None</span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#select_reads_multithreads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.select_reads_multithreads" title="Link to this definition"></a></dt>
<dd><p>Select reads from a given matrix and alignment files. Two groups of alignment files are produced (.bam):
- group1.1.in.bam and group1.2.in.bam : Forward and reverse reads from the source and target genomic interval. These reads are going to be duplicated un each intervals.
- group1.1.out.bam and group1.2.out.bam : Forward and reverse reads not included the source and target genomic interval. These reads are going to be conserved and not duplicated.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bam_for</strong> (<em>str</em><em>, </em><em>optional</em>) – Forward alignment file to select reads from, by default “group1.1.bam”</p></li>
<li><p><strong>bam_rev</strong> (<em>str</em><em>, </em><em>optional</em>) – Reverse alignment file to select reads from, by default “group1.2.bam”</p></li>
<li><p><strong>matrix_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the matrix from which the selection will be based (.cool format), by default “unrescued_map.cool”</p></li>
<li><p><strong>position</strong> (<em>int</em><em>, </em><em>optional</em>) – Genomic coordinates (0-based) used to defined the source genomic interval, by default 0</p></li>
<li><p><strong>chromosome</strong> (<em>str</em><em> or </em><em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Chromosome associated to the 0-based coordinates given by “position’, by default “”</p></li>
<li><p><strong>bin_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the given matrix, by default 2000</p></li>
<li><p><strong>chrom_sizes_dict</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to a dictionary containing chromosome sizes as {chromosome : size} saved in .npy format, by default “chromosome_sizes.npy”</p></li>
<li><p><strong>strides</strong> (<em>list</em><em>[</em><em>int</em><em>]</em><em>, </em><em>optional</em>) – List of strides to apply from “position” to define target genomics intervals (in bp.), by default [0]</p></li>
<li><p><strong>trans_chromosome</strong> (<em>str</em><em>, </em><em>optional</em>) – Chromosomes to consider as trans-chromosomes to define genomic target intervals, by default None</p></li>
<li><p><strong>trans_position</strong> (<em>list</em><em>[</em><em>int</em><em>]</em><em>, </em><em>optional</em>) – Genomic coordinates (0-based) used to defined the trans-chromosomal target genomic interval, by default None</p></li>
<li><p><strong>nb_bins</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of bins to consider on each side of the defined source and target intervals, by default 1</p></li>
<li><p><strong>random</strong> (<em>bool</em><em>, </em><em>optional</em>) – Set wether or not the source and target intervals are defined at random, by default False</p></li>
<li><p><strong>auto</strong> (<em>int</em><em>, </em><em>optional</em>) – Set the number of intervals to get while picking at random, by default None</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save alignments, by default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary of intervals as key : chromosome, value : [(low_limit_0, up_limit_0), (low_limit_1, up_limit_1), …, (low_limit_n, up_limit_n)].</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, list[(int, int)]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.eval.select_reads_multithreads_bckp">
<span class="sig-prename descclassname"><span class="pre">hicberg.eval.</span></span><span class="sig-name descname"><span class="pre">select_reads_multithreads_bckp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam_couple</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matrix_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'unrescued_map.cool'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_sizes_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'chromosome_sizes.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strides</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[0]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trans_chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trans_position</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">auto</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/eval.html#select_reads_multithreads_bckp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.eval.select_reads_multithreads_bckp" title="Link to this definition"></a></dt>
<dd><p>Select reads from a given matrix and alignment files. Two groups of alignment files are produced (.bam):
- group1.1.in.bam and group1.2.in.bam : Forward and reverse reads from the source and target genomic interval. These reads are going to be duplicated un each intervals.
- group1.1.out.bam and group1.2.out.bam : Forward and reverse reads not included the source and target genomic interval. These reads are going to be conserved and not duplicated.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bam_for</strong> (<em>str</em><em>, </em><em>optional</em>) – Forward alignment file to select reads from, by default “group1.1.bam”</p></li>
<li><p><strong>bam_rev</strong> (<em>str</em><em>, </em><em>optional</em>) – Reverse alignment file to select reads from, by default “group1.2.bam”</p></li>
<li><p><strong>matrix_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the matrix from which the selection will be based (.cool format), by default “unrescued_map.cool”</p></li>
<li><p><strong>position</strong> (<em>int</em><em>, </em><em>optional</em>) – Genomic coordinates (0-based) used to defined the source genomic interval, by default 0</p></li>
<li><p><strong>chromosome</strong> (<em>str</em><em> or </em><em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Chromosome associated to the 0-based coordinates given by “position’, by default “”</p></li>
<li><p><strong>bin_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the given matrix, by default 2000</p></li>
<li><p><strong>chrom_sizes_dict</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to a dictionary containing chromosome sizes as {chromosome : size} saved in .npy format, by default “chromosome_sizes.npy”</p></li>
<li><p><strong>strides</strong> (<em>list</em><em>[</em><em>int</em><em>]</em><em>, </em><em>optional</em>) – List of strides to apply from “position” to define target genomics intervals (in bp.), by default [0]</p></li>
<li><p><strong>trans_chromosome</strong> (<em>str</em><em>, </em><em>optional</em>) – Chromosomes to consider as trans-chromosomes to define genomic target intervals, by default None</p></li>
<li><p><strong>trans_position</strong> (<em>list</em><em>[</em><em>int</em><em>]</em><em>, </em><em>optional</em>) – Genomic coordinates (0-based) used to defined the trans-chromosomal target genomic interval, by default None</p></li>
<li><p><strong>nb_bins</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of bins to consider on each side of the defined source and target intervals, by default 1</p></li>
<li><p><strong>random</strong> (<em>bool</em><em>, </em><em>optional</em>) – Set wether or not the source and target intervals are defined at random, by default False</p></li>
<li><p><strong>auto</strong> (<em>int</em><em>, </em><em>optional</em>) – Set the number of intervals to get while picking at random, by default None</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save alignments, by default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary of intervals as key : chromosome, value : [(low_limit_0, up_limit_0), (low_limit_1, up_limit_1), …, (low_limit_n, up_limit_n)].</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, list[(int, int)]]</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-hicberg.filter">
<span id="hicberg-filter-module"></span><h2>hicberg.filter module<a class="headerlink" href="#module-hicberg.filter" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="hicberg.filter.filter_library">
<span class="sig-prename descclassname"><span class="pre">hicberg.filter.</span></span><span class="sig-name descname"><span class="pre">filter_library</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/hicberg/filter.html#filter_library"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.filter.filter_library" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.filter.is_N_rich">
<span class="sig-prename descclassname"><span class="pre">hicberg.filter.</span></span><span class="sig-name descname"><span class="pre">is_N_rich</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/hicberg/filter.html#is_N_rich"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.filter.is_N_rich" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.filter.is_low_complexity">
<span class="sig-prename descclassname"><span class="pre">hicberg.filter.</span></span><span class="sig-name descname"><span class="pre">is_low_complexity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/hicberg/filter.html#is_low_complexity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.filter.is_low_complexity" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.filter.is_low_quality">
<span class="sig-prename descclassname"><span class="pre">hicberg.filter.</span></span><span class="sig-name descname"><span class="pre">is_low_quality</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/hicberg/filter.html#is_low_quality"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.filter.is_low_quality" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-hicberg.io">
<span id="hicberg-io-module"></span><h2>hicberg.io module<a class="headerlink" href="#module-hicberg.io" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="hicberg.io.build_matrix">
<span class="sig-prename descclassname"><span class="pre">hicberg.io.</span></span><span class="sig-name descname"><span class="pre">build_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'fragments_fixed_sizes.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pairs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group1.pairs'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">balance</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cpus</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/io.html#build_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.io.build_matrix" title="Link to this definition"></a></dt>
<dd><p>Take table of bins and .pairs file and build a matrix in .cool format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bins</strong> (<em>str</em>) – Path to the bin file.</p></li>
<li><p><strong>pairs</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to pairs file, by default “group1.pairs”</p></li>
<li><p><strong>mode</strong> (<em>bool</em><em>, </em><em>optional</em>) – Choose wether the mode is rescued or un-rescued to construct associated .cool file, by default False</p></li>
<li><p><strong>balance</strong> (<em>bool</em><em>, </em><em>optional</em>) – Set wether or not to balance the matrix, by default False</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the cooler matrix file, by default None, by default None</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.io.build_pairs">
<span class="sig-prename descclassname"><span class="pre">hicberg.io.</span></span><span class="sig-name descname"><span class="pre">build_pairs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam_for</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group1.1.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bam_rev</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group1.2.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bam_for_rescued</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group2.1.rescued.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bam_rev_rescued</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group2.2.rescued.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/io.html#build_pairs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.io.build_pairs" title="Link to this definition"></a></dt>
<dd><p>Build pairs of reads from the aligned reads.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bam_for</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to forward .bam file for the construction of the .pairs equivalent file (non rescued)., by default “group1.1.bam”</p></li>
<li><p><strong>bam_rev</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to reverse .bam file for the construction of the .pairs equivalent file (non rescued)., by default None, by default “group1.2.bam”</p></li>
<li><p><strong>bam_for_rescued</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to forward .bam file for the construction of the .pairs equivalent file (rescued)., by default “group2.1.rescued.bam”</p></li>
<li><p><strong>bam_rev_rescued</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to reverse .bam file for the construction of the .pairs equivalent file (rescued)., by default “group2.2.rescued.bam”</p></li>
<li><p><strong>mode</strong> (<em>bool</em><em>, </em><em>optional</em>) – Choose wether the mode is rescued or unrescued to construct associated .pairs file, by default False</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path where the alignment files (.sam) should be stored, by default None</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.io.create_folder">
<span class="sig-prename descclassname"><span class="pre">hicberg.io.</span></span><span class="sig-name descname"><span class="pre">create_folder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/io.html#create_folder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.io.create_folder" title="Link to this definition"></a></dt>
<dd><p>Creates folder architecture to store results and intermediate files for the full HiC-BERG pipeline.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_name</strong> (<em>str</em>) – Name of the folder to be created.</p></li>
<li><p><strong>force</strong> (<em>bool</em>) – Set if existing folder has to be deleted before folder creation.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – Path where the folder will be created.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Path of the folder created</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.io.load_cooler">
<span class="sig-prename descclassname"><span class="pre">hicberg.io.</span></span><span class="sig-name descname"><span class="pre">load_cooler</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Cooler</span></span></span><a class="reference internal" href="_modules/hicberg/io.html#load_cooler"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.io.load_cooler" title="Link to this definition"></a></dt>
<dd><p>Load cooler matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>matrix</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to a cooler matrix, by default None</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Cooler matrix object</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cooler.Cooler</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.io.load_dictionary">
<span class="sig-prename descclassname"><span class="pre">hicberg.io.</span></span><span class="sig-name descname"><span class="pre">load_dictionary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dictionary</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="reference internal" href="_modules/hicberg/io.html#load_dictionary"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.io.load_dictionary" title="Link to this definition"></a></dt>
<dd><p>Load dictionary save into numpy format (.npy).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dictionary</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to a the dictionary saved in numpy (.npy) format, by default None</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Python native dictionary</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.io.merge_predictions">
<span class="sig-prename descclassname"><span class="pre">hicberg.io.</span></span><span class="sig-name descname"><span class="pre">merge_predictions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clean</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stage</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'prediction'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cpus</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/io.html#merge_predictions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.io.merge_predictions" title="Link to this definition"></a></dt>
<dd><p>Merge predictions of all chunks of ambiguous reads predictions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the fused alignment file, by default None</p></li>
<li><p><strong>clean</strong> (<em>bool</em><em>, </em><em>optional</em>) – Set weither or not to remove temporary chunks, by default True</p></li>
<li><p><strong>stage</strong> (<em>str</em><em>, </em><em>optional</em>) – Set the stage of the merging, by default “prediction”. Can be “prediction” or “classification”</p></li>
<li><p><strong>cpus</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of cpus to use for the merging, by default 1</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.io.tidy_folder">
<span class="sig-prename descclassname"><span class="pre">hicberg.io.</span></span><span class="sig-name descname"><span class="pre">tidy_folder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/io.html#tidy_folder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.io.tidy_folder" title="Link to this definition"></a></dt>
<dd><p>Tidy all the files in the output folder.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the fused alignment file, by default None</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-hicberg.pipeline">
<span id="hicberg-pipeline-module"></span><h2>hicberg.pipeline module<a class="headerlink" href="#module-hicberg.pipeline" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="hicberg.pipeline.check_tool">
<span class="sig-prename descclassname"><span class="pre">hicberg.pipeline.</span></span><span class="sig-name descname"><span class="pre">check_tool</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hicberg/pipeline.html#check_tool"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.pipeline.check_tool" title="Link to this definition"></a></dt>
<dd><p>Check whether <cite>name</cite> is on PATH and marked as executable.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.pipeline.pipeline">
<span class="sig-prename descclassname"><span class="pre">hicberg.pipeline.</span></span><span class="sig-name descname"><span class="pre">pipeline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'sample'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_stage</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'fastq'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exit_stage</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'None'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fq_for</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fq_rev</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sensitivity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'very-sensitive'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_alignment</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mapq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">35</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enzyme</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['DpnII',</span> <span class="pre">'HinfI']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rate</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_chunks</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'full'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kernel_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">11</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">deviation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cpus</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">blacklist</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/pipeline.html#pipeline"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.pipeline.pipeline" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-hicberg.plot">
<span id="hicberg-plot-module"></span><h2>hicberg.plot module<a class="headerlink" href="#module-hicberg.plot" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="hicberg.plot.plot_benchmark">
<span class="sig-prename descclassname"><span class="pre">hicberg.plot.</span></span><span class="sig-name descname"><span class="pre">plot_benchmark</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">original_matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">depleted_matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rescued_matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosomes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/plot.html#plot_benchmark"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.plot.plot_benchmark" title="Link to this definition"></a></dt>
<dd><p>Plot benchmark results (original, depleted and rescued matrices with associated log ratios). One plot per chromosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>original_matrix</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the original matrix, by default None</p></li>
<li><p><strong>rescued_matrix</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the rescued matrix (re-attributed reads), by default None</p></li>
<li><p><strong>chromosomes</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – List of chromosomes to plot, by default None</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to where to save plots, by default None</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.plot.plot_couple_repartition">
<span class="sig-prename descclassname"><span class="pre">hicberg.plot.</span></span><span class="sig-name descname"><span class="pre">plot_couple_repartition</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">forward_bam_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group2.1.rescued.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse_bam_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group2.2.rescued.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/plot.html#plot_couple_repartition"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.plot.plot_couple_repartition" title="Link to this definition"></a></dt>
<dd><p>Plot read couples sizes distribution</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>forward_bam_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to forward .bam alignment file, by default 1.sorted.bam</p></li>
<li><p><strong>reverse_bam_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to reverse .bam alignment file, by default 2.sorted.bam
Minimal read quality under which a Hi-C read pair will not be kept, by default 30</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the plot, by default None</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.plot.plot_coverages">
<span class="sig-prename descclassname"><span class="pre">hicberg.plot.</span></span><span class="sig-name descname"><span class="pre">plot_coverages</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/plot.html#plot_coverages"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.plot.plot_coverages" title="Link to this definition"></a></dt>
<dd><p>Plot coverages of chromosomes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bins</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of the desired bin., by default 2000</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the plot, by default None, by default None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.plot.plot_d1d2">
<span class="sig-prename descclassname"><span class="pre">hicberg.plot.</span></span><span class="sig-name descname"><span class="pre">plot_d1d2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/plot.html#plot_d1d2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.plot.plot_d1d2" title="Link to this definition"></a></dt>
<dd><p>Plot d1d2 law</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the plot, by default None, by default None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.plot.plot_density">
<span class="sig-prename descclassname"><span class="pre">hicberg.plot.</span></span><span class="sig-name descname"><span class="pre">plot_density</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/plot.html#plot_density"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.plot.plot_density" title="Link to this definition"></a></dt>
<dd><p>Plot density maps</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the plots (one plot per chromosome couple), by default None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.plot.plot_laws">
<span class="sig-prename descclassname"><span class="pre">hicberg.plot.</span></span><span class="sig-name descname"><span class="pre">plot_laws</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/plot.html#plot_laws"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.plot.plot_laws" title="Link to this definition"></a></dt>
<dd><p>Plot P(s) patterns laws</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the plot, by default None, by default None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.plot.plot_matrix">
<span class="sig-prename descclassname"><span class="pre">hicberg.plot.</span></span><span class="sig-name descname"><span class="pre">plot_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">unrescued_matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'unrescued_map.cool'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rescued_matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'rescued_map.cool'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">restriction_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'restriction_map.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/plot.html#plot_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.plot.plot_matrix" title="Link to this definition"></a></dt>
<dd><p>Plot matrix with additional trackss</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>unrescued_matrix</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the unrescued map file, by default unrescued_map.cool</p></li>
<li><p><strong>rescued_matrix</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to rescued map file, by default rescued_map.cool</p></li>
<li><p><strong>restriction_map</strong> (<em>dict</em><em>, </em><em>optional</em>) – Restriction map saved as a dictionary like chrom_name : list of restriction sites’ position, by default dist.frag.npy</p></li>
<li><p><strong>genome</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the genome to digest, by default None, by default None</p></li>
<li><p><strong>vmin</strong> (<em>float</em><em>, </em><em>optional</em>) – Inferior limit for the colorscale, by default 0.0</p></li>
<li><p><strong>vmax</strong> (<em>float</em><em>, </em><em>optional</em>) – Superior limit for the colorscale, by default 3.5</p></li>
<li><p><strong>bins</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of the desired bin., by default 2000</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the plot, by default None</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.plot.plot_pattern_reconstruction">
<span class="sig-prename descclassname"><span class="pre">hicberg.plot.</span></span><span class="sig-name descname"><span class="pre">plot_pattern_reconstruction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">original_cool</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rescued_cool</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">case</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/plot.html#plot_pattern_reconstruction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.plot.plot_pattern_reconstruction" title="Link to this definition"></a></dt>
<dd><p>Create a plot of pattern reconstruction quality.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>table</strong> (<em>pd.DataFrame</em><em>, </em><em>optional</em>) – Table containing either true positives, false positives or false negatives patterns, by default None</p></li>
<li><p><strong>original_cool</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the original matrix in .cool format, by default None</p></li>
<li><p><strong>rescued_cool</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the rescued matrix in .cool format, by default None</p></li>
<li><p><strong>chromosome</strong> (<em>str</em><em>, </em><em>optional</em>) – Selected chromosome, by default None</p></li>
<li><p><strong>threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Threshold for pattern score significance, by default 0.0</p></li>
<li><p><strong>case</strong> (<em>str</em><em>, </em><em>optional</em>) – Mode to consider, either true positives, false positives or false negatives, by default “”</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to save plots, by default None</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.plot.plot_trans_ps">
<span class="sig-prename descclassname"><span class="pre">hicberg.plot.</span></span><span class="sig-name descname"><span class="pre">plot_trans_ps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/plot.html#plot_trans_ps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.plot.plot_trans_ps" title="Link to this definition"></a></dt>
<dd><p>Plot P(s) patterns laws</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the plot, by default None, by default None.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-hicberg.statistics">
<span id="hicberg-statistics-module"></span><h2>hicberg.statistics module<a class="headerlink" href="#module-hicberg.statistics" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.attribute_xs">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">attribute_xs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">xs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#attribute_xs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.attribute_xs" title="Link to this definition"></a></dt>
<dd><p>Attibute genomic distance to the corresponding log bin of xs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>xs</strong> (<em>np.ndarray</em><em>[</em><em>int</em><em>]</em>) – Array containing the log bins.</p></li>
<li><p><strong>distance</strong> (<em>int</em>) – Genomic distance in bp.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Index of the corresponding bins where distance has to be attributed.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.choice">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">choice</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hicberg.statistics.choice" title="Link to this definition"></a></dt>
<dd><p>Generates a random sample from a given 1-D array</p>
<div class="versionadded">
<p><span class="versionmodified added">Added in version 1.7.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>New code should use the <code class="docutils literal notranslate"><span class="pre">choice</span></code> method of a <code class="docutils literal notranslate"><span class="pre">default_rng()</span></code>
instance instead; please see the <span class="xref std std-ref">random-quick-start</span>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a</strong> (<em>1-D array-like</em><em> or </em><em>int</em>) – If an ndarray, a random sample is generated from its elements.
If an int, the random sample is generated as if it were <code class="docutils literal notranslate"><span class="pre">np.arange(a)</span></code></p></li>
<li><p><strong>size</strong> (<em>int</em><em> or </em><em>tuple</em><em> of </em><em>ints</em><em>, </em><em>optional</em>) – Output shape.  If the given shape is, e.g., <code class="docutils literal notranslate"><span class="pre">(m,</span> <span class="pre">n,</span> <span class="pre">k)</span></code>, then
<code class="docutils literal notranslate"><span class="pre">m</span> <span class="pre">*</span> <span class="pre">n</span> <span class="pre">*</span> <span class="pre">k</span></code> samples are drawn.  Default is None, in which case a
single value is returned.</p></li>
<li><p><strong>replace</strong> (<em>boolean</em><em>, </em><em>optional</em>) – Whether the sample is with or without replacement. Default is True,
meaning that a value of <code class="docutils literal notranslate"><span class="pre">a</span></code> can be selected multiple times.</p></li>
<li><p><strong>p</strong> (<em>1-D array-like</em><em>, </em><em>optional</em>) – The probabilities associated with each entry in a.
If not given, the sample assumes a uniform distribution over all
entries in <code class="docutils literal notranslate"><span class="pre">a</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>samples</strong> – The generated random samples</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>single item or ndarray</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If a is an int and less than zero, if a or p are not 1-dimensional,
    if a is an array-like of size 0, if p is not a vector of
    probabilities, if a and p have different lengths, or if
    replace=False and the sample size is greater than the population
    size</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">randint</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">shuffle</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">permutation</span></code></p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">random.Generator.choice</span></code></dt><dd><p>which should be used in new code</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>Setting user-specified probabilities through <code class="docutils literal notranslate"><span class="pre">p</span></code> uses a more general but less
efficient sampler than the default. The general sampler produces a different sample
than the optimized sampler even if each element of <code class="docutils literal notranslate"><span class="pre">p</span></code> is 1 / len(a).</p>
<p>Sampling random rows from a 2-D array is not possible with this function,
but is possible with <cite>Generator.choice</cite> through its <code class="docutils literal notranslate"><span class="pre">axis</span></code> keyword.</p>
<p class="rubric">Examples</p>
<p>Generate a uniform random sample from np.arange(5) of size 3:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="go">array([0, 3, 4]) # random</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#This is equivalent to np.random.randint(0,5,3)</span>
</pre></div>
</div>
<p>Generate a non-uniform random sample from np.arange(5) of size 3:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">array([3, 3, 0]) # random</span>
</pre></div>
</div>
<p>Generate a uniform random sample from np.arange(5) of size 3 without
replacement:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">array([3,1,0]) # random</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#This is equivalent to np.random.permutation(np.arange(5))[:3]</span>
</pre></div>
</div>
<p>Generate a non-uniform random sample from np.arange(5) of size
3 without replacement:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">array([2, 3, 0]) # random</span>
</pre></div>
</div>
<p>Any of the above can be repeated with an arbitrary array-like
instead of just integers. For instance:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aa_milne_arr</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;pooh&#39;</span><span class="p">,</span> <span class="s1">&#39;rabbit&#39;</span><span class="p">,</span> <span class="s1">&#39;piglet&#39;</span><span class="p">,</span> <span class="s1">&#39;Christopher&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">aa_milne_arr</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">])</span>
<span class="go">array([&#39;pooh&#39;, &#39;pooh&#39;, &#39;pooh&#39;, &#39;Christopher&#39;, &#39;piglet&#39;], # random</span>
<span class="go">      dtype=&#39;&lt;U11&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.compute_density">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">compute_density</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cooler_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'unrescued_map.cool'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kernel_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">11</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">deviation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#compute_density"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.compute_density" title="Link to this definition"></a></dt>
<dd><p>Create density map from a Hi-C matrix. Return a dictionary where keys are chromosomes names and values are density maps.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cooler_file</strong> (<em>str</em><em>, </em><em>optional</em>) – [description], by default None</p></li>
<li><p><strong>threads</strong> (<em>int</em><em>, </em><em>optional</em>) – [description], by default 2</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – [description], by default None</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.compute_propensity">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">compute_propensity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read_forward</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_reverse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">restriction_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weirds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uncuts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loops</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trans_ps</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coverage</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d1d2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">density_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'full'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#compute_propensity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.compute_propensity" title="Link to this definition"></a></dt>
<dd><p>Compute propensity for read pair to be selected among all plausible pairs related to multi-mapping reads.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read_forward</strong> (<em>pysam.AlignedSegment</em>) – Forward read to compare with the reverse read.</p></li>
<li><p><strong>read_reverse</strong> (<em>pysam.AlignedSegment</em>) – Reverse read to compare with the forward read.</p></li>
<li><p><strong>restriction_map</strong> (<em>dict</em><em>, </em><em>optional</em>) – Restriction map saved as a dictionary like chrom_name : list of restriction sites’ position, by default None</p></li>
<li><p><strong>xs</strong> (<em>dict</em>) – Dictionary containing log binning values for each chromosome.</p></li>
<li><p><strong>weirds</strong> (<em>dict</em>) – Dictionary containing number of weird events considering distance for each chromosome.</p></li>
<li><p><strong>uncuts</strong> (<em>dict</em>) – Dictionary containing number of uncuts events considering distance for each chromosome.</p></li>
<li><p><strong>loops</strong> (<em>dict</em>) – Dictionary containing number of loops events considering distance for each chromosome.</p></li>
<li><p><strong>circular</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the chromosomes to consider as circular, by default None, by default “”.</p></li>
<li><p><strong>trans_ps</strong> (<em>dict</em>) – Dictionary of trans-chromosomal P(s)</p></li>
<li><p><strong>coverage</strong> (<em>dict</em>) – Dictionary containing the coverage of each chromosome.</p></li>
<li><p><strong>bins</strong> (<em>int</em>) – Size of the desired bin, by default 2000</p></li>
<li><p><strong>d1d2</strong> (<em>np.array</em><em>, </em><em>optional</em>) – Distribution of d1d2 values, by default None</p></li>
<li><p><strong>density</strong> (<em>dict</em>) – Dictionary of contact density by chromosome couple.</p></li>
<li><p><strong>mode</strong> (<em>str</em><em>, </em><em>optional</em>) – Mode to use to compute propensity among, by default “full”</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Propensity to use for read couple drawing</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.draw_read_couple">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">draw_read_couple</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">propensities</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#draw_read_couple"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.draw_read_couple" title="Link to this definition"></a></dt>
<dd><p>Draw an index respecting distribution of propensities. This function is used to draw a couple of reads considering the propensity of each couple.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>propensities</strong> (<em>np.array</em>) – Array containing all the propensities of each couple of reads.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Index of the couple of reads drawn.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.generate_coverages">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">generate_coverages</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">forward_bam_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group1.1.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse_bam_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group1.2.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#generate_coverages"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.generate_coverages" title="Link to this definition"></a></dt>
<dd><p>Take a genome and both forward and reverse bam files for unambiguous group and return a dictionary containing the coverage in terms of reads overs chromosomes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the genome file to get coverage on, by default None</p></li>
<li><p><strong>bins</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of the desired bin., by default 2000</p></li>
<li><p><strong>forward_bam_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to forward .bam alignment file, by default None, by default group1.1.bam</p></li>
<li><p><strong>reverse_bam_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to reverse .bam alignment file, by default None, by default group1.2.bam</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the classified alignment files, by default None, by default None</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.generate_d1d2">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">generate_d1d2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">forward_bam_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group1.1.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse_bam_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group1.2.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">restriction_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'restriction_map.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#generate_d1d2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.generate_d1d2" title="Link to this definition"></a></dt>
<dd><p>Compute d1d2 distance laws with the given alignments and restriction map.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>forward_bam_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to forward .bam alignment file, by default None, by default group1.1.bam, by default “group1.1.bam”</p></li>
<li><p><strong>reverse_bam_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to reverse .bam alignment file, by default None, by default group1.1.bam, by default “group1.2.bam”</p></li>
<li><p><strong>restriction_map</strong> (<em>str</em><em>, </em><em>optional</em>) – Restriction map saved as a dictionary like chrom_name : list of restriction sites’ position, by default “dist.frag.npy”</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the dictionary, by default None, by default None</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.generate_density_map">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">generate_density_map</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'unrescued_map.cool'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sigma</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_mads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">ndarray</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#generate_density_map"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.generate_density_map" title="Link to this definition"></a></dt>
<dd><p>Create density map from a Hi-C matrix. Return a dictionary where keys are chromosomes names and values are density maps.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>matrix</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to a cooler matrix, by default None</p></li>
<li><p><strong>rounds</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of times the matrix has to be shaken, by default 1</p></li>
<li><p><strong>magnitude</strong> (<em>float</em><em>, </em><em>optional</em>) – Blending ratio between the native matrix and the shaken one, by default 1.0</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the density map, by default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Density maps as a dictionary where keys are chromosomes names couples as tuples and values are density maps.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, np.ndarray[float]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.generate_density_map_backup">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">generate_density_map_backup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'unrescued_map.cool'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rounds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">magnitude</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">ndarray</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#generate_density_map_backup"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.generate_density_map_backup" title="Link to this definition"></a></dt>
<dd><p>Create density map from a Hi-C matrix. Return a dictionary where keys are chromosomes names and values are density maps.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>matrix</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to a cooler matrix, by default None</p></li>
<li><p><strong>rounds</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of times the matrix has to be shaken, by default 1</p></li>
<li><p><strong>magnitude</strong> (<em>float</em><em>, </em><em>optional</em>) – Blending ratio between the native matrix and the shaken one, by default 1.0</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the density map, by default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Density maps as a dictionary where keys are chromosomes names couples as tuples and values are density maps.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, np.ndarray[float]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.generate_trans_ps">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">generate_trans_ps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'unrescued_map.cool'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_sizes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'chromosome_sizes.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#generate_trans_ps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.generate_trans_ps" title="Link to this definition"></a></dt>
<dd><p>Generate pseudo-P(s) while considering reads originating from different chromosomes.
Pseudo-P(s) are computed as the number of interactions between two chromosomes divided by the product of the length of these two chromosomes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>matrix</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to a cooler matrix, by default “unrescued_map.cool”</p></li>
<li><p><strong>chrom_sizes</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the chromosome sizes dictionary, by default “chromosome_sizes.npy”</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the dictionary, by default None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.generate_xs">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">generate_xs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">ndarray</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#generate_xs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.generate_xs" title="Link to this definition"></a></dt>
<dd><p>Generate xs array for computing P(s). Return xs array which is log space.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chromosome_size</strong> (<em>int</em>) – Size of the chromosome to be binned in bp.</p></li>
<li><p><strong>base</strong> (<em>float</em><em>, </em><em>optional</em>) – Base of the log space., by default 1.1</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Array of log bins related to the chromosome.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray[int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.get_d1d2">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">get_d1d2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read_forward</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_reverse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">restriction_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d1d2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#get_d1d2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.get_d1d2" title="Link to this definition"></a></dt>
<dd><p>Get the d1d2 value of a pair of reads.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read_forward</strong> (<em>pysam.AlignedSegment</em>) – Forward read to compare with the reverse read.</p></li>
<li><p><strong>read_reverse</strong> (<em>pysam.AlignedSegment</em>) – Reverse read to compare with the forward read.</p></li>
<li><p><strong>restriction_map</strong> (<em>dict</em><em>, </em><em>optional</em>) – Restriction map saved as a dictionary like chrom_name : list of restriction sites’ position, by default None</p></li>
<li><p><strong>d1d2</strong> (<em>np.array</em><em>, </em><em>optional</em>) – Distribution of d1d2 values, by default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>d1d2 value</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.get_density">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">get_density</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">read_forward:</span> <span class="pre">~pysam.libcalignedsegment.AlignedSegment,</span> <span class="pre">read_reverse:</span> <span class="pre">~pysam.libcalignedsegment.AlignedSegment,</span> <span class="pre">density_map:</span> <span class="pre">dict[slice((&lt;class</span> <span class="pre">'str'&gt;,</span> <span class="pre">&lt;class</span> <span class="pre">'str'&gt;),</span> <span class="pre">&lt;built-in</span> <span class="pre">function</span> <span class="pre">array&gt;,</span> <span class="pre">None)],</span> <span class="pre">bin_size:</span> <span class="pre">int</span> <span class="pre">=</span> <span class="pre">2000</span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#get_density"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.get_density" title="Link to this definition"></a></dt>
<dd><p>Get density from density map dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read_forward</strong> (<em>pysam.AlignedSegment</em>) – Forward read to get density from.</p></li>
<li><p><strong>read_reverse</strong> (<em>pysam.AlignedSegment</em>) – Reverse read to get density from.</p></li>
<li><p><strong>density_map</strong> (<em>dict</em>) – Dictionary containing density maps for each chromosome couple.</p></li>
<li><p><strong>bin_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Resolution of the matrix on which density map has been computed, by default 2000</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Density corresponding to the pair of reads.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.get_dist_frags">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">get_dist_frags</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">restriction_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rate</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#get_dist_frags"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.get_dist_frags" title="Link to this definition"></a></dt>
<dd><p>Get the distribution of fragments’ distance from a genome distribution .</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the genome, by default None</p></li>
<li><p><strong>restriction_map</strong> (<em>dict</em><em>, </em><em>optional</em>) – Restriction map saved as a dictionary like chrom_name : list of restriction sites’ position, by default None</p></li>
<li><p><strong>circular</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the chromosomes to consider as circular, by default “”</p></li>
<li><p><strong>rate</strong> (<em>float</em><em>, </em><em>optional</em>) – Set the proportion of restriction sites to consider. Avoid memory overflow when restriction maps are very dense, by default 1.0</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the dictionary, by default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary of sub-sampled restriction map with keys as chromosome names and values as lists of restriction sites’ position.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.get_pair_cover">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">get_pair_cover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read_forward</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_reverse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coverage</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#get_pair_cover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.get_pair_cover" title="Link to this definition"></a></dt>
<dd><p>Get the coverage of a pair of reads.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read_forward</strong> (<em>pysam.AlignedSegment</em>) – Forward read to compare with the reverse read.</p></li>
<li><p><strong>read_reverse</strong> (<em>pysam.AlignedSegment</em>) – Reverse read to compare with the forward read.</p></li>
<li><p><strong>coverage</strong> (<em>dict</em>) – Dictionary containing the coverage of each chromosome.</p></li>
<li><p><strong>bins</strong> (<em>int</em>) – Size of the desired bin.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Product of the coverage of the pair of reads.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.get_pair_ps">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">get_pair_ps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read_forward</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_reverse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weirds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uncuts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loops</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#get_pair_ps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.get_pair_ps" title="Link to this definition"></a></dt>
<dd><p>Take two reads and return the P(s) value depending on event type (intra-chromosomal case only).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read_forward</strong> (<em>pysam.AlignedSegment</em>) – Forward read to compare with the reverse read.</p></li>
<li><p><strong>read_reverse</strong> (<em>pysam.AlignedSegment</em>) – Reverse read to compare with the forward read.</p></li>
<li><p><strong>xs</strong> (<em>dict</em>) – Dictionary containing log binning values for each chromosome.</p></li>
<li><p><strong>weirds</strong> (<em>dict</em>) – Dictionary containing number of weird events considering distance for each chromosome.</p></li>
<li><p><strong>uncuts</strong> (<em>dict</em>) – Dictionary containing number of uncuts events considering distance for each chromosome.</p></li>
<li><p><strong>loops</strong> (<em>dict</em>) – Dictionary containing number of loops events considering distance for each chromosome.</p></li>
<li><p><strong>circular</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the chromosomes to consider as circular, by default None, by default “”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>P(s) of the pair considering the event type.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.get_patterns">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">get_patterns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">forward_bam_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group1.1.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse_bam_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group1.2.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'xs.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_sizes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'chromosome_sizes.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">blacklist</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#get_patterns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.get_patterns" title="Link to this definition"></a></dt>
<dd><p>Get the patterns distribution from read pairs alignment. .</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>forward_bam_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to forward .bam alignment file, by default None, by default group1.1.bam, by default “group1.1.bam”, by default “group1.1.bam”</p></li>
<li><p><strong>reverse_bam_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to reverse .bam alignment file, by default None, by default group1.1.bam, by default “group1.1.bam”, by default “group1.2.bam”</p></li>
<li><p><strong>xs</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the dictionary containing the xs values, by default “xs.npy”</p></li>
<li><p><strong>dist_frag</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the dictionary containing the inter fragment distances, by default “dist.frag.npy”</p></li>
<li><p><strong>circular</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the chromosomes to consider as circular, by default “”</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the dictionary, by default None, by default None, by default None</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.get_restriction_map">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">get_restriction_map</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enzyme</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['DpnII']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">ndarray</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#get_restriction_map"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.get_restriction_map" title="Link to this definition"></a></dt>
<dd><p>Get ordered restriction map (including 0 and n) from a chromosome sequence.
Return a dictionary where keys are chromosomes names and values are restrictions sites positions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the genome to digest, by default None, by default None</p></li>
<li><p><strong>enzyme</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Enzyme or list of enzyme to digest the genome with. If integer passed, micro-C mode using Mnase is used, and the integer correspond to the size of nucleosomal fragment, by default None, by default [“DpnII”]</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the plot, by default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary of the product of digestion where keys are chromosomes names and values are restrictions sites positions.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.get_top_pattern">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">get_top_pattern</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#get_top_pattern"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.get_top_pattern" title="Link to this definition"></a></dt>
<dd><p>Get top patterns from a dataframe</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em><em>, </em><em>optional</em>) – Dataframe containing patterns given by Chromosight, by default None</p></li>
<li><p><strong>top</strong> (<em>int</em><em>, </em><em>optional</em>) – Percentage of top patterns to get, by default 10</p></li>
<li><p><strong>chromosome</strong> (<em>str</em><em>, </em><em>optional</em>) – Chromosome to consider, by default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe containing top percentage patterns.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.get_trans_ps">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">get_trans_ps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read_forward</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_reverse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trans_ps</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#get_trans_ps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.get_trans_ps" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read_forward</strong> (<em>pysam.AlignedSegment</em>) – Forward read to compare with the reverse read.</p></li>
<li><p><strong>read_reverse</strong> (<em>pysam.AlignedSegment</em>) – Reverse read to compare with the forward read.</p></li>
<li><p><strong>trans_ps</strong> (<em>dict</em>) – Dictionary of trans-chromosomal P(s)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Trans P(s) value</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.log_bin_genome">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">log_bin_genome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">ndarray</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#log_bin_genome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.log_bin_genome" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.pearson_score">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">pearson_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">original_matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Cooler</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rescued_matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Cooler</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">markers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#pearson_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.pearson_score" title="Link to this definition"></a></dt>
<dd><p>Compute Pearson correlation between concatenated matrix bins which have been deleted and reconstructed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>original_matrix</strong> (<em>cooler.Cooler</em>) – Cooler object containing the original matrix.</p></li>
<li><p><strong>rescued_matrix</strong> (<em>cooler.Cooler</em>) – Cooler object containing the reconstructed matrix.</p></li>
<li><p><strong>markers</strong> (<em>list</em><em>[</em><em>int</em><em>]</em>) – List of markers to consider. Markers are the bins which have been deleted.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Pearson correlation between original and reconstructed matrix.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.statistics.reattribute_reads">
<span class="sig-prename descclassname"><span class="pre">hicberg.statistics.</span></span><span class="sig-name descname"><span class="pre">reattribute_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reads_couple</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">('group2.1.bam',</span> <span class="pre">'group2.2.bam')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">restriction_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'xs.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weirds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'weirds.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uncuts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'uncuts.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loops</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'loops.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trans_ps</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'trans_ps.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coverage</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'coverage.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d1d2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'d1d2.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">density_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'density_map.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'full'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/statistics.html#reattribute_reads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.statistics.reattribute_reads" title="Link to this definition"></a></dt>
<dd><p>Re-attribute multi-mapping (ambiguous) reads considering sets of statistical laws.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>reads_couple</strong> (<em>tuple</em><em>[</em><em>str</em><em>, </em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Paths to ambiguous reads alignment files (.bam), by default (“group2.1.bam”, “group2.2.bam”)</p></li>
<li><p><strong>restriction_map</strong> (<em>dict</em><em>, </em><em>optional</em>) – Restriction map saved as a dictionary like chrom_name : list of restriction sites’ position, by default None</p></li>
<li><p><strong>xs</strong> (<em>dict</em>) – Dictionary containing log binning values for each chromosome.</p></li>
<li><p><strong>weirds</strong> (<em>dict</em>) – Dictionary containing number of weird events considering distance for each chromosome.</p></li>
<li><p><strong>uncuts</strong> (<em>dict</em>) – Dictionary containing number of uncuts events considering distance for each chromosome.</p></li>
<li><p><strong>loops</strong> (<em>dict</em>) – Dictionary containing number of loops events considering distance for each chromosome.</p></li>
<li><p><strong>circular</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the chromosomes to consider as circular, by default None, by default “”.</p></li>
<li><p><strong>trans_ps</strong> (<em>dict</em>) – Dictionary of trans-chromosomal P(s)</p></li>
<li><p><strong>coverage</strong> (<em>dict</em>) – Dictionary containing the coverage of each chromosome.</p></li>
<li><p><strong>bins</strong> (<em>int</em>) – Size of the desired bin, by default 2000</p></li>
<li><p><strong>d1d2</strong> (<em>np.array</em><em>, </em><em>optional</em>) – Distribution of d1d2 values, by default None</p></li>
<li><p><strong>density</strong> (<em>np.array</em><em>, </em><em>optional</em>) – Dictionary containing density maps per chromosome couples as, by default None</p></li>
<li><p><strong>mode</strong> (<em>str</em><em>, </em><em>optional</em>) – Mode to use to compute propensity among, by default “full”</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the re-attributed ambiguous reads alignment files are saved, by default None</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-hicberg.utils">
<span id="hicberg-utils-module"></span><h2>hicberg.utils module<a class="headerlink" href="#module-hicberg.utils" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.bam_iterator">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">bam_iterator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Iterator</span><span class="p"><span class="pre">[</span></span><span class="pre">AlignedSegment</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#bam_iterator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.bam_iterator" title="Link to this definition"></a></dt>
<dd><p>Returns an iterator for the given SAM/BAM file (must be query-sorted).
In each call, the alignments of a single read are yielded as a 3-tuple: (list of primary pysam.AlignedSegment, list of supplementary pysam.AlignedSegment, list of secondary pysam.AlignedSegment).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>bam</strong> (<em>[</em><em>str</em><em>]</em>) – Path to alignment file in .sam or .bam format.</p>
</dd>
<dt class="field-even">Yields<span class="colon">:</span></dt>
<dd class="field-even"><p><em>Iterator[pysam.AlignedSegment]</em> – Yields a list containing pysam AlignmentSegment objects, within which all the reads have the same id.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.block_counter">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">block_counter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">forward_bam_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse_bam_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#block_counter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.block_counter" title="Link to this definition"></a></dt>
<dd><p>Return as a tuple the number of blocks in the forward and reverse bam files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>forward_bam_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to forward .bam alignment file.</p></li>
<li><p><strong>reverse_bam_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to reverse .bam alignment file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Number of blocks in the forward and reverse bam files.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[int, int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.choice">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">choice</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#hicberg.utils.choice" title="Link to this definition"></a></dt>
<dd><p>Generates a random sample from a given 1-D array</p>
<div class="versionadded">
<p><span class="versionmodified added">Added in version 1.7.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>New code should use the <code class="docutils literal notranslate"><span class="pre">choice</span></code> method of a <code class="docutils literal notranslate"><span class="pre">default_rng()</span></code>
instance instead; please see the <span class="xref std std-ref">random-quick-start</span>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a</strong> (<em>1-D array-like</em><em> or </em><em>int</em>) – If an ndarray, a random sample is generated from its elements.
If an int, the random sample is generated as if it were <code class="docutils literal notranslate"><span class="pre">np.arange(a)</span></code></p></li>
<li><p><strong>size</strong> (<em>int</em><em> or </em><em>tuple</em><em> of </em><em>ints</em><em>, </em><em>optional</em>) – Output shape.  If the given shape is, e.g., <code class="docutils literal notranslate"><span class="pre">(m,</span> <span class="pre">n,</span> <span class="pre">k)</span></code>, then
<code class="docutils literal notranslate"><span class="pre">m</span> <span class="pre">*</span> <span class="pre">n</span> <span class="pre">*</span> <span class="pre">k</span></code> samples are drawn.  Default is None, in which case a
single value is returned.</p></li>
<li><p><strong>replace</strong> (<em>boolean</em><em>, </em><em>optional</em>) – Whether the sample is with or without replacement. Default is True,
meaning that a value of <code class="docutils literal notranslate"><span class="pre">a</span></code> can be selected multiple times.</p></li>
<li><p><strong>p</strong> (<em>1-D array-like</em><em>, </em><em>optional</em>) – The probabilities associated with each entry in a.
If not given, the sample assumes a uniform distribution over all
entries in <code class="docutils literal notranslate"><span class="pre">a</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>samples</strong> – The generated random samples</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>single item or ndarray</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If a is an int and less than zero, if a or p are not 1-dimensional,
    if a is an array-like of size 0, if p is not a vector of
    probabilities, if a and p have different lengths, or if
    replace=False and the sample size is greater than the population
    size</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">randint</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">shuffle</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">permutation</span></code></p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">random.Generator.choice</span></code></dt><dd><p>which should be used in new code</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>Setting user-specified probabilities through <code class="docutils literal notranslate"><span class="pre">p</span></code> uses a more general but less
efficient sampler than the default. The general sampler produces a different sample
than the optimized sampler even if each element of <code class="docutils literal notranslate"><span class="pre">p</span></code> is 1 / len(a).</p>
<p>Sampling random rows from a 2-D array is not possible with this function,
but is possible with <cite>Generator.choice</cite> through its <code class="docutils literal notranslate"><span class="pre">axis</span></code> keyword.</p>
<p class="rubric">Examples</p>
<p>Generate a uniform random sample from np.arange(5) of size 3:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="go">array([0, 3, 4]) # random</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#This is equivalent to np.random.randint(0,5,3)</span>
</pre></div>
</div>
<p>Generate a non-uniform random sample from np.arange(5) of size 3:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">array([3, 3, 0]) # random</span>
</pre></div>
</div>
<p>Generate a uniform random sample from np.arange(5) of size 3 without
replacement:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">array([3,1,0]) # random</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#This is equivalent to np.random.permutation(np.arange(5))[:3]</span>
</pre></div>
</div>
<p>Generate a non-uniform random sample from np.arange(5) of size
3 without replacement:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">array([2, 3, 0]) # random</span>
</pre></div>
</div>
<p>Any of the above can be repeated with an arbitrary array-like
instead of just integers. For instance:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aa_milne_arr</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;pooh&#39;</span><span class="p">,</span> <span class="s1">&#39;rabbit&#39;</span><span class="p">,</span> <span class="s1">&#39;piglet&#39;</span><span class="p">,</span> <span class="s1">&#39;Christopher&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">aa_milne_arr</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">])</span>
<span class="go">array([&#39;pooh&#39;, &#39;pooh&#39;, &#39;pooh&#39;, &#39;Christopher&#39;, &#39;piglet&#39;], # random</span>
<span class="go">      dtype=&#39;&lt;U11&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.chunk_bam">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">chunk_bam</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">forward_bam_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group2.1.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse_bam_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group2.2.bam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_chunks</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#chunk_bam"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.chunk_bam" title="Link to this definition"></a></dt>
<dd><p>Split a .bam file into chunks .bam files.
:param forward_bam_file: Path to forward .bam alignment file, by default group2.1.bam
:type forward_bam_file: str, optional
:param reverse_bam_file: Path to reverse .bam alignment file, by default group2.2.bam
:type reverse_bam_file: str, optional
:param nb_chunks: Number of chunks to create, by default 2
:type nb_chunks: int, optional
:param output_dir: Path to the folder where to save the classified alignment files, by default None
:type output_dir: str, optional</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.classify_reads">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">classify_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam_couple</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">('1.sorted.bam',</span> <span class="pre">'2.sorted.bam')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome_sizes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'chromosome_sizes.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mapq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">35</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#classify_reads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.classify_reads" title="Link to this definition"></a></dt>
<dd><dl class="simple">
<dt>Classification of pairs of reads in 2 different groups:</dt><dd><p>Group 0) –&gt; (Unmappable) - files :group0.1.bam and group0.2.bam
Group 1) –&gt; (Uniquely Mapped  Uniquely Mapped) - files :group1.1.bam and group1.2.bam
Group 2) –&gt; (Uniquely Mapped Multi Mapped) or (Multi Mapped  Multi Mapped).- files :group2.1.bam and group2.2.bam</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bam_couple</strong> (<em>tuple</em><em>[</em><em>str</em><em>, </em><em>str</em><em>]</em>) – Tuple containing the paths to the forward and reverse alignment files. By default (“1.sorted.bam”, “2.sorted.bam”)</p></li>
<li><p><strong>chromosome_sizes</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to a chromosome size dictionary save in .npy format, by default chromosome_sizes.npy</p></li>
<li><p><strong>mapq</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimal read quality under which a Hi-C read pair will not be kept, by default 30</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the classified alignment files, by default None</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.detrend_matrix">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">detrend_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">array</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#detrend_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.detrend_matrix" title="Link to this definition"></a></dt>
<dd><p>Detrend a matrix by P(s).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>matrix</strong> (<em>np.array</em>) – Hi-C matrix to detrend.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Detrended Hi-C matrix.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.format_blacklist">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">format_blacklist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">blacklist</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#format_blacklist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.format_blacklist" title="Link to this definition"></a></dt>
<dd><p>Format a blacklist file into a dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>blacklist</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the blacklist file. If set to -1 this is equivalent to None for workflow managers purpose, by default None</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary with chromosome name as key and tuple of start and end as value.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, Tuple[int, int]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.generate_gaussian_kernel">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">generate_gaussian_kernel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sigma</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">array</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#generate_gaussian_kernel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.generate_gaussian_kernel" title="Link to this definition"></a></dt>
<dd><p>Generate a 2D Gaussian kernel of a given size and standard deviation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>size</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of the kernel, by default 1</p></li>
<li><p><strong>sigma</strong> (<em>int</em><em>, </em><em>optional</em>) – Standard deviation to use for the kernel build, by default 2</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>2D Gaussian kernel.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.get_bad_bins">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">get_bad_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_mads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">array</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#get_bad_bins"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.get_bad_bins" title="Link to this definition"></a></dt>
<dd><p>Detect bad bins (poor interacting bins) in a normalized Hi-C matrix and return their indexes.
Bins where the nan sum of interactions is zero  are considered as bad bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>matrix</strong> (<em>Normalized Hi-C matrix to detect bad bins from</em><em>, </em><em>by default None.</em>)</p></li>
<li><p><strong>n_mads</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of median absolute deviations to set poor interacting bins threshold, by default 2</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Indexes of bad bins.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.get_bin_table">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">get_bin_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chrom_sizes_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'chromosome_sizes.npy'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#get_bin_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.get_bin_table" title="Link to this definition"></a></dt>
<dd><p>Create bin table containing start and end position for fixed size bin per chromosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom_sizes_dict</strong> (<em>str</em>) – Path to a dictionary containing chromosome sizes as {chromosome : size} saved in .npy format. By default chromosome_sizes.npy</p></li>
<li><p><strong>bins</strong> (<em>int</em>) – Size of the desired bin, by default 2000.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the dictionary, by default None</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.get_chromosomes_sizes">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">get_chromosomes_sizes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#get_chromosomes_sizes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.get_chromosomes_sizes" title="Link to this definition"></a></dt>
<dd><p>Generate a dictionary save in .npy format where keys are chromosome name and value are size in bp.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the genome, by default None</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the folder where to save the dictionary, by default None</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.get_chunks">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">get_chunks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hicberg/utils.html#get_chunks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.get_chunks" title="Link to this definition"></a></dt>
<dd><p>Return a tuple containing the paths to the forward and reverse chunks.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to get chunks from, by default None</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Tuple containing the paths to the forward and reverse chunks.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple([List[str], List[str]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.get_cis_distance">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">get_cis_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read_forward</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_reverse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#get_cis_distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.get_cis_distance" title="Link to this definition"></a></dt>
<dd><p>Calculate the distance between two reads in the same pairwise alignment .</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read_forward</strong> (<em>pysam.aligned_segment</em>) – Forward read of the pair</p></li>
<li><p><strong>read_reverse</strong> (<em>pysam.AlignedSegment</em>) – Reverse read of the pair</p></li>
<li><p><strong>circular</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the chromosomes to consider as circular, by default None, by default “”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Genomic distance separating the two reads (bp).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.get_local_density">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">get_local_density</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cooler_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(None,</span> <span class="pre">None)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">11</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sigma</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_mads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">array</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#get_local_density"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.get_local_density" title="Link to this definition"></a></dt>
<dd><p>Create density map from a Hi-C matrix. Return a dictionary where keys are chromosomes names and values are density maps.
Density is obtained by getting the local density of each pairwise bin using a gaussian kernel convolution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cooler_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to Hi-C matrix (or sub-matrix) to dget density from, by default None, by default None</p></li>
<li><p><strong>chrom_name</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Tuple containing the sub-matrix to fetch, by default (None, None)</p></li>
<li><p><strong>size</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of the gaussian kernel to use, by default 5</p></li>
<li><p><strong>sigma</strong> (<em>int</em><em>, </em><em>optional</em>) – Standard deviation to use for the kernel, by default 2</p></li>
<li><p><strong>n_mads</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of median absolute deviations to set poor interacting bins threshold, by default 2</p></li>
<li><p><strong>nan_threshold</strong> (<em>bool</em><em>, </em><em>optional</em>) – Set wether or not convolution return nan if not enough value are caught, by default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Map of local contact density.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.get_ordered_reads">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">get_ordered_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read_forward</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_reverse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">AlignedSegment</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AlignedSegment</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#get_ordered_reads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.get_ordered_reads" title="Link to this definition"></a></dt>
<dd><p>Returns the ordered pair of reads in the same chromosome as the two reads .</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read_forward</strong> (<em>pysam.AlignedSegment</em>) – Forward read to compare with the reverse read.</p></li>
<li><p><strong>read_reverse</strong> (<em>pysam.AlignedSegment</em>) – Reverse read to compare with the forward read.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The ordered pair of reads in the same chromosome as the two reads.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[pysam.AlignedSegment, pysam.AlignedSegment]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.is_blacklisted">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">is_blacklisted</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read_forward</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_reverse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">blacklist</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#is_blacklisted"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.is_blacklisted" title="Link to this definition"></a></dt>
<dd><p>Check if a read pair is blacklisted based on a list of coordiantes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read_forward</strong> (<em>pysam.AlignmentSegment</em>) – Forward read of the pair.</p></li>
<li><p><strong>read_reverse</strong> (<em>pysam.AlignedSegment</em>) – Reverse read of the pair.</p></li>
<li><p><strong>blacklist</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em><em>]</em>) – Blacklist of coordinates to check against. Chromsome name as key and tuple of start and end as value.
Chromosome names should be formatted as ‘chr1_A’, ‘chr1_B’, etc. With A and B being the index of the coordinates to blacklist in a given chromosome.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the read pair is blacklisted, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.is_circle">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">is_circle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read_forward</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_reverse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#is_circle"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.is_circle" title="Link to this definition"></a></dt>
<dd><p>Check if two reads are forming a loop pattern .</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read_forward</strong> (<em>pysam.AlignedSegment</em>) – Forward read of the pair</p></li>
<li><p><strong>read_reverse</strong> (<em>pysam.AlignedSegment</em>) – Reverse read of the pair</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the two reads are forming a loop pattern, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.is_duplicated">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">is_duplicated</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#is_duplicated"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.is_duplicated" title="Link to this definition"></a></dt>
<dd><p>Check if read from pysam AlignmentFile is mapping more than once along the genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>read</strong> (<em>pysam.AlignedSegment</em>) – pysam AlignedSegment object.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the read is duplicated i.e. mapping to more than one position.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.is_empty_alignment">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">is_empty_alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alignment_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#is_empty_alignment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.is_empty_alignment" title="Link to this definition"></a></dt>
<dd><p>Check if an alignment file is empty.
If empty, return True, else return False.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>alignment_file</strong> (<em>str</em>) – Path to the alignment file to check.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Return True if the file is empty, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.is_intra_chromosome">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">is_intra_chromosome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read_forward</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_reverse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#is_intra_chromosome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.is_intra_chromosome" title="Link to this definition"></a></dt>
<dd><p>Return True if two reads of a pair came from the same chromosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read_forward</strong> (<em>pysam.AlignedSegment</em>) – Forward read of the pair.</p></li>
<li><p><strong>read_reverse</strong> (<em>pysam.AlignedSegment</em>) – Reverse read of the pair.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the pair is intra-chromosomic, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.is_poor_quality">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">is_poor_quality</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mapq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#is_poor_quality"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.is_poor_quality" title="Link to this definition"></a></dt>
<dd><p>Check if read from pysam AlignmentFile is under mapping quality threshold</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read</strong> (<em>pysam.AlignedSegment</em>) – pysam AlignedSegment object.</p></li>
<li><p><strong>mapq</strong> (<em>int</em>) – Mapping quality threshold.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the read quality is below mapq threshold.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.is_reverse">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">is_reverse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#is_reverse"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.is_reverse" title="Link to this definition"></a></dt>
<dd><p>Check if read from pysam AlignmentFile is reverse</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>read</strong> (<em>pysam.AlignedSegment</em>) – pysam AlignedSegment object.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the read is reverse.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.is_uncut">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">is_uncut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read_forward</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_reverse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#is_uncut"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.is_uncut" title="Link to this definition"></a></dt>
<dd><p>Check if two reads are forming an uncut pattern .</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read_forward</strong> (<em>pysam.AlignedSegment</em>) – Forward read of the pair</p></li>
<li><p><strong>read_reverse</strong> (<em>pysam.AlignedSegment</em>) – Reverse read of the pair</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the two reads are forming an uncut pattern, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.is_unmapped">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">is_unmapped</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#is_unmapped"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.is_unmapped" title="Link to this definition"></a></dt>
<dd><p>Check if read from pysam AlignmentFile is unmapped</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>read</strong> (<em>pysam.AlignedSegment</em>) – pysam AlignedSegment object.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the read is unmapped.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.is_unqualitative">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">is_unqualitative</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#is_unqualitative"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.is_unqualitative" title="Link to this definition"></a></dt>
<dd><p>Check if the read is not qualitative.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>read</strong> (<em>pysam.AlignedSegment</em>) – pysam AlignedSegment object.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the read is not qualitative, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.is_weird">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">is_weird</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read_forward</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_reverse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignedSegment</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#is_weird"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.is_weird" title="Link to this definition"></a></dt>
<dd><p>Check if two reads are forming a weird pattern .</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read_forward</strong> (<em>pysam.AlignedSegment</em>) – Forward read of the pair</p></li>
<li><p><strong>read_reverse</strong> (<em>pysam.AlignedSegment</em>) – Reverse read of the pair</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the two reads are forming a weird pattern, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.mad_smoothing">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">mad_smoothing</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vector</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'auto'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nmads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">ndarray</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#mad_smoothing"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.mad_smoothing" title="Link to this definition"></a></dt>
<dd><p>Apply MAD smoothing to an vector .</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vector</strong> (<em>np.ndarray</em><em>[</em><em>int</em><em>]</em><em>, </em><em>optional</em>) – Data to smooth, by default None</p></li>
<li><p><strong>window_size</strong> (<em>int</em><em> or </em><em>str</em><em>, </em><em>optional</em>) – Size of the window to perform mean sliding average in. Window is center on current value as [current_value - window_size/2] U [current_value + window_size/2], by default “auto”</p></li>
<li><p><strong>nmads</strong> (<em>int</em><em>, </em><em>optional</em>) – number of median absolute deviation tu use, by default 1</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>MAD smoothed vector.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray[int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.max_consecutive_nans">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">max_consecutive_nans</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vector</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#max_consecutive_nans"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.max_consecutive_nans" title="Link to this definition"></a></dt>
<dd><p>Return the maximum number of consecutive NaN values in a vector.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>vector</strong> (<em>np.ndarray</em>) – Vector to get the maximum number of consecutive NaN values from.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Number of maximum consecutive NaN values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.nan_conv">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">nan_conv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kernel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">array</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#nan_conv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.nan_conv" title="Link to this definition"></a></dt>
<dd><p>Custom convolution function that takes into account nan values when convolving.
Used to compute the local density of a Hi-C matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>matrix</strong> (<em>np.array</em><em>, </em><em>optional</em>) – Hi-C matrix to detect bad bins from, by default None</p></li>
<li><p><strong>kernel</strong> (<em>np.array</em><em>, </em><em>optional</em>) – Kernel to use for convolution (dimension must be odd), by default None</p></li>
<li><p><strong>nan_threshold</strong> (<em>bool</em><em>, </em><em>optional</em>) – Set wether or not convolution return nan if not enough value are caught, by default False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Convolution product of the matrix and the kernel.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.replace_consecutive_zeros_with_mean">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">replace_consecutive_zeros_with_mean</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vector</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">ndarray</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#replace_consecutive_zeros_with_mean"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.replace_consecutive_zeros_with_mean" title="Link to this definition"></a></dt>
<dd><p>Replace consecutive zeros in a vector with the mean of the flanking values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>vector</strong> (<em>np.ndarray</em><em>[</em><em>float</em><em>]</em>) – Array to replace consecutive zeros in.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Array with consecutive zeros replaced by the mean of the flanking values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray[float]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.subsample_restriction_map">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">subsample_restriction_map</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">restriction_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rate</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">ndarray</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#subsample_restriction_map"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.subsample_restriction_map" title="Link to this definition"></a></dt>
<dd><p>Subsample a restriction map by a given rate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>restriction_map</strong> (<em>dict</em><em>, </em><em>optional</em>) – Restriction map saved as a dictionary like chrom_name : list of restriction sites’ position, by default None</p></li>
<li><p><strong>rate</strong> (<em>float</em><em>, </em><em>optional</em>) – Set the proportion of restriction sites to consider. Avoid memory overflow when restriction maps are very dense, by default 1.0</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary of sub-sampled restriction map with keys as chromosome names and values as lists of restriction sites’ position.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, np.ndarray[int]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="hicberg.utils.sum_mat_bins">
<span class="sig-prename descclassname"><span class="pre">hicberg.utils.</span></span><span class="sig-name descname"><span class="pre">sum_mat_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">array</span></span></span><a class="reference internal" href="_modules/hicberg/utils.html#sum_mat_bins"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#hicberg.utils.sum_mat_bins" title="Link to this definition"></a></dt>
<dd><p>Adapted from : <a class="reference external" href="https://github.com/koszullab/hicstuff/tree/master/hicstuff">https://github.com/koszullab/hicstuff/tree/master/hicstuff</a>
Compute the sum of matrices bins (i.e. rows or columns) using
only the upper triangle, assuming symmetrical matrices.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mat</strong> (<em>scipy.sparse.coo_matrix</em>) – Contact map in sparse format, either in upper triangle or
full matrix.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>1D array of bin sums.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-hicberg.version">
<span id="hicberg-version-module"></span><h2>hicberg.version module<a class="headerlink" href="#module-hicberg.version" title="Link to this heading"></a></h2>
</section>
<section id="module-hicberg">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-hicberg" title="Link to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="modules.html" class="btn btn-neutral float-left" title="hicberg" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Sébastien Gradit.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>