

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>hicberg.benchmark &mdash; hicberg 0.1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=01f34227"></script>
      <script src="../../_static/doctools.js?v=888ff710"></script>
      <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            hicberg
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../modules.html">hicberg</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorial:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../notebooks/tutorial.html">Hicberg tutorial</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../hicberg.html">hicberg package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">hicberg</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">hicberg.benchmark</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for hicberg.benchmark</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">uuid</span>
<span class="kn">from</span> <span class="nn">os</span> <span class="kn">import</span> <span class="n">mkdir</span> <span class="c1">#path</span>
<span class="kn">from</span> <span class="nn">shutil</span> <span class="kn">import</span> <span class="n">rmtree</span><span class="p">,</span> <span class="n">copy</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span> <span class="nn">functools</span> <span class="kn">import</span> <span class="n">partial</span>

<span class="kn">from</span> <span class="nn">multiprocessing</span> <span class="kn">import</span> <span class="n">Process</span><span class="p">,</span> <span class="n">Pool</span>


<span class="kn">from</span> <span class="nn">glob</span> <span class="kn">import</span> <span class="n">glob</span>
<span class="c1"># import tempfile as tmpf</span>
<span class="c1"># import multiprocessing as mp</span>
<span class="kn">import</span> <span class="nn">subprocess</span> <span class="k">as</span> <span class="nn">sp</span>
<span class="kn">from</span> <span class="nn">datetime</span> <span class="kn">import</span> <span class="n">datetime</span>


<span class="kn">from</span> <span class="nn">itertools</span> <span class="kn">import</span> <span class="n">product</span>

<span class="kn">from</span> <span class="nn">shutil</span> <span class="kn">import</span> <span class="n">rmtree</span>
<span class="kn">import</span> <span class="nn">click</span>
<span class="kn">import</span> <span class="nn">logging</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">cooler</span>

<span class="kn">import</span> <span class="nn">hicberg.io</span> <span class="k">as</span> <span class="nn">hio</span>
<span class="kn">import</span> <span class="nn">hicberg.utils</span> <span class="k">as</span> <span class="nn">hut</span>
<span class="c1"># import hicberg.align as hal</span>
<span class="kn">import</span> <span class="nn">hicberg.statistics</span> <span class="k">as</span> <span class="nn">hst</span>
<span class="kn">import</span> <span class="nn">hicberg.eval</span> <span class="k">as</span> <span class="nn">hev</span>
<span class="kn">import</span> <span class="nn">hicberg.plot</span> <span class="k">as</span> <span class="nn">hpl</span>

<span class="kn">from</span> <span class="nn">hicberg</span> <span class="kn">import</span> <span class="n">logger</span>

<span class="c1"># BAM_FOR = &quot;group1.1_subsampled.bam&quot; </span>
<span class="c1"># BAM_REV = &#39;group1.2_subsampled.bam&#39;</span>
<span class="n">BAM_FOR</span> <span class="o">=</span> <span class="s2">&quot;group1.1.bam&quot;</span> 
<span class="n">BAM_REV</span> <span class="o">=</span> <span class="s1">&#39;group1.2.bam&#39;</span>
<span class="n">BASE_MATRIX</span> <span class="o">=</span> <span class="s2">&quot;original_map.cool&quot;</span>
<span class="n">FRAGMENTS</span> <span class="o">=</span> <span class="s2">&quot;fragments_fixed_sizes.txt&quot;</span>
<span class="n">DIST_FRAGS</span> <span class="o">=</span> <span class="s2">&quot;dist.frag.npy&quot;</span>
<span class="n">CHROMOSOME_SIZES</span> <span class="o">=</span> <span class="s2">&quot;chromosome_sizes.npy&quot;</span>
<span class="n">XS</span> <span class="o">=</span> <span class="s2">&quot;xs.npy&quot;</span>
<span class="n">UNRESCUED_MATRIX</span> <span class="o">=</span> <span class="s2">&quot;unrescued_map.cool&quot;</span>
<span class="n">RESCUED_MATRIX</span> <span class="o">=</span> <span class="s2">&quot;rescued_map.cool&quot;</span>
<span class="n">RESTRICTION_MAP</span> <span class="o">=</span> <span class="s2">&quot;restriction_map.npy&quot;</span>
<span class="n">FORWARD_IN_FILE</span> <span class="o">=</span> <span class="s2">&quot;group1.1.in.bam&quot;</span>
<span class="n">REVERSE_IN_FILE</span> <span class="o">=</span> <span class="s2">&quot;group1.2.in.bam&quot;</span>
<span class="n">FORWARD_OUT_FILE</span> <span class="o">=</span> <span class="s2">&quot;group1.1.out.bam&quot;</span>
<span class="n">REVERSE_OUT_FILE</span> <span class="o">=</span> <span class="s2">&quot;group1.2.out.bam&quot;</span>

<span class="c1"># TODO : Complete docstring</span>
<div class="viewcode-block" id="benchmark">
<a class="viewcode-back" href="../../hicberg.html#hicberg.benchmark.benchmark">[docs]</a>
<span class="k">def</span> <span class="nf">benchmark</span><span class="p">(</span><span class="n">output_dir</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">chromosome</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">position</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">trans_chromosome</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">trans_position</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">strides</span> <span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">mode</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;full&quot;</span><span class="p">,</span> <span class="n">auto</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">kernel_size</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">11</span><span class="p">,</span> <span class="n">deviation</span> <span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">bins</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">circular</span> <span class="p">:</span><span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">genome</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">pattern</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">threshold</span> <span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span> <span class="n">jitter</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">trend</span> <span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">top</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>  <span class="n">force</span> <span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">iterations</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">cpus</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">8</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Performs a benchmark of the HiC-Berg method by simulating the deletion of a genomic region and evaluating the method&#39;s ability to recover the original signal.</span>

<span class="sd">    The benchmark simulates the deletion of a genomic region by removing reads that map to that region from a Hi-C dataset. </span>
<span class="sd">    It then applies the HiC-Berg method to the depleted dataset to try to recover the original signal. </span>
<span class="sd">    The benchmark evaluates the method&#39;s performance by comparing the rescued contact map to the original contact map.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    output_dir : str, optional</span>
<span class="sd">        Path to the output directory.</span>
<span class="sd">    chromosome : str, optional</span>
<span class="sd">        Chromosome to perform the benchmark on.</span>
<span class="sd">    position : int, optional</span>
<span class="sd">        Position of the deletion on the chromosome.</span>
<span class="sd">    trans_chromosome : str, optional</span>
<span class="sd">        Chromosome to consider for trans interactions.</span>
<span class="sd">    trans_position : int, optional</span>
<span class="sd">        Position of the deletion on the trans chromosome.</span>
<span class="sd">    strides : list[int], optional</span>
<span class="sd">        List of strides to use for the deletion.</span>
<span class="sd">    mode : str, optional</span>
<span class="sd">        Mode of the HiC-Berg method to use. Can be &quot;full&quot; or &quot;density&quot;.</span>
<span class="sd">    auto : int, optional</span>
<span class="sd">        Automatically determine the size of the deletion based on the given number of bins.</span>
<span class="sd">    kernel_size : int, optional</span>
<span class="sd">        Size of the kernel to use for the density estimation.</span>
<span class="sd">    deviation : float, optional</span>
<span class="sd">        Standard deviation of the kernel to use for the density estimation.</span>
<span class="sd">    bins : int, optional</span>
<span class="sd">        Number of bins to use for the deletion.</span>
<span class="sd">    circular : str, optional</span>
<span class="sd">        Whether the chromosome is circular.</span>
<span class="sd">    genome : str, optional</span>
<span class="sd">        Genome assembly to use.</span>
<span class="sd">    pattern : str, optional</span>
<span class="sd">        Pattern to use for Chromosight pre-call.</span>
<span class="sd">    threshold : float, optional</span>
<span class="sd">        Threshold to use for Chromosight pre-call.</span>
<span class="sd">    jitter : int, optional</span>
<span class="sd">        Jitter to use for Chromosight pre-call.</span>
<span class="sd">    trend : bool, optional</span>
<span class="sd">        Whether to use trend for Chromosight pre-call.</span>
<span class="sd">    top : int, optional</span>
<span class="sd">        Top patterns to use for Chromosight pre-call.</span>
<span class="sd">    force : bool, optional</span>
<span class="sd">        Whether to force the benchmark to run even if the output directory already exists.</span>
<span class="sd">    iterations : int, optional</span>
<span class="sd">        Number of iterations to run the HiC-Berg method.</span>
<span class="sd">    cpus : int, optional</span>
<span class="sd">        Number of CPUs to use for parallel processing.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        If the output directory does not exist.</span>
<span class="sd">    ValueError</span>
<span class="sd">        If the restriction map file does not exist.</span>
<span class="sd">    &quot;&quot;&quot;</span>    
    <span class="c1"># logger.addHandler(&#39;hicberg_benchmark.log&#39;)</span>

    <span class="n">args</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span>
    <span class="c1"># Keep track of the arguments used</span>
    <span class="k">for</span> <span class="n">arg</span> <span class="ow">in</span> <span class="n">args</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">arg</span><span class="p">,</span> <span class="n">args</span><span class="p">[</span><span class="n">arg</span><span class="p">])</span>

    <span class="n">learning_status</span> <span class="o">=</span> <span class="kc">False</span> <span class="c1">#False bckp</span>
    <span class="n">picking_status</span> <span class="o">=</span> <span class="kc">False</span> <span class="c1"># False bckp</span>

    <span class="c1"># Setting files paths</span>
    <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>
    <span class="n">output_path_chunks</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="s2">&quot;chunks&quot;</span><span class="p">)</span>
    <span class="n">restriction_map_path</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="n">RESTRICTION_MAP</span>
    <span class="n">bam_for_path</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="n">BAM_FOR</span>
    <span class="n">bam_rev_path</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="n">BAM_REV</span>

    <span class="n">fragments_path</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="n">FRAGMENTS</span> 
    <span class="n">chromosome_size_path</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="n">CHROMOSOME_SIZES</span>
    <span class="n">dist_frags_path</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="n">DIST_FRAGS</span>
    <span class="n">xs_path</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="n">XS</span>


    <span class="c1"># Define unique id to keep track of the experiments</span>
    <span class="n">id_tag</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">uuid</span><span class="o">.</span><span class="n">uuid4</span><span class="p">())[:</span><span class="mi">8</span><span class="p">]</span>

    <span class="c1"># Define output child directory</span>
    <span class="n">output_uniq_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">id_tag</span><span class="p">)</span>
    <span class="n">output_data_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">id_tag</span><span class="p">,</span> <span class="s2">&quot;data&quot;</span><span class="p">)</span>
    <span class="n">output_plot_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">id_tag</span><span class="p">,</span> <span class="s2">&quot;plots&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">output_uniq_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">mkdir</span><span class="p">(</span><span class="n">output_uniq_path</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">output_data_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">mkdir</span><span class="p">(</span><span class="n">output_data_path</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">output_plot_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">mkdir</span><span class="p">(</span><span class="n">output_plot_path</span><span class="p">)</span>


    <span class="k">if</span> <span class="ow">not</span> <span class="n">output_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Output directory </span><span class="si">{</span><span class="n">output_dir</span><span class="si">}</span><span class="s2"> does not exist.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">restriction_map_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Restriction map file </span><span class="si">{</span><span class="n">restriction_map_path</span><span class="si">}</span><span class="s2"> does not exist. PLease provide an existing restriction map file.&quot;</span><span class="p">)</span>
    
    <span class="c1"># Define file to store results</span>
    <span class="n">header</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;id</span><span class="se">\t</span><span class="s2">date</span><span class="se">\t</span><span class="s2">chrom</span><span class="se">\t</span><span class="s2">pos</span><span class="se">\t</span><span class="s2">stride</span><span class="se">\t</span><span class="s2">trans_chrom</span><span class="se">\t</span><span class="s2">trans_pos</span><span class="se">\t</span><span class="s2">auto</span><span class="se">\t</span><span class="s2">bins</span><span class="se">\t</span><span class="s2">mode</span><span class="se">\t</span><span class="s2">nb_reads</span><span class="se">\t</span><span class="s2">pattern</span><span class="se">\t</span><span class="s2">precision</span><span class="se">\t</span><span class="s2">recall</span><span class="se">\t</span><span class="s2">f1_score</span><span class="se">\t</span><span class="s2">score</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;benchmark.csv&quot;</span>

    <span class="c1"># Copy files to data_path</span>
    <span class="n">copy</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span><span class="n">BASE_MATRIX</span><span class="p">,</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">BASE_MATRIX</span><span class="p">)</span>
    <span class="n">copy</span><span class="p">(</span><span class="n">bam_for_path</span><span class="p">,</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">BAM_FOR</span><span class="p">)</span>
    <span class="n">copy</span><span class="p">(</span><span class="n">bam_rev_path</span><span class="p">,</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">BAM_REV</span><span class="p">)</span>
    <span class="n">copy</span><span class="p">(</span><span class="n">restriction_map_path</span><span class="p">,</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">RESTRICTION_MAP</span><span class="p">)</span>
    <span class="n">copy</span><span class="p">(</span><span class="n">fragments_path</span><span class="p">,</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">FRAGMENTS</span><span class="p">)</span>
    <span class="n">copy</span><span class="p">(</span><span class="n">chromosome_size_path</span><span class="p">,</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">CHROMOSOME_SIZES</span><span class="p">)</span>
    <span class="n">copy</span><span class="p">(</span><span class="n">dist_frags_path</span><span class="p">,</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">DIST_FRAGS</span><span class="p">)</span>
    <span class="n">copy</span><span class="p">(</span><span class="n">xs_path</span><span class="p">,</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">XS</span><span class="p">)</span>

    <span class="n">base_matrix</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_cooler</span><span class="p">(</span><span class="n">output_data_path</span> <span class="o">/</span> <span class="n">BASE_MATRIX</span><span class="p">)</span>
    <span class="n">bin_size</span> <span class="o">=</span> <span class="n">base_matrix</span><span class="o">.</span><span class="n">binsize</span>
    <span class="n">flag_file</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;flag.txt&quot;</span>



    <span class="n">forward_in_path</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">FORWARD_IN_FILE</span>
    <span class="n">reverse_in_path</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">REVERSE_IN_FILE</span>
    <span class="n">forward_out_path</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">FORWARD_OUT_FILE</span>
    <span class="n">reverse_out_path</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">REVERSE_OUT_FILE</span>

    <span class="c1">## Chromosomsight pre-call       </span>
    <span class="c1"># reformat inputs</span>
    <span class="k">if</span> <span class="n">pattern</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">pattern</span> <span class="o">==</span> <span class="s2">&quot;-1&quot;</span><span class="p">:</span>
    
        <span class="n">chromosome</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">chromosome</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)]</span>

        <span class="k">if</span> <span class="n">trans_chromosome</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">trans_chromosome</span> <span class="o">!=</span> <span class="s2">&quot;-1&quot;</span><span class="p">:</span>
            <span class="n">trans_chromosome</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">t</span><span class="p">)</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">trans_chromosome</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)]</span>

        <span class="k">else</span> <span class="p">:</span> 
            <span class="n">trans_chromosome</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="n">trans_position</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">trans_position</span> <span class="o">!=</span> <span class="s2">&quot;-1&quot;</span><span class="p">:</span> <span class="c1"># -1 usefull when not using trans cases with benchmark calling through Snakemake</span>
            <span class="n">trans_position</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">trans_position</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)]</span>

        <span class="k">else</span> <span class="p">:</span>
            <span class="n">trans_position</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="n">nb_bins</span> <span class="o">=</span> <span class="n">bins</span>

        <span class="n">strides</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">strides</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)]</span>

    <span class="k">elif</span> <span class="n">pattern</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">pattern</span> <span class="o">!=</span> <span class="s2">&quot;-1&quot;</span><span class="p">:</span> <span class="c1"># Pattern based benchmarking</span>
        <span class="n">pre_recall_cmd</span> <span class="o">=</span> <span class="n">hev</span><span class="o">.</span><span class="n">chromosight_cmd_generator</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">BASE_MATRIX</span><span class="p">,</span> <span class="n">pattern</span> <span class="o">=</span> <span class="n">pattern</span><span class="p">,</span> <span class="n">untrend</span> <span class="o">=</span> <span class="n">trend</span><span class="p">,</span> <span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Starting Chromosight pre-call&quot;</span><span class="p">)</span>
        <span class="n">sp</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">pre_recall_cmd</span><span class="p">,</span> <span class="n">shell</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

        <span class="n">chromosome</span> <span class="o">=</span> <span class="n">chromosome</span>

        <span class="k">if</span> <span class="n">trans_chromosome</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">trans_chromosome</span> <span class="o">!=</span> <span class="s2">&quot;-1&quot;</span><span class="p">:</span>
            <span class="n">trans_chromosome</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">t</span><span class="p">)</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">trans_chromosome</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)]</span>

        <span class="k">else</span> <span class="p">:</span> 
            <span class="n">trans_chromosome</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="n">trans_position</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">trans_position</span> <span class="o">!=</span> <span class="s2">&quot;-1&quot;</span><span class="p">:</span> <span class="c1"># -1 usefull when not using trans cases with benchmark calling through Snakemake</span>
            <span class="n">trans_position</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">trans_position</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)]</span>

        <span class="k">else</span> <span class="p">:</span>
            <span class="n">trans_position</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">hev</span><span class="o">.</span><span class="n">get_top_pattern</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="s2">&quot;original.tsv&quot;</span><span class="p">,</span> <span class="n">top</span> <span class="o">=</span> <span class="n">top</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="n">threshold</span><span class="p">,</span> <span class="n">chromosome</span> <span class="o">=</span> <span class="n">chromosome</span><span class="p">)</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;start1&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

        <span class="n">position</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">start1</span> <span class="c1"># select top score pattern</span>

        <span class="k">if</span> <span class="n">strides</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">strides</span> <span class="o">==</span> <span class="s2">&quot;-1&quot;</span><span class="p">:</span>

            <span class="n">strides</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">start1</span> <span class="o">-</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;start1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span>
        
        <span class="k">else</span> <span class="p">:</span> 

            <span class="n">strides</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">strides</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)]</span>
        
        <span class="n">nb_bins</span> <span class="o">=</span> <span class="n">bins</span>

    <span class="k">for</span> <span class="n">sub_mode</span> <span class="ow">in</span> <span class="n">mode</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">):</span>

        <span class="n">precision</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">recall</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">f1_score</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># Pick reads</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">picking_status</span><span class="p">:</span>
            
            <span class="c1"># # chunk bam files</span>
            <span class="n">forward_chunks</span><span class="p">,</span> <span class="n">reverse_chunks</span> <span class="o">=</span> <span class="n">hut</span><span class="o">.</span><span class="n">get_chunks</span><span class="p">(</span><span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_path</span><span class="o">.</span><span class="n">as_posix</span><span class="p">())</span>

            <span class="c1"># Multithread part </span>

            <span class="n">coordinates</span> <span class="o">=</span> <span class="n">hev</span><span class="o">.</span><span class="n">generate_dict_coordinates</span><span class="p">(</span><span class="n">matrix_file</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="n">BASE_MATRIX</span><span class="p">,</span> <span class="n">position</span> <span class="o">=</span> <span class="n">position</span><span class="p">,</span> <span class="n">chromosome</span> <span class="o">=</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">strides</span> <span class="o">=</span> <span class="n">strides</span><span class="p">,</span> <span class="n">trans_chromosome</span> <span class="o">=</span> <span class="n">trans_chromosome</span><span class="p">,</span> <span class="n">trans_position</span> <span class="o">=</span> <span class="n">trans_position</span><span class="p">,</span> <span class="n">auto</span> <span class="o">=</span> <span class="n">auto</span><span class="p">,</span> <span class="n">nb_bins</span> <span class="o">=</span> <span class="n">nb_bins</span><span class="p">,</span> <span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">)</span>

            <span class="c1"># TODO : pass coordinates to select_reads_multithreads to avoid seeding problems through multithreading</span>

            <span class="c1"># Pick reads reads</span>
            <span class="k">with</span> <span class="n">Pool</span><span class="p">(</span><span class="n">processes</span> <span class="o">=</span> <span class="n">cpus</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span> <span class="c1"># cpus</span>

                <span class="n">res</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">partial</span><span class="p">(</span><span class="n">hev</span><span class="o">.</span><span class="n">select_reads_multithreads</span><span class="p">,</span> <span class="n">interval_dictionary</span> <span class="o">=</span>  <span class="n">coordinates</span><span class="p">,</span> <span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">),</span>
                <span class="nb">zip</span><span class="p">(</span><span class="n">forward_chunks</span><span class="p">,</span> <span class="n">reverse_chunks</span><span class="p">))</span>
                <span class="n">pool</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
                <span class="n">pool</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>

            <span class="c1"># Get dictionrary of intervals from Pool</span>
            <span class="n">intervals_dictionary</span> <span class="o">=</span> <span class="n">coordinates</span>

            <span class="n">hio</span><span class="o">.</span><span class="n">merge_predictions</span><span class="p">(</span><span class="n">output_dir</span>  <span class="o">=</span> <span class="n">output_data_path</span><span class="p">,</span> <span class="n">clean</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">stage</span> <span class="o">=</span> <span class="s2">&quot;benchmark&quot;</span><span class="p">,</span> <span class="n">cpus</span> <span class="o">=</span> <span class="n">cpus</span><span class="p">)</span>

            <span class="c1"># TODO : put aside in function</span>
            <span class="n">indexes</span> <span class="o">=</span> <span class="n">hev</span><span class="o">.</span><span class="n">get_bin_indexes</span><span class="p">(</span><span class="n">matrix</span> <span class="o">=</span> <span class="n">base_matrix</span><span class="p">,</span> <span class="n">dictionary</span> <span class="o">=</span> <span class="n">intervals_dictionary</span><span class="p">,</span> <span class="p">)</span>
            <span class="n">picking_status</span> <span class="o">=</span> <span class="kc">True</span>


            <span class="c1"># Get corresponding indexes to the duplicated reads coordinates.</span>
            <span class="c1"># Re-build pairs and cooler matrix </span>
            <span class="n">hio</span><span class="o">.</span><span class="n">build_pairs</span><span class="p">(</span><span class="n">bam_for</span> <span class="o">=</span> <span class="n">forward_out_path</span><span class="p">,</span> <span class="n">bam_rev</span> <span class="o">=</span> <span class="n">reverse_out_path</span><span class="p">,</span> <span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">)</span>
            <span class="n">hio</span><span class="o">.</span><span class="n">build_matrix</span><span class="p">(</span><span class="n">balance</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">)</span>

        <span class="n">unrescued_map_path</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="n">UNRESCUED_MATRIX</span>
        
        <span class="c1"># Reattribute reads from inner group</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">learning_status</span> <span class="p">:</span> 
            <span class="c1">## Compute statistics</span>
            <span class="n">p1</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">target</span> <span class="o">=</span> <span class="n">hst</span><span class="o">.</span><span class="n">get_patterns</span><span class="p">,</span> <span class="n">kwargs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">forward_bam_file</span>  <span class="o">=</span> <span class="n">forward_out_path</span><span class="p">,</span> <span class="n">reverse_bam_file</span> <span class="o">=</span> <span class="n">reverse_out_path</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="n">circular</span><span class="p">,</span> <span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">))</span>
            <span class="n">p2</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">target</span> <span class="o">=</span> <span class="n">hst</span><span class="o">.</span><span class="n">generate_trans_ps</span><span class="p">,</span> <span class="n">kwargs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">))</span>
            <span class="n">p3</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">target</span> <span class="o">=</span> <span class="n">hst</span><span class="o">.</span><span class="n">generate_coverages</span><span class="p">,</span> <span class="n">kwargs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">forward_bam_file</span> <span class="o">=</span> <span class="n">forward_out_path</span><span class="p">,</span> <span class="n">reverse_bam_file</span>  <span class="o">=</span> <span class="n">reverse_out_path</span><span class="p">,</span> <span class="n">genome</span> <span class="o">=</span> <span class="n">genome</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="n">bin_size</span><span class="p">,</span> <span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">))</span>
            <span class="n">p4</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">target</span> <span class="o">=</span> <span class="n">hst</span><span class="o">.</span><span class="n">generate_d1d2</span><span class="p">,</span> <span class="n">kwargs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">forward_bam_file</span> <span class="o">=</span> <span class="n">forward_out_path</span><span class="p">,</span> <span class="n">reverse_bam_file</span>  <span class="o">=</span> <span class="n">reverse_out_path</span><span class="p">,</span> <span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">))</span>
            
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Full mode selected. Learning step will be performed.&quot;</span><span class="p">)</span>

            <span class="c1"># Launch processes</span>
            <span class="k">for</span> <span class="n">process</span> <span class="ow">in</span> <span class="p">[</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">,</span> <span class="n">p3</span><span class="p">,</span> <span class="n">p4</span><span class="p">]:</span>
                <span class="n">process</span><span class="o">.</span><span class="n">start</span><span class="p">()</span>

            <span class="c1"># for process in [p1, p2, p3, p4, p5]:</span>
            <span class="k">for</span> <span class="n">process</span> <span class="ow">in</span> <span class="p">[</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">,</span> <span class="n">p3</span><span class="p">,</span> <span class="n">p4</span><span class="p">]:</span>
                <span class="n">process</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Learning step completed&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="s2">&quot;density&quot;</span> <span class="ow">in</span> <span class="n">mode</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">):</span>
                <span class="n">hst</span><span class="o">.</span><span class="n">compute_density</span><span class="p">(</span><span class="n">cooler_file</span> <span class="o">=</span> <span class="n">UNRESCUED_MATRIX</span><span class="p">,</span> <span class="n">kernel_size</span> <span class="o">=</span> <span class="n">kernel_size</span><span class="p">,</span> <span class="n">deviation</span> <span class="o">=</span> <span class="n">deviation</span><span class="p">,</span> <span class="n">threads</span> <span class="o">=</span> <span class="n">cpus</span><span class="p">,</span> <span class="n">output_dir</span>  <span class="o">=</span> <span class="n">output_data_path</span><span class="p">)</span>

        <span class="n">learning_status</span> <span class="o">=</span> <span class="kc">True</span>


        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">iterations</span><span class="p">):</span>
            <span class="c1"># Reattribute reads</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Re-attributing reads&quot;</span><span class="p">)</span>
            
            <span class="c1"># Get chunk_for_*.in.bam/chunk_rev_*.in.bam</span>

            <span class="n">forward_chunks</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">glob</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">output_data_path</span> <span class="o">/</span> <span class="s2">&quot;chunk_for_*.in.bam&quot;</span><span class="p">)))</span>
            <span class="n">reverse_chunks</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">glob</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">output_data_path</span> <span class="o">/</span> <span class="s2">&quot;chunk_rev_*.in.bam&quot;</span><span class="p">)))</span>

            <span class="c1"># Check if chunks are empty</span>
            <span class="k">for</span> <span class="n">forward_chunk</span><span class="p">,</span> <span class="n">reverse_chunk</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">forward_chunks</span><span class="p">,</span> <span class="n">reverse_chunks</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_empty_alignment</span><span class="p">(</span><span class="n">forward_chunk</span><span class="p">)</span> <span class="ow">or</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_empty_alignment</span><span class="p">(</span><span class="n">reverse_chunk</span><span class="p">):</span>
                    <span class="n">forward_chunks</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">forward_chunk</span><span class="p">)</span>
                    <span class="n">reverse_chunks</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">reverse_chunk</span><span class="p">)</span>

            <span class="c1"># Reattribute reads</span>
            <span class="k">with</span> <span class="n">Pool</span><span class="p">(</span><span class="n">processes</span> <span class="o">=</span> <span class="n">cpus</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span> <span class="c1"># cpus</span>

                <span class="n">res</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">partial</span><span class="p">(</span><span class="n">hst</span><span class="o">.</span><span class="n">reattribute_reads</span><span class="p">,</span> <span class="n">mode</span> <span class="o">=</span> <span class="n">sub_mode</span><span class="p">,</span> <span class="n">restriction_map</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">RESTRICTION_MAP</span><span class="p">,</span> <span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">),</span>
                <span class="nb">zip</span><span class="p">(</span><span class="n">forward_chunks</span><span class="p">,</span> <span class="n">reverse_chunks</span><span class="p">))</span>
                <span class="n">pool</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
                <span class="n">pool</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>


            <span class="c1"># hst.reattribute_reads(reads_couple = (forward_in_path, reverse_in_path), mode = sub_mode, output_dir = output_data_path)</span>
            <span class="n">hio</span><span class="o">.</span><span class="n">merge_predictions</span><span class="p">(</span><span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">,</span> <span class="n">clean</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">cpus</span> <span class="o">=</span> <span class="n">cpus</span><span class="p">)</span>

            

            <span class="n">hio</span><span class="o">.</span><span class="n">build_pairs</span><span class="p">(</span><span class="n">bam_for</span>  <span class="o">=</span> <span class="s2">&quot;group1.1.out.bam&quot;</span><span class="p">,</span> <span class="n">bam_rev</span> <span class="o">=</span> <span class="s2">&quot;group1.2.out.bam&quot;</span><span class="p">,</span> <span class="n">bam_for_rescued</span>  <span class="o">=</span> <span class="s2">&quot;group2.1.rescued.bam&quot;</span><span class="p">,</span> <span class="n">bam_rev_rescued</span> <span class="o">=</span> <span class="s2">&quot;group2.2.rescued.bam&quot;</span><span class="p">,</span> <span class="n">mode</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">)</span>
            <span class="n">hio</span><span class="o">.</span><span class="n">build_matrix</span><span class="p">(</span><span class="n">mode</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">balance</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">)</span>

            <span class="n">rescued_matrix</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_cooler</span><span class="p">(</span><span class="n">output_data_path</span> <span class="o">/</span> <span class="n">RESCUED_MATRIX</span><span class="p">)</span>

            <span class="n">rescued_matrix_path</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">RESCUED_MATRIX</span>

            <span class="n">pearson</span> <span class="o">=</span> <span class="n">hst</span><span class="o">.</span><span class="n">pearson_score</span><span class="p">(</span><span class="n">original_matrix</span> <span class="o">=</span> <span class="n">base_matrix</span><span class="p">,</span> <span class="n">rescued_matrix</span> <span class="o">=</span> <span class="n">rescued_matrix</span> <span class="p">,</span> <span class="n">markers</span> <span class="o">=</span> <span class="n">indexes</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">pattern</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">pattern</span> <span class="o">==</span> <span class="s2">&quot;-1&quot;</span><span class="p">:</span>
                <span class="n">chromosome_set</span> <span class="o">=</span> <span class="p">[</span><span class="o">*</span><span class="n">chromosome</span><span class="p">,</span> <span class="o">*</span><span class="n">trans_chromosome</span><span class="p">]</span> <span class="k">if</span> <span class="n">trans_chromosome</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">chromosome</span>
            
            <span class="k">else</span> <span class="p">:</span> 
                <span class="n">chromosome_set</span> <span class="o">=</span> <span class="p">[</span><span class="n">chromosome</span><span class="p">,</span> <span class="o">*</span><span class="n">trans_chromosome</span><span class="p">]</span> <span class="k">if</span> <span class="n">trans_chromosome</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">chromosome</span>

                <span class="n">rescued_matrix_path</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">RESCUED_MATRIX</span>
                <span class="n">post_recall_cmd</span> <span class="o">=</span> <span class="n">hev</span><span class="o">.</span><span class="n">chromosight_cmd_generator</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="n">rescued_matrix_path</span><span class="p">,</span> <span class="n">pattern</span> <span class="o">=</span> <span class="n">pattern</span><span class="p">,</span> <span class="n">untrend</span> <span class="o">=</span> <span class="n">trend</span><span class="p">,</span> <span class="n">mode</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">)</span>

                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Starting Chromosight post-call&quot;</span><span class="p">)</span>
                <span class="n">sp</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">post_recall_cmd</span><span class="p">,</span> <span class="n">shell</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
                
                <span class="c1"># TODO : move to top</span>

                <span class="n">df_original</span> <span class="o">=</span> <span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;original.tsv&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">as_posix</span><span class="p">()</span>
                <span class="n">df_rescued</span> <span class="o">=</span> <span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;rescued.tsv&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">as_posix</span><span class="p">()</span>

                <span class="n">true_positives</span> <span class="o">=</span> <span class="n">hev</span><span class="o">.</span><span class="n">get_TP_table</span><span class="p">(</span><span class="n">df_pattern</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="s2">&quot;original.tsv&quot;</span><span class="p">,</span> <span class="n">df_pattern_recall</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="s2">&quot;rescued.tsv&quot;</span><span class="p">,</span> <span class="n">chromosome</span> <span class="o">=</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">bin_size</span> <span class="o">=</span> <span class="n">bin_size</span><span class="p">,</span> <span class="n">jitter</span> <span class="o">=</span> <span class="n">jitter</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="n">threshold</span><span class="p">)</span>
                <span class="n">false_positives</span> <span class="o">=</span> <span class="n">hev</span><span class="o">.</span><span class="n">get_FP_table</span><span class="p">(</span><span class="n">df_pattern</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="s2">&quot;original.tsv&quot;</span><span class="p">,</span> <span class="n">df_pattern_recall</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="s2">&quot;rescued.tsv&quot;</span><span class="p">,</span> <span class="n">chromosome</span> <span class="o">=</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">bin_size</span> <span class="o">=</span> <span class="n">bin_size</span><span class="p">,</span> <span class="n">jitter</span> <span class="o">=</span> <span class="n">jitter</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="n">threshold</span><span class="p">)</span>
                <span class="n">false_negatives</span> <span class="o">=</span> <span class="n">hev</span><span class="o">.</span><span class="n">get_FN_table</span><span class="p">(</span><span class="n">df_pattern</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="s2">&quot;original.tsv&quot;</span><span class="p">,</span> <span class="n">df_pattern_recall</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="s2">&quot;rescued.tsv&quot;</span><span class="p">,</span> <span class="n">chromosome</span> <span class="o">=</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">bin_size</span> <span class="o">=</span> <span class="n">bin_size</span><span class="p">,</span> <span class="n">jitter</span> <span class="o">=</span> <span class="n">jitter</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="n">threshold</span><span class="p">)</span>

                <span class="c1"># Get scores</span>
                <span class="n">precision</span> <span class="o">=</span> <span class="n">hev</span><span class="o">.</span><span class="n">get_precision</span><span class="p">(</span><span class="n">df_pattern</span> <span class="o">=</span> <span class="n">df_original</span><span class="p">,</span> <span class="n">df_pattern_recall</span> <span class="o">=</span> <span class="n">df_rescued</span><span class="p">,</span> <span class="n">chromosome</span> <span class="o">=</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">bin_size</span> <span class="o">=</span> <span class="n">bin_size</span><span class="p">,</span> <span class="n">jitter</span> <span class="o">=</span> <span class="n">jitter</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="n">threshold</span><span class="p">)</span>
                <span class="n">recall</span> <span class="o">=</span> <span class="n">hev</span><span class="o">.</span><span class="n">get_recall</span><span class="p">(</span><span class="n">df_pattern</span> <span class="o">=</span> <span class="n">df_original</span><span class="p">,</span> <span class="n">df_pattern_recall</span> <span class="o">=</span> <span class="n">df_rescued</span><span class="p">,</span> <span class="n">chromosome</span> <span class="o">=</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">bin_size</span> <span class="o">=</span> <span class="n">bin_size</span><span class="p">,</span> <span class="n">jitter</span> <span class="o">=</span> <span class="n">jitter</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="n">threshold</span><span class="p">)</span>
                <span class="n">f1_score</span> <span class="o">=</span> <span class="n">hev</span><span class="o">.</span><span class="n">get_f1_score</span><span class="p">(</span><span class="n">df_pattern</span> <span class="o">=</span> <span class="n">df_original</span><span class="p">,</span> <span class="n">df_pattern_recall</span> <span class="o">=</span> <span class="n">df_rescued</span><span class="p">,</span> <span class="n">chromosome</span> <span class="o">=</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">bin_size</span> <span class="o">=</span> <span class="n">bin_size</span><span class="p">,</span> <span class="n">jitter</span> <span class="o">=</span> <span class="n">jitter</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="n">threshold</span><span class="p">)</span>

                <span class="c1"># Get plots</span>
                <span class="n">hpl</span><span class="o">.</span><span class="n">plot_pattern_reconstruction</span><span class="p">(</span><span class="n">table</span> <span class="o">=</span> <span class="n">true_positives</span><span class="p">,</span> <span class="n">original_cool</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">BASE_MATRIX</span><span class="p">,</span> <span class="n">rescued_cool</span> <span class="o">=</span> <span class="n">rescued_matrix_path</span><span class="p">,</span> <span class="n">chromosome</span> <span class="o">=</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="n">threshold</span><span class="p">,</span> <span class="n">case</span>  <span class="o">=</span> <span class="s2">&quot;true_positives&quot;</span><span class="p">,</span>  <span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">)</span>
                <span class="n">hpl</span><span class="o">.</span><span class="n">plot_pattern_reconstruction</span><span class="p">(</span><span class="n">table</span> <span class="o">=</span> <span class="n">false_positives</span><span class="p">,</span> <span class="n">original_cool</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">BASE_MATRIX</span><span class="p">,</span> <span class="n">rescued_cool</span> <span class="o">=</span> <span class="n">rescued_matrix_path</span><span class="p">,</span> <span class="n">chromosome</span> <span class="o">=</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="n">threshold</span><span class="p">,</span> <span class="n">case</span>  <span class="o">=</span> <span class="s2">&quot;false_positives&quot;</span><span class="p">,</span>  <span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">)</span>
                <span class="n">hpl</span><span class="o">.</span><span class="n">plot_pattern_reconstruction</span><span class="p">(</span><span class="n">table</span> <span class="o">=</span> <span class="n">false_negatives</span><span class="p">,</span> <span class="n">original_cool</span> <span class="o">=</span> <span class="n">output_data_path</span> <span class="o">/</span> <span class="n">BASE_MATRIX</span><span class="p">,</span> <span class="n">rescued_cool</span> <span class="o">=</span> <span class="n">rescued_matrix_path</span><span class="p">,</span> <span class="n">chromosome</span> <span class="o">=</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="n">threshold</span><span class="p">,</span> <span class="n">case</span>  <span class="o">=</span> <span class="s2">&quot;false_negatives&quot;</span><span class="p">,</span>  <span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">)</span>

            <span class="n">hpl</span><span class="o">.</span><span class="n">plot_benchmark</span><span class="p">(</span><span class="n">original_matrix</span> <span class="o">=</span> <span class="n">BASE_MATRIX</span><span class="p">,</span> <span class="n">depleted_matrix</span> <span class="o">=</span> <span class="n">UNRESCUED_MATRIX</span><span class="p">,</span> <span class="n">rescued_matrix</span> <span class="o">=</span> <span class="n">RESCUED_MATRIX</span><span class="p">,</span> <span class="n">chromosomes</span> <span class="o">=</span> <span class="n">chromosome_set</span><span class="p">,</span> <span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_data_path</span><span class="p">)</span>

            <span class="n">number_reads</span> <span class="o">=</span> <span class="mi">10</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Pearson score : </span><span class="si">{</span><span class="n">pearson</span><span class="si">:</span><span class="s2">9.4f</span><span class="si">}</span><span class="s2"> in mode </span><span class="si">{</span><span class="n">sub_mode</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">results</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_out</span><span class="p">:</span>
                    <span class="n">f_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>

                    <span class="n">date</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%d</span><span class="s2">/%m/%Y %H:%M:%S&quot;</span><span class="p">)</span>
                    <span class="n">f_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">id_tag</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">date</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">chromosome</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">position</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">strides</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">trans_chromosome</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">trans_position</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">auto</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">bins</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">sub_mode</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">number_reads</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">pattern</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">precision</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">recall</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">f1_score</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">pearson</span><span class="si">:</span><span class="s2">9.4f</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="n">f_out</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

            <span class="k">else</span> <span class="p">:</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_out</span><span class="p">:</span>
                    <span class="n">date</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%d</span><span class="s2">/%m/%Y %H:%M:%S&quot;</span><span class="p">)</span>
                    <span class="n">f_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">id_tag</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">date</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">chromosome</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">position</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">strides</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">trans_chromosome</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">trans_position</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">auto</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">bins</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">sub_mode</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">number_reads</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">pattern</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">precision</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">recall</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">f1_score</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">pearson</span><span class="si">:</span><span class="s2">9.4f</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="n">f_out</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Ending benchmark&quot;</span><span class="p">)</span>

    <span class="c1"># # Clean up</span>
    <span class="c1"># forward_in_path.unlink()</span>
    <span class="c1"># reverse_in_path.unlink()</span>
    <span class="c1"># forward_out_path.unlink()</span>
    <span class="c1"># reverse_out_path.unlink()</span>
    <span class="c1"># (output_data_path / BASE_MATRIX).unlink()</span>
    <span class="c1"># (output_data_path / BAM_FOR).unlink()</span>
    <span class="c1"># (output_data_path / BAM_REV).unlink()</span>
    <span class="c1"># (output_data_path / RESTRICTION_MAP).unlink()</span>
    <span class="c1"># (output_data_path / FRAGMENTS).unlink()</span>
    <span class="c1"># (output_data_path / CHROMOSOME_SIZES).unlink()</span>
    <span class="c1"># (output_data_path / DIST_FRAGS).unlink()</span>
    <span class="c1"># (output_data_path / XS).unlink()</span>

    <span class="n">files</span> <span class="o">=</span> <span class="p">[</span><span class="n">p</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">output_data_path</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">)]</span>

    <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span> 

        <span class="k">if</span> <span class="n">Path</span><span class="p">(</span><span class="n">file</span><span class="p">)</span><span class="o">.</span><span class="n">suffix</span> <span class="ow">in</span> <span class="p">[</span> <span class="s2">&quot;.npy&quot;</span><span class="p">,</span> <span class="s2">&quot;.tsv&quot;</span><span class="p">,</span> <span class="s2">&quot;.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;.pairs&quot;</span><span class="p">]:</span>

            <span class="n">Path</span><span class="p">(</span><span class="n">file</span><span class="p">)</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>

    <span class="nb">open</span><span class="p">(</span><span class="n">flag_file</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">return</span></div>


</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Sébastien Gradit.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>