

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>hicberg.statistics &mdash; hicberg 0.1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=01f34227"></script>
      <script src="../../_static/doctools.js?v=888ff710"></script>
      <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            hicberg
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../modules.html">hicberg</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorial:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../notebooks/tutorial.html">Hicberg tutorial</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../hicberg.html">hicberg package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">hicberg</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">hicberg.statistics</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for hicberg.statistics</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">from</span> <span class="nn">os</span> <span class="kn">import</span> <span class="n">getcwd</span>
<span class="kn">from</span> <span class="nn">os.path</span> <span class="kn">import</span> <span class="n">join</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span> <span class="nn">uuid</span>
<span class="kn">import</span> <span class="nn">multiprocessing</span> <span class="k">as</span> <span class="nn">mp</span>
<span class="kn">from</span> <span class="nn">functools</span> <span class="kn">import</span> <span class="n">partial</span>

<span class="kn">import</span> <span class="nn">itertools</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">numpy.random</span> <span class="kn">import</span> <span class="n">choice</span>
<span class="kn">from</span> <span class="nn">scipy.spatial.distance</span> <span class="kn">import</span> <span class="n">pdist</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">pearsonr</span>
<span class="kn">from</span> <span class="nn">scipy.ndimage</span> <span class="kn">import</span> <span class="n">gaussian_filter</span>
<span class="kn">import</span> <span class="nn">statsmodels.api</span> <span class="k">as</span> <span class="nn">sm</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="kn">import</span> <span class="nn">pysam</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span><span class="p">,</span> <span class="n">Restriction</span>
<span class="kn">from</span> <span class="nn">Bio.Restriction</span> <span class="kn">import</span> <span class="o">*</span>

<span class="kn">import</span> <span class="nn">cooler</span>
<span class="kn">import</span> <span class="nn">hicberg.utils</span> <span class="k">as</span> <span class="nn">hut</span>
<span class="kn">import</span> <span class="nn">hicberg.io</span> <span class="k">as</span> <span class="nn">hio</span>

<span class="kn">from</span> <span class="nn">hicberg</span> <span class="kn">import</span> <span class="n">logger</span>

<span class="n">lowess</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">nonparametric</span><span class="o">.</span><span class="n">lowess</span>

<span class="n">DIST_FRAG</span> <span class="o">=</span> <span class="s2">&quot;dist.frag.npy&quot;</span>
<span class="n">XS</span> <span class="o">=</span> <span class="s2">&quot;xs.npy&quot;</span>
<span class="n">COVERAGE_DICO</span> <span class="o">=</span> <span class="s2">&quot;coverage.npy&quot;</span>
<span class="n">D1D2</span> <span class="o">=</span> <span class="s2">&quot;d1d2.npy&quot;</span>
<span class="n">UNCUTS</span> <span class="o">=</span> <span class="s2">&quot;uncuts.npy&quot;</span>
<span class="n">WEIRDS</span> <span class="o">=</span> <span class="s2">&quot;weirds.npy&quot;</span>
<span class="n">LOOPS</span> <span class="o">=</span> <span class="s2">&quot;loops.npy&quot;</span>
<span class="n">TRANS_PS</span> <span class="o">=</span> <span class="s2">&quot;trans_ps.npy&quot;</span>
<span class="n">RESTRICTION_MAP</span> <span class="o">=</span> <span class="s2">&quot;restriction_map.npy&quot;</span>
<span class="n">DENSITY_MAP</span> <span class="o">=</span> <span class="s2">&quot;density_map.npy&quot;</span>

<span class="c1"># TODO : check if keeping is necessary</span>
<div class="viewcode-block" id="generate_density_map_backup">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.generate_density_map_backup">[docs]</a>
<span class="k">def</span> <span class="nf">generate_density_map_backup</span><span class="p">(</span><span class="n">matrix</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;unrescued_map.cool&quot;</span><span class="p">,</span> <span class="n">rounds</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">magnitude</span> <span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span> <span class="n">output_dir</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="nb">float</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create density map from a Hi-C matrix. Return a dictionary where keys are chromosomes names and values are density maps.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    matrix : str, optional</span>
<span class="sd">        Path to a cooler matrix, by default None</span>
<span class="sd">    rounds : int, optional</span>
<span class="sd">        Number of times the matrix has to be shaken, by default 1</span>
<span class="sd">    magnitude : float, optional</span>
<span class="sd">        Blending ratio between the native matrix and the shaken one, by default 1.0</span>
<span class="sd">    output_dir : str, optional</span>
<span class="sd">        Path to the folder where to save the density map, by default None</span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    dict[str, np.ndarray[float]]</span>
<span class="sd">        Density maps as a dictionary where keys are chromosomes names couples as tuples and values are density maps.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Start generating density map...&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">output_dir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">getcwd</span><span class="p">())</span>

    <span class="k">else</span> <span class="p">:</span> 

        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>

    <span class="n">matrix_path</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="n">matrix</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">matrix_path</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>

        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Matrix file </span><span class="si">{</span><span class="n">matrix</span><span class="si">}</span><span class="s2"> not found. Please provide a valid path to a matrix file.&quot;</span><span class="p">)</span>
    
    <span class="n">density_map</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Load cooler matrix</span>

    <span class="n">matrix</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_cooler</span><span class="p">(</span><span class="n">matrix_path</span><span class="p">)</span>

    <span class="c1"># Get chromosomes names</span>

    <span class="n">chromosomes</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">chromnames</span>

    <span class="n">chromosomes_combination</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span><span class="n">chromosomes</span><span class="p">,</span> <span class="n">chromosomes</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">combination</span> <span class="ow">in</span> <span class="n">chromosomes_combination</span><span class="p">:</span>

        <span class="c1"># Get matrix for each chromosome pair</span>

        <span class="n">matrix_chromosome</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">balance</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">combination</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">combination</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

        <span class="c1"># Get density map for each chromosome pair</span>

        <span class="k">if</span> <span class="n">combination</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">combination</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>

            <span class="n">density_map</span><span class="p">[</span><span class="n">combination</span><span class="p">]</span> <span class="o">=</span> <span class="n">hut</span><span class="o">.</span><span class="n">diffuse_matrix</span><span class="p">(</span><span class="n">matrix</span> <span class="o">=</span> <span class="n">matrix_chromosome</span><span class="p">,</span> <span class="n">rounds</span> <span class="o">=</span> <span class="n">rounds</span><span class="p">,</span> <span class="n">magnitude</span> <span class="o">=</span> <span class="n">magnitude</span><span class="p">,</span> <span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;intra&quot;</span><span class="p">)</span>

        <span class="k">elif</span> <span class="n">combination</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">combination</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>

            <span class="n">density_map</span><span class="p">[</span><span class="n">combination</span><span class="p">]</span> <span class="o">=</span> <span class="n">hut</span><span class="o">.</span><span class="n">diffuse_matrix</span><span class="p">(</span><span class="n">matrix</span> <span class="o">=</span> <span class="n">matrix_chromosome</span><span class="p">,</span> <span class="n">rounds</span> <span class="o">=</span> <span class="n">rounds</span><span class="p">,</span> <span class="n">magnitude</span> <span class="o">=</span> <span class="n">magnitude</span><span class="p">,</span> <span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;inter&quot;</span><span class="p">)</span>

        <span class="n">blured_density</span>  <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">density_map</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>

        <span class="k">if</span> <span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>

            <span class="n">blured_density</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">gaussian_filter</span><span class="p">(</span><span class="n">density_map</span><span class="p">[</span><span class="n">key</span><span class="p">],</span> <span class="n">sigma</span> <span class="o">=</span> <span class="mi">5</span><span class="p">)</span>

        <span class="k">else</span> <span class="p">:</span>

            <span class="n">blured_density</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">gaussian_filter</span><span class="p">(</span><span class="n">density_map</span><span class="p">[</span><span class="n">key</span><span class="p">],</span> <span class="n">sigma</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>

    <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">DENSITY_MAP</span><span class="p">,</span> <span class="n">blured_density</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved density map at : </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">blured_density</span></div>


<span class="c1"># TODO : to correct docstrings</span>
<div class="viewcode-block" id="generate_density_map">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.generate_density_map">[docs]</a>
<span class="k">def</span> <span class="nf">generate_density_map</span><span class="p">(</span><span class="n">matrix</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;unrescued_map.cool&quot;</span><span class="p">,</span> <span class="n">size</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">sigma</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">n_mads</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">nan_threshold</span> <span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">output_dir</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="nb">float</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create density map from a Hi-C matrix. Return a dictionary where keys are chromosomes names and values are density maps.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    matrix : str, optional</span>
<span class="sd">        Path to a cooler matrix, by default None</span>
<span class="sd">    rounds : int, optional</span>
<span class="sd">        Number of times the matrix has to be shaken, by default 1</span>
<span class="sd">    magnitude : float, optional</span>
<span class="sd">        Blending ratio between the native matrix and the shaken one, by default 1.0</span>
<span class="sd">    output_dir : str, optional</span>
<span class="sd">        Path to the folder where to save the density map, by default None</span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    dict[str, np.ndarray[float]]</span>
<span class="sd">        Density maps as a dictionary where keys are chromosomes names couples as tuples and values are density maps.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Start generating density map...&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">output_dir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">getcwd</span><span class="p">())</span>

    <span class="k">else</span> <span class="p">:</span> 

        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>

    <span class="n">matrix_path</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="n">matrix</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">matrix_path</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>

        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Matrix file </span><span class="si">{</span><span class="n">matrix</span><span class="si">}</span><span class="s2"> not found. Please provide a valid path to a matrix file.&quot;</span><span class="p">)</span>
    
    <span class="n">density_map</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Load cooler matrix</span>

    <span class="n">matrix</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_cooler</span><span class="p">(</span><span class="n">matrix_path</span><span class="p">)</span>

    <span class="c1"># Get chromosomes names</span>

    <span class="n">chromosomes</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">chromnames</span>

    <span class="n">chromosomes_combination</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span><span class="n">chromosomes</span><span class="p">,</span> <span class="n">chromosomes</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">combination</span> <span class="ow">in</span> <span class="n">chromosomes_combination</span><span class="p">:</span>

        <span class="c1"># Get matrix for each chromosome pair</span>
        <span class="n">matrix_chromosome</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">balance</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">combination</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">combination</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

        <span class="c1"># Get density map for each chromosome pair</span>
        <span class="n">density_map</span><span class="p">[</span><span class="n">combination</span><span class="p">]</span> <span class="o">=</span> <span class="n">hut</span><span class="o">.</span><span class="n">get_local_density</span><span class="p">(</span><span class="n">matrix</span> <span class="o">=</span> <span class="n">matrix_chromosome</span><span class="p">,</span> <span class="n">size</span>  <span class="o">=</span> <span class="n">size</span><span class="p">,</span> <span class="n">sigma</span>  <span class="o">=</span> <span class="n">sigma</span><span class="p">,</span> <span class="n">n_mads</span>  <span class="o">=</span> <span class="n">n_mads</span><span class="p">,</span> <span class="n">nan_threshold</span>  <span class="o">=</span> <span class="n">nan_threshold</span><span class="p">)</span>

    <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">DENSITY_MAP</span><span class="p">,</span> <span class="n">density_map</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved density map at : </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">density_map</span></div>


<div class="viewcode-block" id="compute_density">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.compute_density">[docs]</a>
<span class="k">def</span> <span class="nf">compute_density</span><span class="p">(</span><span class="n">cooler_file</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;unrescued_map.cool&quot;</span><span class="p">,</span> <span class="n">threads</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">kernel_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">11</span><span class="p">,</span> <span class="n">deviation</span> <span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">output_dir</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span> 
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create density map from a Hi-C matrix. Return a dictionary where keys are chromosomes names and values are density maps.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    cooler_file : str, optional</span>
<span class="sd">        [description], by default None</span>
<span class="sd">    threads : int, optional</span>
<span class="sd">        [description], by default 2</span>
<span class="sd">    output_dir : str, optional</span>
<span class="sd">        [description], by default None</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Start generating density map...&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">output_dir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">getcwd</span><span class="p">())</span>

    <span class="k">else</span> <span class="p">:</span> 
        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>

    <span class="n">matrix_path</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="n">cooler_file</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">matrix_path</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>

        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Matrix file </span><span class="si">{</span><span class="n">matrix</span><span class="si">}</span><span class="s2"> not found. Please provide a valid path to a matrix file.&quot;</span><span class="p">)</span>

    <span class="c1">#Load cooler file</span>
    <span class="n">matrix</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_cooler</span><span class="p">(</span><span class="n">matrix</span> <span class="o">=</span> <span class="n">matrix_path</span><span class="p">)</span>

    <span class="c1">#Get chromosomes names</span>
    <span class="n">chromosomes</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">chromnames</span>

    <span class="c1">#Get chromosomes couples</span>
    <span class="n">chromosomes_couples</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span><span class="n">chromosomes</span><span class="p">,</span> <span class="n">repeat</span> <span class="o">=</span> <span class="mi">2</span><span class="p">))</span>

    <span class="c1"># Get chromsomes maps</span>
    <span class="n">chromosomes_maps</span> <span class="o">=</span> <span class="p">[</span><span class="n">matrix</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">balance</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">chrom1</span><span class="p">,</span> <span class="n">chrom2</span><span class="p">)</span> <span class="k">for</span> <span class="n">chrom1</span><span class="p">,</span> <span class="n">chrom2</span> <span class="ow">in</span> <span class="n">chromosomes_couples</span><span class="p">]</span>

    <span class="n">pool</span> <span class="o">=</span> <span class="n">mp</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">threads</span><span class="p">)</span>

    <span class="n">results</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">partial</span><span class="p">(</span><span class="n">hut</span><span class="o">.</span><span class="n">get_local_density</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">matrix_path</span><span class="p">),</span> <span class="n">size</span>  <span class="o">=</span> <span class="n">kernel_size</span><span class="p">,</span> <span class="n">sigma</span>  <span class="o">=</span> <span class="n">deviation</span><span class="p">,</span> <span class="n">n_mads</span>  <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">nan_threshold</span>  <span class="o">=</span> <span class="kc">False</span><span class="p">),</span>
            <span class="n">chromosomes_couples</span><span class="p">)</span>

    <span class="c1"># Close the pool and wait for the work to finish</span>
    <span class="n">pool</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">pool</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>

    <span class="n">results_dict</span> <span class="o">=</span>  <span class="p">{</span><span class="n">key</span> <span class="p">:</span> <span class="n">value</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">results</span><span class="p">}</span>

    <span class="k">for</span> <span class="n">chrom_pair</span> <span class="ow">in</span> <span class="n">results_dict</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">chrom_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">chrom_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
            <span class="k">pass</span>

        <span class="k">else</span> <span class="p">:</span>
            <span class="n">results_dict</span><span class="p">[(</span><span class="n">chrom_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">chrom_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span>  <span class="o">=</span> <span class="n">results_dict</span><span class="p">[</span><span class="n">chrom_pair</span><span class="p">]</span><span class="o">.</span><span class="n">T</span>

    <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">DENSITY_MAP</span><span class="p">,</span> <span class="n">results_dict</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved density map at : </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="get_restriction_map">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.get_restriction_map">[docs]</a>
<span class="k">def</span> <span class="nf">get_restriction_map</span><span class="p">(</span><span class="n">genome</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">enzyme</span> <span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;DpnII&quot;</span><span class="p">],</span> <span class="n">output_dir</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="nb">int</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get ordered restriction map (including 0 and n) from a chromosome sequence.</span>
<span class="sd">    Return a dictionary where keys are chromosomes names and values are restrictions sites positions.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    genome : str, optional</span>
<span class="sd">        Path to the genome to digest, by default None, by default None</span>
<span class="sd">    enzyme : list[str], optional</span>
<span class="sd">        Enzyme or list of enzyme to digest the genome with. If integer passed, micro-C mode using Mnase is used, and the integer correspond to the size of nucleosomal fragment, by default None, by default [&quot;DpnII&quot;]</span>
<span class="sd">    output_dir : str, optional</span>
<span class="sd">        Path to the folder where to save the plot, by default None</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    dict</span>
<span class="sd">        Dictionary of the product of digestion where keys are chromosomes names and values are restrictions sites positions.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Generating restriction map...&quot;</span><span class="p">)</span>

    <span class="n">genome_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">genome_path</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>

        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Genome file </span><span class="si">{</span><span class="n">genome</span><span class="si">}</span><span class="s2"> not found. Please provide a valid path to a genome file.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">output_dir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">getcwd</span><span class="p">())</span>

    <span class="k">else</span> <span class="p">:</span> 

        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>
    
    <span class="n">restriction_map_dictionary</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">enzyme</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">enzyme</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">isnumeric</span><span class="p">():</span>
        <span class="c1"># print(f&quot;Enabled micro-c with enzyme : {enzyme}&quot;)</span>
        <span class="n">enzyme</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">enzyme</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

        <span class="k">for</span> <span class="n">seq_record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>

            <span class="c1"># Get restriction map from the restriction batch.</span>
            <span class="n">restriction_map</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">seq</span><span class="p">),</span> <span class="n">enzyme</span><span class="p">)</span>
            <span class="n">restriction_map</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span>
                <span class="n">restriction_map</span><span class="p">,</span>
                <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">restriction_map</span><span class="p">)],</span>
                <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">seq</span><span class="p">)],</span>
            <span class="p">)</span>
            <span class="n">restriction_map_dictionary</span><span class="p">[</span><span class="n">seq_record</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">restriction_map</span>

    <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">enzyme</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span> <span class="ow">or</span> <span class="nb">type</span><span class="p">(</span><span class="n">enzyme</span><span class="p">)</span> <span class="o">==</span> <span class="nb">list</span> <span class="ow">or</span> <span class="nb">type</span><span class="p">(</span><span class="n">enzyme</span><span class="p">)</span> <span class="o">==</span> <span class="nb">tuple</span><span class="p">:</span>

        <span class="n">restriction_batch</span> <span class="o">=</span> <span class="n">Restriction</span><span class="o">.</span><span class="n">RestrictionBatch</span><span class="p">()</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">enzyme</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">enzyme</span> <span class="o">=</span> <span class="n">enzyme</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">enz</span> <span class="ow">in</span> <span class="n">enzyme</span><span class="p">:</span>
            <span class="n">restriction_batch</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">enz</span><span class="p">)</span>

        <span class="c1"># parse sequence from fasta file</span>
        <span class="k">for</span> <span class="n">seq_record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>

            <span class="c1"># Get restriction map from the restriction batch.</span>
            <span class="n">restriction_map</span> <span class="o">=</span> <span class="n">restriction_batch</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>

            <span class="c1"># Convert dictionary values to numpy array</span>
            <span class="n">restriction_map_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">pos</span> <span class="k">for</span> <span class="n">sites</span> <span class="ow">in</span> <span class="n">restriction_map</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">sites</span><span class="p">])</span>
            <span class="p">)</span>
            <span class="n">restriction_map_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span>
                <span class="n">restriction_map_array</span><span class="p">,</span>
                <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">restriction_map_array</span><span class="p">)],</span>
                <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">seq</span><span class="p">)],</span>
            <span class="p">)</span>

            <span class="n">restriction_map_dictionary</span><span class="p">[</span><span class="n">seq_record</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">restriction_map_array</span>

    <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">RESTRICTION_MAP</span><span class="p">,</span> <span class="n">restriction_map_dictionary</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved restriction map at : </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    

    <span class="k">return</span> <span class="n">restriction_map_dictionary</span></div>



<div class="viewcode-block" id="generate_xs">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.generate_xs">[docs]</a>
<span class="k">def</span> <span class="nf">generate_xs</span><span class="p">(</span><span class="n">chromosome_size</span> <span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">base</span> <span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate xs array for computing P(s). Return xs array which is log space.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    chromosome_size : int</span>
<span class="sd">        Size of the chromosome to be binned in bp.</span>
<span class="sd">    base : float, optional</span>
<span class="sd">        Base of the log space., by default 1.1</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    np.ndarray[int]</span>
<span class="sd">        Array of log bins related to the chromosome.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">n_bins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">chromosome_size</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">base</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">xs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">logspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">n_bins</span><span class="p">,</span> <span class="n">num</span><span class="o">=</span><span class="n">n_bins</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">base</span><span class="o">=</span><span class="n">base</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
        <span class="p">)</span>
    
    <span class="c1"># Add 0 to the beginning of the array (allow distance 0 to be counted in the first bin)</span>
    <span class="n">xs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">xs</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">xs</span></div>

    <span class="c1"># # TODO : recent correction to fuse bins n-1 and n    (to delete ?)</span>
    <span class="c1"># return np.delete(xs, -2)</span>

<div class="viewcode-block" id="log_bin_genome">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.log_bin_genome">[docs]</a>
<span class="k">def</span> <span class="nf">log_bin_genome</span><span class="p">(</span><span class="n">genome</span> <span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">base</span> <span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.1</span><span class="p">,</span> <span class="n">output_dir</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="nb">int</span><span class="p">]]:</span>
    
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Start log binning of genome...&quot;</span><span class="p">)</span>

    <span class="n">genome_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">genome_path</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>

        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Genome file </span><span class="si">{</span><span class="n">genome</span><span class="si">}</span><span class="s2"> not found. Please provide a valid path to a genome file.&quot;</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">output_dir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            
        <span class="n">folder_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">getcwd</span><span class="p">())</span>

    <span class="k">else</span><span class="p">:</span>
            
        <span class="n">folder_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>

    <span class="n">genome_parser</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
    <span class="n">xs_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">seq_record</span><span class="o">.</span><span class="n">id</span> <span class="p">:</span> <span class="n">generate_xs</span><span class="p">(</span><span class="n">chromosome_size</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">seq</span><span class="p">),</span> <span class="n">base</span> <span class="o">=</span> <span class="n">base</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq_record</span> <span class="ow">in</span> <span class="n">genome_parser</span><span class="p">}</span>

    <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">folder_path</span> <span class="o">/</span> <span class="n">XS</span><span class="p">,</span> <span class="n">xs_dict</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Log binning of genome </span><span class="si">{</span><span class="n">genome</span><span class="si">}</span><span class="s2"> saved in </span><span class="si">{</span><span class="n">folder_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">XS</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="attribute_xs">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.attribute_xs">[docs]</a>
<span class="k">def</span> <span class="nf">attribute_xs</span><span class="p">(</span><span class="n">xs</span> <span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span> <span class="n">distance</span> <span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Attibute genomic distance to the corresponding log bin of xs.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    xs : np.ndarray[int]</span>
<span class="sd">        Array containing the log bins.</span>
<span class="sd">    distance : int</span>
<span class="sd">        Genomic distance in bp.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    int</span>
<span class="sd">        Index of the corresponding bins where distance has to be attributed.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="n">xs</span><span class="p">,</span> <span class="n">distance</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
    
    <span class="k">return</span> <span class="n">idx</span> <span class="k">if</span> <span class="n">distance</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span> </div>


<div class="viewcode-block" id="get_dist_frags">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.get_dist_frags">[docs]</a>
<span class="k">def</span> <span class="nf">get_dist_frags</span><span class="p">(</span><span class="n">genome</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">restriction_map</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">circular</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">rate</span> <span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span> <span class="n">output_dir</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the distribution of fragments&#39; distance from a genome distribution .</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    genome : str, optional</span>
<span class="sd">        Path to the genome, by default None</span>
<span class="sd">    restriction_map : dict, optional</span>
<span class="sd">        Restriction map saved as a dictionary like chrom_name : list of restriction sites&#39; position, by default None</span>
<span class="sd">    circular : str, optional</span>
<span class="sd">        Name of the chromosomes to consider as circular, by default &quot;&quot;</span>
<span class="sd">    rate : float, optional</span>
<span class="sd">        Set the proportion of restriction sites to consider. Avoid memory overflow when restriction maps are very dense, by default 1.0</span>
<span class="sd">    output_dir : str, optional</span>
<span class="sd">        Path to the folder where to save the dictionary, by default None</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    dict</span>
<span class="sd">        Dictionary of sub-sampled restriction map with keys as chromosome names and values as lists of restriction sites&#39; position.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Start generating distribution of fragments&#39; distance...&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">output_dir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

        <span class="n">folder_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">getcwd</span><span class="p">())</span>

    <span class="k">else</span><span class="p">:</span>

        <span class="n">folder_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>

    <span class="k">if</span> <span class="p">(</span><span class="mf">0.0</span> <span class="o">&gt;</span> <span class="n">rate</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">rate</span> <span class="o">&gt;</span> <span class="mf">1.0</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Subsampling rate must be between 0.0 and 1.0.&quot;</span><span class="p">)</span>

    <span class="n">genome_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">genome_path</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
            
        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Genome file </span><span class="si">{</span><span class="n">genome</span><span class="si">}</span><span class="s2"> not found. Please provide a valid path to a genome file.&quot;</span><span class="p">)</span>
    
    <span class="n">dist_frag</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="n">xs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    
    <span class="k">if</span> <span class="n">rate</span> <span class="o">!=</span> <span class="mf">1.0</span><span class="p">:</span>

        <span class="n">restriction_map</span> <span class="o">=</span> <span class="n">hut</span><span class="o">.</span><span class="n">subsample_restriction_map</span><span class="p">(</span><span class="n">restriction_map</span> <span class="o">=</span> <span class="n">restriction_map</span><span class="p">,</span> <span class="n">rate</span> <span class="o">=</span> <span class="n">rate</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">seq_record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>

        <span class="n">seq_name</span> <span class="o">=</span> <span class="n">seq_record</span><span class="o">.</span><span class="n">id</span>

        
        <span class="k">if</span> <span class="n">seq_record</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">circular</span><span class="p">:</span>

            <span class="n">map_size</span> <span class="o">=</span> <span class="n">restriction_map</span><span class="p">[</span><span class="n">seq_name</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

            <span class="n">forward_distances</span> <span class="o">=</span> <span class="n">pdist</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">restriction_map</span><span class="p">[</span><span class="n">seq_name</span><span class="p">],</span> <span class="p">(</span><span class="n">map_size</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
            <span class="p">)</span>
            <span class="n">max_size_vector</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span>
                <span class="n">forward_distances</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">restriction_map</span><span class="p">[</span><span class="n">seq_name</span><span class="p">])</span>
            <span class="p">)</span>
            <span class="n">backward_distances</span> <span class="o">=</span> <span class="n">max_size_vector</span> <span class="o">-</span> <span class="n">forward_distances</span>
            <span class="n">pairwise_distances</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">minimum</span><span class="p">(</span><span class="n">forward_distances</span><span class="p">,</span> <span class="n">backward_distances</span><span class="p">)</span>
            <span class="n">pairwise_distances</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span>
                <span class="n">pairwise_distances</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">pairwise_distances</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
            <span class="p">)</span>

            <span class="c1"># freeing memory</span>
            <span class="k">del</span> <span class="n">forward_distances</span>
            <span class="k">del</span> <span class="n">backward_distances</span>
            <span class="k">del</span> <span class="n">max_size_vector</span>

        <span class="k">else</span> <span class="p">:</span>

            <span class="n">pairwise_distances</span> <span class="o">=</span> <span class="n">pdist</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span>
                    <span class="n">restriction_map</span><span class="p">[</span><span class="n">seq_name</span><span class="p">],</span>
                    <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">restriction_map</span><span class="p">[</span><span class="n">seq_name</span><span class="p">]),</span> <span class="mi">1</span><span class="p">),</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="n">pairwise_distances</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span>
                <span class="n">pairwise_distances</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">pairwise_distances</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="c1"># Computing xs</span>
        <span class="n">xs</span><span class="p">[</span><span class="n">seq_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">generate_xs</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">seq</span><span class="p">),</span> <span class="n">base</span><span class="o">=</span><span class="mf">1.1</span><span class="p">)</span>
        <span class="n">dist_frag</span><span class="p">[</span><span class="n">seq_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">xs</span><span class="p">[</span><span class="n">seq_name</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
        <span class="c1"># Parse distances</span>
        <span class="k">for</span> <span class="n">distance</span> <span class="ow">in</span> <span class="n">pairwise_distances</span><span class="p">:</span>
            <span class="n">dist_frag</span><span class="p">[</span><span class="n">seq_name</span><span class="p">][</span><span class="n">attribute_xs</span><span class="p">(</span><span class="n">xs</span><span class="p">[</span><span class="n">seq_name</span><span class="p">],</span> <span class="n">distance</span><span class="p">)]</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="c1"># Save dictionaries</span>
    <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span>
        <span class="n">folder_path</span> <span class="o">/</span> <span class="n">DIST_FRAG</span><span class="p">,</span>
        <span class="n">dist_frag</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved restriction map at : </span><span class="si">{</span><span class="n">folder_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">DIST_FRAG</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="generate_trans_ps">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.generate_trans_ps">[docs]</a>
<span class="k">def</span> <span class="nf">generate_trans_ps</span><span class="p">(</span><span class="n">matrix</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;unrescued_map.cool&quot;</span><span class="p">,</span> <span class="n">chrom_sizes</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;chromosome_sizes.npy&quot;</span><span class="p">,</span> <span class="n">output_dir</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate pseudo-P(s) while considering reads originating from different chromosomes.</span>
<span class="sd">    Pseudo-P(s) are computed as the number of interactions between two chromosomes divided by the product of the length of these two chromosomes.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    matrix : str, optional</span>
<span class="sd">        Path to a cooler matrix, by default &quot;unrescued_map.cool&quot;</span>
<span class="sd">    chrom_sizes : str, optional</span>
<span class="sd">        Path to the chromosome sizes dictionary, by default &quot;chromosome_sizes.npy&quot;</span>
<span class="sd">    output_dir : str, optional</span>
<span class="sd">        Path to the folder where to save the dictionary, by default None.</span>

<span class="sd">    &quot;&quot;&quot;</span>    

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Start getting trans-P(s)&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">output_dir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">getcwd</span><span class="p">())</span>

    <span class="k">else</span><span class="p">:</span>

        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>

    <span class="n">chrom_size_dict</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_dictionary</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">chrom_sizes</span><span class="p">)</span>

    <span class="n">matrix_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">matrix</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">matrix_path</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>

        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Matrix file </span><span class="si">{</span><span class="n">matrix</span><span class="si">}</span><span class="s2"> not found. Please provide a valid path to a matrix file.&quot;</span><span class="p">)</span>
        
    <span class="n">matrix</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">matrix_path</span><span class="o">.</span><span class="n">as_posix</span><span class="p">())</span>

    <span class="n">chromosome_sets</span> <span class="o">=</span> <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">((</span><span class="n">chrom_size_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="n">repeat</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

    <span class="n">trans_ps</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">all_interaction_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">chrom_size_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
    <span class="n">n_frags_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">chrom_size_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">chromosome_sets</span><span class="p">):</span>

        <span class="n">all_interactions</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">balance</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">s</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">s</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="n">n_frags</span> <span class="o">=</span> <span class="n">chrom_size_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">s</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="n">chrom_size_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">s</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        
        <span class="n">trans_ps</span><span class="p">[</span><span class="n">s</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">all_interactions</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">n_frags</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span> <span class="c1"># Multiplied by 4 to balance 4 configurations of reads orientation (++/+-/-+/--)</span>
        
        <span class="n">all_interaction_matrix</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_interactions</span>
        <span class="n">n_frags_matrix</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">n_frags</span>

    <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">TRANS_PS</span><span class="p">,</span> <span class="n">trans_ps</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Trans P(s) saved in </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="generate_coverages">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.generate_coverages">[docs]</a>
<span class="k">def</span> <span class="nf">generate_coverages</span><span class="p">(</span><span class="n">genome</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">bins</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span> <span class="n">forward_bam_file</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;group1.1.bam&quot;</span><span class="p">,</span> <span class="n">reverse_bam_file</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;group1.2.bam&quot;</span><span class="p">,</span> <span class="n">output_dir</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Take a genome and both forward and reverse bam files for unambiguous group and return a dictionary containing the coverage in terms of reads overs chromosomes.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    genome : str, optional</span>
<span class="sd">        Path to the genome file to get coverage on, by default None</span>
<span class="sd">    bins : int, optional</span>
<span class="sd">        Size of the desired bin., by default 2000</span>
<span class="sd">    forward_bam_file : str, optional</span>
<span class="sd">        Path to forward .bam alignment file, by default None, by default group1.1.bam</span>
<span class="sd">    reverse_bam_file : str, optional</span>
<span class="sd">        Path to reverse .bam alignment file, by default None, by default group1.2.bam</span>
<span class="sd">    output_dir : str, optional</span>
<span class="sd">        Path to the folder where to save the classified alignment files, by default None, by default None</span>
<span class="sd">    &quot;&quot;&quot;</span>        
    
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Start generating coverages...&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">output_dir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">getcwd</span><span class="p">())</span>

    <span class="k">else</span><span class="p">:</span>

        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>
    
    <span class="n">genome_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">genome_path</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
                
        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Genome file </span><span class="si">{</span><span class="n">genome</span><span class="si">}</span><span class="s2"> not found. Please provide a valid path to a genome file.&quot;</span><span class="p">)</span>   
    
    <span class="n">genome_parser</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

    <span class="n">genome_coverages</span> <span class="o">=</span> <span class="p">{</span><span class="n">seq_record</span><span class="o">.</span><span class="n">id</span> <span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">seq</span><span class="p">),</span> <span class="n">bins</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">))</span> <span class="k">for</span> <span class="n">seq_record</span> <span class="ow">in</span> <span class="n">genome_parser</span><span class="p">}</span>

    <span class="n">forward_bam_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">forward_bam_file</span><span class="p">)</span>
    <span class="n">reverse_bam_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">reverse_bam_file</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">forward_bam_path</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>

        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Forward .bam file </span><span class="si">{</span><span class="n">forward_bam_file</span><span class="si">}</span><span class="s2"> not found. Please provide a valid path to a forward .bam file.&quot;</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="ow">not</span> <span class="n">reverse_bam_path</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>

        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reverse .bam file </span><span class="si">{</span><span class="n">reverse_bam_file</span><span class="si">}</span><span class="s2"> not found. Please provide a valid path to a reverse .bam file.&quot;</span><span class="p">)</span>
    
    <span class="n">forward_bam_handler</span><span class="p">,</span> <span class="n">reverse_bam_handler</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">forward_bam_path</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">),</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">reverse_bam_path</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">forward_read</span><span class="p">,</span> <span class="n">reverse_read</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">forward_bam_handler</span><span class="p">,</span> <span class="n">reverse_bam_handler</span><span class="p">):</span>

        <span class="n">genome_coverages</span><span class="p">[</span><span class="n">forward_read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">][</span>
            <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">forward_read</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span> <span class="n">bins</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
        <span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">genome_coverages</span><span class="p">[</span><span class="n">reverse_read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">][</span>
            <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">reverse_read</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span> <span class="n">bins</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
        <span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="c1"># close files</span>
    <span class="n">forward_bam_handler</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">reverse_bam_handler</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Smooth coverages</span>
    <span class="n">smoothed_coverages</span> <span class="o">=</span> <span class="p">{</span><span class="n">seq_name</span> <span class="p">:</span> <span class="n">hut</span><span class="o">.</span><span class="n">mad_smoothing</span><span class="p">(</span><span class="n">coverage</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq_name</span><span class="p">,</span> <span class="n">coverage</span> <span class="ow">in</span> <span class="n">genome_coverages</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

    <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">COVERAGE_DICO</span><span class="p">,</span> <span class="n">smoothed_coverages</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Coverage dictionary saved in </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="generate_d1d2">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.generate_d1d2">[docs]</a>
<span class="k">def</span> <span class="nf">generate_d1d2</span><span class="p">(</span><span class="n">forward_bam_file</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;group1.1.bam&quot;</span><span class="p">,</span> <span class="n">reverse_bam_file</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;group1.2.bam&quot;</span><span class="p">,</span> <span class="n">restriction_map</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;restriction_map.npy&quot;</span><span class="p">,</span> <span class="n">output_dir</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute d1d2 distance laws with the given alignments and restriction map.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    forward_bam_file : str, optional</span>
<span class="sd">        Path to forward .bam alignment file, by default None, by default group1.1.bam, by default &quot;group1.1.bam&quot;</span>
<span class="sd">    reverse_bam_file : str, optional</span>
<span class="sd">        Path to reverse .bam alignment file, by default None, by default group1.1.bam, by default &quot;group1.2.bam&quot;</span>
<span class="sd">    restriction_map : str, optional</span>
<span class="sd">        Restriction map saved as a dictionary like chrom_name : list of restriction sites&#39; position, by default &quot;dist.frag.npy&quot;</span>
<span class="sd">    output_dir : str, optional</span>
<span class="sd">        Path to the folder where to save the dictionary, by default None, by default None</span>
<span class="sd">    &quot;&quot;&quot;</span>    
    
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Start generating d1d2 law...&quot;</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">output_dir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">getcwd</span><span class="p">())</span>

    <span class="k">else</span><span class="p">:</span>

        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>

    <span class="n">forward_bam_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">forward_bam_file</span><span class="p">)</span>
    <span class="n">reverse_bam_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">reverse_bam_file</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">forward_bam_path</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
            
        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Forward .bam file </span><span class="si">{</span><span class="n">forward_bam_file</span><span class="si">}</span><span class="s2"> not found. Please provide a valid path to a forward .bam file.&quot;</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="ow">not</span> <span class="n">reverse_bam_path</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
            
        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reverse .bam file </span><span class="si">{</span><span class="n">reverse_bam_file</span><span class="si">}</span><span class="s2"> not found. Please provide a valid path to a reverse .bam file.&quot;</span><span class="p">)</span>
    
    <span class="n">forward_bam_handler</span><span class="p">,</span> <span class="n">reverse_bam_handler</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">forward_bam_path</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">),</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">reverse_bam_path</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>

    <span class="c1"># Ensure that the restriction map is a dictionary to be loaded</span>
    <span class="k">try</span><span class="p">:</span>

        <span class="n">restriction_map</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_dictionary</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">restriction_map</span><span class="p">)</span>

    <span class="k">except</span> <span class="p">:</span> 

        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;restriction map not found&quot;</span><span class="p">)</span>
        <span class="k">pass</span>


    <span class="n">list_d1d2</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># list containing the (d1+d2) i.e size of the fragment to sequence</span>

    <span class="k">for</span> <span class="n">forward_read</span><span class="p">,</span> <span class="n">reverse_read</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">forward_bam_handler</span><span class="p">,</span> <span class="n">reverse_bam_handler</span><span class="p">):</span>

        <span class="n">r_sites_forward_read</span> <span class="o">=</span> <span class="n">restriction_map</span><span class="p">[</span><span class="n">forward_read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">]</span>
        <span class="n">r_sites_reverse_read</span> <span class="o">=</span> <span class="n">restriction_map</span><span class="p">[</span><span class="n">reverse_read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">forward_read</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">forward_read</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">256</span><span class="p">:</span>

            <span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="n">r_sites_forward_read</span><span class="p">,</span> <span class="n">forward_read</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span>

            <span class="n">distance_1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">r_sites_forward_read</span><span class="p">[</span><span class="n">index</span><span class="p">],</span> <span class="n">forward_read</span><span class="o">.</span><span class="n">pos</span><span class="p">)</span>

        <span class="k">elif</span> <span class="n">forward_read</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">16</span> <span class="ow">or</span> <span class="n">forward_read</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">272</span><span class="p">:</span>

            <span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="n">r_sites_forward_read</span><span class="p">,</span> <span class="n">forward_read</span><span class="o">.</span><span class="n">reference_end</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>

            <span class="n">distance_1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">forward_read</span><span class="o">.</span><span class="n">reference_end</span><span class="p">,</span> <span class="n">r_sites_forward_read</span><span class="p">[</span><span class="n">index</span><span class="p">])</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">reverse_read</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">reverse_read</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">256</span><span class="p">:</span>

            <span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span>
                <span class="n">r_sites_reverse_read</span><span class="p">,</span> <span class="n">reverse_read</span><span class="o">.</span><span class="n">reference_start</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="s2">&quot;right&quot;</span>
            <span class="p">)</span>  <span class="c1"># right</span>


            <span class="n">distance_2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">r_sites_reverse_read</span><span class="p">[</span><span class="n">index</span><span class="p">],</span> <span class="n">reverse_read</span><span class="o">.</span><span class="n">reference_start</span><span class="p">)</span>

        <span class="k">elif</span> <span class="n">reverse_read</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">16</span> <span class="ow">or</span> <span class="n">reverse_read</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">272</span><span class="p">:</span>

            <span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span>
                <span class="n">r_sites_reverse_read</span><span class="p">,</span> <span class="n">reverse_read</span><span class="o">.</span><span class="n">reference_end</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="s2">&quot;left&quot;</span>
            <span class="p">)</span>  <span class="c1"># left</span>

            <span class="n">distance_2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">reverse_read</span><span class="o">.</span><span class="n">reference_end</span><span class="p">,</span> <span class="n">r_sites_reverse_read</span><span class="p">[</span><span class="n">index</span><span class="p">])</span>
            <span class="p">)</span>

        <span class="c1"># Correction for uncuts with no restriction sites inside</span>
        <span class="k">if</span> <span class="n">forward_read</span><span class="o">.</span><span class="n">reference_name</span> <span class="o">==</span> <span class="n">reverse_read</span><span class="o">.</span><span class="n">reference_name</span> <span class="ow">and</span> <span class="n">np</span><span class="o">.</span><span class="n">add</span><span class="p">(</span>
            <span class="n">distance_1</span><span class="p">,</span> <span class="n">distance_2</span>
        <span class="p">)</span> <span class="o">&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">reverse_read</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span> <span class="n">forward_read</span><span class="o">.</span><span class="n">pos</span><span class="p">)):</span>
            <span class="n">list_d1d2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">reverse_read</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span> <span class="n">forward_read</span><span class="o">.</span><span class="n">pos</span><span class="p">)))</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="n">list_d1d2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">distance_1</span><span class="p">,</span> <span class="n">distance_2</span><span class="p">))</span>

    <span class="n">histo</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">histogram</span><span class="p">(</span><span class="n">list_d1d2</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">list_d1d2</span><span class="p">)))</span>

    <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">D1D2</span><span class="p">,</span> <span class="n">histo</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved d1d2 law at : </span><span class="si">{</span><span class="n">output_path</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">D1D2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="get_patterns">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.get_patterns">[docs]</a>
<span class="k">def</span> <span class="nf">get_patterns</span><span class="p">(</span><span class="n">forward_bam_file</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;group1.1.bam&quot;</span><span class="p">,</span> <span class="n">reverse_bam_file</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;group1.2.bam&quot;</span><span class="p">,</span> <span class="n">xs</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;xs.npy&quot;</span><span class="p">,</span> <span class="n">chrom_sizes</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;chromosome_sizes.npy&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">blacklist</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">output_dir</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the patterns distribution from read pairs alignment. .</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    forward_bam_file : str, optional</span>
<span class="sd">        Path to forward .bam alignment file, by default None, by default group1.1.bam, by default &quot;group1.1.bam&quot;, by default &quot;group1.1.bam&quot;</span>
<span class="sd">    reverse_bam_file : str, optional</span>
<span class="sd">        Path to reverse .bam alignment file, by default None, by default group1.1.bam, by default &quot;group1.1.bam&quot;, by default &quot;group1.2.bam&quot;</span>
<span class="sd">    xs : str, optional</span>
<span class="sd">        Path to the dictionary containing the xs values, by default &quot;xs.npy&quot;</span>
<span class="sd">    dist_frag : str, optional</span>
<span class="sd">        Path to the dictionary containing the inter fragment distances, by default &quot;dist.frag.npy&quot;</span>
<span class="sd">    circular : str, optional</span>
<span class="sd">        Name of the chromosomes to consider as circular, by default &quot;&quot;</span>
<span class="sd">    output_dir : str, optional</span>
<span class="sd">        Path to the folder where to save the dictionary, by default None, by default None, by default None</span>
<span class="sd">    &quot;&quot;&quot;</span>    

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Start generating patterns distribution...&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">output_dir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            
        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">getcwd</span><span class="p">())</span>

    <span class="k">else</span><span class="p">:</span>

        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>

    <span class="n">forward_bam_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">forward_bam_file</span><span class="p">)</span>
    <span class="n">reverse_bam_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">reverse_bam_file</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">forward_bam_path</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
        
        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Forward .bam file </span><span class="si">{</span><span class="n">forward_bam_file</span><span class="si">}</span><span class="s2"> not found. Please provide a valid path to a forward .bam file.&quot;</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="ow">not</span> <span class="n">reverse_bam_path</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>

        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reverse .bam file </span><span class="si">{</span><span class="n">reverse_bam_file</span><span class="si">}</span><span class="s2"> not found. Please provide a valid path to a reverse .bam file.&quot;</span><span class="p">)</span>
    
    <span class="c1">#Load xs</span>
    <span class="n">xs</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_dictionary</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">XS</span><span class="p">)</span>
    <span class="c1"># dist_frag = hio.load_dictionary(output_path / dist_frag)</span>
    <span class="n">chrom_size_dict</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_dictionary</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">chrom_sizes</span><span class="p">)</span>

    <span class="c1"># Create placeholders for the dictionaries</span>
    <span class="n">weirds</span> <span class="o">=</span> <span class="p">{</span><span class="n">seq_name</span> <span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">xs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">seq_name</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq_name</span> <span class="ow">in</span> <span class="n">xs</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
    <span class="n">uncuts</span> <span class="o">=</span> <span class="p">{</span><span class="n">seq_name</span> <span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">xs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">seq_name</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq_name</span> <span class="ow">in</span> <span class="n">xs</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
    <span class="n">loops</span> <span class="o">=</span> <span class="p">{</span><span class="n">seq_name</span> <span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">xs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">seq_name</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq_name</span> <span class="ow">in</span> <span class="n">xs</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>

    <span class="c1"># Create placeholder for area to divide logbins counts</span>
    <span class="n">trapezoids_area</span> <span class="o">=</span> <span class="p">{</span><span class="n">seq_name</span> <span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">xs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">seq_name</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq_name</span> <span class="ow">in</span> <span class="n">xs</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>

    <span class="c1"># Compute areas of trapezoids</span>

    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">xs</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">xs_</span> <span class="o">=</span> <span class="n">xs</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span>
        <span class="n">chrom_size_</span> <span class="o">=</span> <span class="n">chrom_size_dict</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span>

        <span class="n">trapezoids_area</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">=</span> <span class="p">[(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">chrom_size_</span> <span class="o">-</span> <span class="n">xs_</span><span class="p">[</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">xs_</span><span class="p">[</span><span class="n">j</span><span class="p">])</span> <span class="o">*</span> <span class="p">(</span><span class="n">xs_</span><span class="p">[</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">xs_</span><span class="p">[</span><span class="n">j</span><span class="p">])</span> <span class="o">*</span> <span class="mf">0.5</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">xs_</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)]</span>
        <span class="n">trapezoids_area</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="p">(((</span><span class="n">chrom_size_</span> <span class="o">-</span> <span class="n">xs_</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">))</span>

    <span class="n">forward_bam_handler</span><span class="p">,</span> <span class="n">reverse_bam_handler</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">forward_bam_path</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">),</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">reverse_bam_path</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">forward_read</span><span class="p">,</span> <span class="n">reverse_read</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">forward_bam_handler</span><span class="p">,</span> <span class="n">reverse_bam_handler</span><span class="p">):</span>
                <span class="c1"># TODO : Add blacklisting system to avoid counting reads from blacklisted regions</span>
                <span class="c1"># if not is_blacklisted(forward_read, reverse_read): --&gt; TO BE ADDED TO UTILS</span>
                <span class="c1">#    continue</span>
                <span class="c1"># blacklisting system per coordinates or per file as a list of coordinates (BED file)</span>
        <span class="k">if</span>  <span class="n">hut</span><span class="o">.</span><span class="n">is_blacklisted</span><span class="p">(</span><span class="n">read_forward</span> <span class="o">=</span> <span class="n">forward_read</span><span class="p">,</span> <span class="n">read_reverse</span> <span class="o">=</span> <span class="n">reverse_read</span><span class="p">,</span> <span class="n">blacklist</span> <span class="o">=</span> <span class="n">blacklist</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Blacklisted read pair : </span><span class="si">{</span><span class="n">forward_read</span><span class="o">.</span><span class="n">query_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="k">if</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_intra_chromosome</span><span class="p">(</span><span class="n">forward_read</span><span class="p">,</span> <span class="n">reverse_read</span><span class="p">):</span>
                
            <span class="k">if</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_weird</span><span class="p">(</span><span class="n">forward_read</span><span class="p">,</span> <span class="n">reverse_read</span><span class="p">):</span>

                <span class="n">weirds</span><span class="p">[</span><span class="n">forward_read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">][</span>
                    <span class="n">attribute_xs</span><span class="p">(</span>
                        <span class="n">xs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">forward_read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">),</span>
                        <span class="n">hut</span><span class="o">.</span><span class="n">get_cis_distance</span><span class="p">(</span><span class="n">forward_read</span><span class="p">,</span> <span class="n">reverse_read</span><span class="p">,</span> <span class="n">circular</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">if</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_uncut</span><span class="p">(</span><span class="n">forward_read</span><span class="p">,</span> <span class="n">reverse_read</span><span class="p">):</span>

                <span class="n">uncuts</span><span class="p">[</span><span class="n">forward_read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">][</span>
                    <span class="n">attribute_xs</span><span class="p">(</span>
                        <span class="n">xs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">forward_read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">),</span>
                        <span class="n">hut</span><span class="o">.</span><span class="n">get_cis_distance</span><span class="p">(</span><span class="n">forward_read</span><span class="p">,</span> <span class="n">reverse_read</span><span class="p">,</span> <span class="n">circular</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">if</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_circle</span><span class="p">(</span><span class="n">forward_read</span><span class="p">,</span> <span class="n">reverse_read</span><span class="p">):</span>
                <span class="n">loops</span><span class="p">[</span><span class="n">forward_read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">][</span>
                    <span class="n">attribute_xs</span><span class="p">(</span>
                        <span class="n">xs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">forward_read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">),</span>
                        <span class="n">hut</span><span class="o">.</span><span class="n">get_cis_distance</span><span class="p">(</span><span class="n">forward_read</span><span class="p">,</span> <span class="n">reverse_read</span><span class="p">,</span> <span class="n">circular</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
    
    <span class="n">forward_bam_handler</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">reverse_bam_handler</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">chromosome</span> <span class="ow">in</span> <span class="n">weirds</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>

        <span class="n">weirds</span><span class="p">[</span><span class="n">chromosome</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span>
            <span class="n">weirds</span><span class="p">[</span><span class="n">chromosome</span><span class="p">],</span>
            <span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">trapezoids_area</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">chromosome</span><span class="p">),</span> <span class="mi">2</span><span class="p">),</span>
            <span class="n">out</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">weirds</span><span class="p">[</span><span class="n">chromosome</span><span class="p">]),</span>
            <span class="n">where</span><span class="o">=</span><span class="n">trapezoids_area</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">chromosome</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">uncuts</span><span class="p">[</span><span class="n">chromosome</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span>
            <span class="n">uncuts</span><span class="p">[</span><span class="n">chromosome</span><span class="p">],</span>
            <span class="n">trapezoids_area</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">chromosome</span><span class="p">),</span>
            <span class="n">out</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">uncuts</span><span class="p">[</span><span class="n">chromosome</span><span class="p">]),</span>
            <span class="n">where</span><span class="o">=</span><span class="n">trapezoids_area</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">chromosome</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">loops</span><span class="p">[</span><span class="n">chromosome</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span>
            <span class="n">loops</span><span class="p">[</span><span class="n">chromosome</span><span class="p">],</span>
            <span class="n">trapezoids_area</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">chromosome</span><span class="p">),</span>
            <span class="n">out</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">loops</span><span class="p">[</span><span class="n">chromosome</span><span class="p">]),</span>
            <span class="n">where</span><span class="o">=</span><span class="n">trapezoids_area</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">chromosome</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="c1"># Correction to avoid empty last point on the curve</span>
    <span class="c1"># Corrected case where P(s) is nan due to trapezoid area equal 0</span>
    <span class="k">for</span> <span class="n">chromosome</span> <span class="ow">in</span> <span class="n">weirds</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>

        <span class="k">if</span> <span class="n">weirds</span><span class="p">[</span><span class="n">chromosome</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">weirds</span><span class="p">[</span><span class="n">chromosome</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]):</span>
            <span class="n">weirds</span><span class="p">[</span><span class="n">chromosome</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">weirds</span><span class="p">[</span><span class="n">chromosome</span><span class="p">][</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">uncuts</span><span class="p">[</span><span class="n">chromosome</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">uncuts</span><span class="p">[</span><span class="n">chromosome</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]):</span>
            <span class="n">uncuts</span><span class="p">[</span><span class="n">chromosome</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">uncuts</span><span class="p">[</span><span class="n">chromosome</span><span class="p">][</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">loops</span><span class="p">[</span><span class="n">chromosome</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">loops</span><span class="p">[</span><span class="n">chromosome</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]):</span>
            <span class="n">loops</span><span class="p">[</span><span class="n">chromosome</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">loops</span><span class="p">[</span><span class="n">chromosome</span><span class="p">][</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span>

    <span class="c1"># Post-processing to avoid nan values</span>
            
    <span class="n">smoothed_weirds</span> <span class="o">=</span> <span class="p">{</span><span class="n">seq_name</span> <span class="p">:</span> <span class="n">hut</span><span class="o">.</span><span class="n">replace_consecutive_zeros_with_mean</span><span class="p">(</span><span class="n">weirds</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">seq_name</span><span class="p">))</span> <span class="k">for</span> <span class="n">seq_name</span> <span class="ow">in</span> <span class="n">weirds</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
    <span class="n">smoothed_uncuts</span> <span class="o">=</span> <span class="p">{</span><span class="n">seq_name</span> <span class="p">:</span> <span class="n">hut</span><span class="o">.</span><span class="n">replace_consecutive_zeros_with_mean</span><span class="p">(</span><span class="n">uncuts</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">seq_name</span><span class="p">))</span> <span class="k">for</span> <span class="n">seq_name</span> <span class="ow">in</span> <span class="n">uncuts</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
    <span class="n">smoothed_loops</span> <span class="o">=</span> <span class="p">{</span><span class="n">seq_name</span> <span class="p">:</span> <span class="n">hut</span><span class="o">.</span><span class="n">replace_consecutive_zeros_with_mean</span><span class="p">(</span><span class="n">loops</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">seq_name</span><span class="p">))</span> <span class="k">for</span> <span class="n">seq_name</span> <span class="ow">in</span> <span class="n">loops</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>

    <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">WEIRDS</span><span class="p">,</span> <span class="n">smoothed_weirds</span><span class="p">)</span>
    <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">UNCUTS</span><span class="p">,</span> <span class="n">smoothed_uncuts</span><span class="p">)</span>
    <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">LOOPS</span><span class="p">,</span> <span class="n">smoothed_loops</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved </span><span class="si">{</span><span class="n">WEIRDS</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">UNCUTS</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">LOOPS</span><span class="si">}</span><span class="s2"> in </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="get_pair_ps">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.get_pair_ps">[docs]</a>
<span class="k">def</span> <span class="nf">get_pair_ps</span><span class="p">(</span><span class="n">read_forward</span> <span class="p">:</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignedSegment</span><span class="p">,</span> <span class="n">read_reverse</span> <span class="p">:</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignedSegment</span><span class="p">,</span> <span class="n">xs</span> <span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">weirds</span> <span class="p">:</span>  <span class="nb">dict</span><span class="p">,</span> <span class="n">uncuts</span> <span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">loops</span> <span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">circular</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Take two reads and return the P(s) value depending on event type (intra-chromosomal case only).</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    read_forward : pysam.AlignedSegment</span>
<span class="sd">        Forward read to compare with the reverse read.</span>
<span class="sd">    read_reverse : pysam.AlignedSegment</span>
<span class="sd">        Reverse read to compare with the forward read.</span>
<span class="sd">    xs : dict</span>
<span class="sd">        Dictionary containing log binning values for each chromosome.</span>
<span class="sd">    weirds : dict</span>
<span class="sd">        Dictionary containing number of weird events considering distance for each chromosome.</span>
<span class="sd">    uncuts : dict</span>
<span class="sd">        Dictionary containing number of uncuts events considering distance for each chromosome.</span>
<span class="sd">    loops : dict</span>
<span class="sd">        Dictionary containing number of loops events considering distance for each chromosome.</span>
<span class="sd">    circular : str, optional</span>
<span class="sd">        Name of the chromosomes to consider as circular, by default None, by default &quot;&quot;.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float</span>
<span class="sd">        P(s) of the pair considering the event type.</span>
<span class="sd">    &quot;&quot;&quot;</span>    
    
    <span class="k">if</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">query_name</span> <span class="o">!=</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">query_name</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Reads are not coming from the same pair.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_intra_chromosome</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Reads are not intra-chromosomal.&quot;</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_weird</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">weirds</span><span class="p">[</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_name</span><span class="p">][</span>
            <span class="n">attribute_xs</span><span class="p">(</span>
                <span class="n">xs</span><span class="p">[</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_name</span><span class="p">],</span>
                <span class="n">hut</span><span class="o">.</span><span class="n">get_cis_distance</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">,</span> <span class="n">circular</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">]</span>

    <span class="k">elif</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_uncut</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">uncuts</span><span class="p">[</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_name</span><span class="p">][</span>
            <span class="n">attribute_xs</span><span class="p">(</span>
                <span class="n">xs</span><span class="p">[</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_name</span><span class="p">],</span>
                <span class="n">hut</span><span class="o">.</span><span class="n">get_cis_distance</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">,</span> <span class="n">circular</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">]</span>

    <span class="k">elif</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_circle</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">loops</span><span class="p">[</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_name</span><span class="p">][</span>
            <span class="n">attribute_xs</span><span class="p">(</span>
                <span class="n">xs</span><span class="p">[</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_name</span><span class="p">],</span>
                <span class="n">hut</span><span class="o">.</span><span class="n">get_cis_distance</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">,</span> <span class="n">circular</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">]</span></div>

    

<div class="viewcode-block" id="get_trans_ps">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.get_trans_ps">[docs]</a>
<span class="k">def</span> <span class="nf">get_trans_ps</span><span class="p">(</span><span class="n">read_forward</span> <span class="p">:</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignedSegment</span><span class="p">,</span> <span class="n">read_reverse</span> <span class="p">:</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignedSegment</span><span class="p">,</span> <span class="n">trans_ps</span> <span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    read_forward : pysam.AlignedSegment</span>
<span class="sd">        Forward read to compare with the reverse read.</span>
<span class="sd">    read_reverse : pysam.AlignedSegment</span>
<span class="sd">        Reverse read to compare with the forward read.</span>
<span class="sd">    trans_ps : dict</span>
<span class="sd">        Dictionary of trans-chromosomal P(s)</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float</span>
<span class="sd">        Trans P(s) value</span>

<span class="sd">    &quot;&quot;&quot;</span>    
    <span class="k">if</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">query_name</span> <span class="o">!=</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">query_name</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Reads are not coming from the same pair.&quot;</span><span class="p">)</span>
    
    <span class="k">if</span>  <span class="n">hut</span><span class="o">.</span><span class="n">is_intra_chromosome</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">):</span>

        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Reads are not inter-chromosomal.&quot;</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">trans_ps</span><span class="p">[</span>
        <span class="nb">tuple</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_name</span><span class="p">,</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_name</span><span class="p">]))</span>
    <span class="p">]</span> </div>


<div class="viewcode-block" id="get_pair_cover">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.get_pair_cover">[docs]</a>
<span class="k">def</span> <span class="nf">get_pair_cover</span><span class="p">(</span><span class="n">read_forward</span> <span class="p">:</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignedSegment</span><span class="p">,</span> <span class="n">read_reverse</span> <span class="p">:</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignedSegment</span><span class="p">,</span> <span class="n">coverage</span> <span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">bins</span> <span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the coverage of a pair of reads.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    read_forward : pysam.AlignedSegment</span>
<span class="sd">        Forward read to compare with the reverse read.</span>
<span class="sd">    read_reverse : pysam.AlignedSegment</span>
<span class="sd">        Reverse read to compare with the forward read.</span>
<span class="sd">    coverage : dict</span>
<span class="sd">        Dictionary containing the coverage of each chromosome.</span>
<span class="sd">    bins : int</span>
<span class="sd">        Size of the desired bin.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float</span>
<span class="sd">        Product of the coverage of the pair of reads.</span>
<span class="sd">    &quot;&quot;&quot;</span>    

    <span class="k">if</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">query_name</span> <span class="o">!=</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">query_name</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Reads are not coming from the same pair.&quot;</span><span class="p">)</span>
    

    <span class="k">if</span> <span class="p">(</span><span class="n">read_forward</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">16</span> <span class="ow">or</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">272</span><span class="p">)</span> <span class="ow">and</span><span class="p">(</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">256</span><span class="p">):</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="n">coverage</span><span class="p">[</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_name</span><span class="p">][</span>
                <span class="nb">int</span><span class="p">(</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_end</span> <span class="o">/</span> <span class="n">bins</span><span class="p">)</span>
            <span class="p">]</span>
            <span class="o">*</span> <span class="n">coverage</span><span class="p">[</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_name</span><span class="p">][</span>
                <span class="nb">int</span><span class="p">(</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_start</span> <span class="o">/</span> <span class="n">bins</span><span class="p">)</span>
            <span class="p">]</span>
        <span class="p">)</span>

    <span class="k">elif</span> <span class="p">(</span><span class="n">read_forward</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">256</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">16</span> <span class="ow">or</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">272</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="n">coverage</span><span class="p">[</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_name</span><span class="p">][</span>
                <span class="nb">int</span><span class="p">(</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_start</span> <span class="o">/</span> <span class="n">bins</span><span class="p">)</span>
            <span class="p">]</span>
            <span class="o">*</span> <span class="n">coverage</span><span class="p">[</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_name</span><span class="p">][</span>
                <span class="nb">int</span><span class="p">(</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_end</span> <span class="o">/</span> <span class="n">bins</span><span class="p">)</span>
            <span class="p">]</span>
        <span class="p">)</span>

    <span class="k">elif</span> <span class="p">(</span><span class="n">read_forward</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">16</span> <span class="ow">or</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">272</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">16</span> <span class="ow">or</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">272</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="n">coverage</span><span class="p">[</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_name</span><span class="p">][</span>
                <span class="nb">int</span><span class="p">(</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_end</span> <span class="o">/</span> <span class="n">bins</span><span class="p">)</span>
            <span class="p">]</span>
            <span class="o">*</span> <span class="n">coverage</span><span class="p">[</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_name</span><span class="p">][</span>
                <span class="nb">int</span><span class="p">(</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_end</span> <span class="o">/</span> <span class="n">bins</span><span class="p">)</span>
            <span class="p">]</span>
        <span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="n">coverage</span><span class="p">[</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_name</span><span class="p">][</span>
                <span class="nb">int</span><span class="p">(</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_start</span> <span class="o">/</span> <span class="n">bins</span><span class="p">)</span>
            <span class="p">]</span>
            <span class="o">*</span> <span class="n">coverage</span><span class="p">[</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_name</span><span class="p">][</span>
                <span class="nb">int</span><span class="p">(</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_start</span> <span class="o">/</span> <span class="n">bins</span><span class="p">)</span>
            <span class="p">]</span>
        <span class="p">)</span></div>



<div class="viewcode-block" id="get_d1d2">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.get_d1d2">[docs]</a>
<span class="k">def</span> <span class="nf">get_d1d2</span><span class="p">(</span><span class="n">read_forward</span> <span class="p">:</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignedSegment</span><span class="p">,</span> <span class="n">read_reverse</span> <span class="p">:</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignedSegment</span><span class="p">,</span> <span class="n">restriction_map</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">d1d2</span> <span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the d1d2 value of a pair of reads.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    read_forward : pysam.AlignedSegment</span>
<span class="sd">        Forward read to compare with the reverse read.</span>
<span class="sd">    read_reverse : pysam.AlignedSegment</span>
<span class="sd">        Reverse read to compare with the forward read.</span>
<span class="sd">    restriction_map : dict, optional</span>
<span class="sd">        Restriction map saved as a dictionary like chrom_name : list of restriction sites&#39; position, by default None</span>
<span class="sd">    d1d2 : np.array, optional</span>
<span class="sd">        Distribution of d1d2 values, by default None</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    int</span>
<span class="sd">        d1d2 value</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">query_name</span> <span class="o">!=</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">query_name</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Reads are not coming from the same pair.&quot;</span><span class="p">)</span>
    
    <span class="c1"># Get appropriate restriction sites vecctors in dicitionary</span>
    <span class="n">r_sites_read_for</span> <span class="o">=</span> <span class="n">restriction_map</span><span class="p">[</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_name</span><span class="p">]</span>
    <span class="n">r_sites_read_rev</span> <span class="o">=</span> <span class="n">restriction_map</span><span class="p">[</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_name</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">256</span><span class="p">:</span>

        <span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="n">r_sites_read_for</span><span class="p">,</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span>
        <span class="n">distance_1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">r_sites_read_for</span><span class="p">[</span><span class="n">index</span><span class="p">],</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">pos</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">16</span> <span class="ow">or</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">272</span><span class="p">:</span>

        <span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="n">r_sites_read_for</span><span class="p">,</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">reference_end</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
        <span class="n">distance_1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_end</span><span class="p">,</span> <span class="n">r_sites_read_for</span><span class="p">[</span><span class="n">index</span><span class="p">])</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">256</span><span class="p">:</span>

        <span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span>
            <span class="n">r_sites_read_rev</span><span class="p">,</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_start</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="s2">&quot;right&quot;</span>
        <span class="p">)</span>  <span class="c1"># right</span>
        <span class="n">distance_2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">r_sites_read_rev</span><span class="p">[</span><span class="n">index</span><span class="p">],</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_start</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">16</span> <span class="ow">or</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">flag</span> <span class="o">==</span> <span class="mi">272</span><span class="p">:</span>

        <span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span>
            <span class="n">r_sites_read_rev</span><span class="p">,</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_end</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="s2">&quot;left&quot;</span>
        <span class="p">)</span>  <span class="c1"># left</span>
        <span class="n">distance_2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_end</span><span class="p">,</span> <span class="n">r_sites_read_rev</span><span class="p">[</span><span class="n">index</span><span class="p">])</span>
        <span class="p">)</span>

    <span class="c1"># Correction for uncuts with no restriction sites inside</span>
    <span class="k">if</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">reference_name</span> <span class="o">==</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_name</span> <span class="ow">and</span> <span class="n">np</span><span class="o">.</span><span class="n">add</span><span class="p">(</span>
        <span class="n">distance_1</span><span class="p">,</span> <span class="n">distance_2</span>
    <span class="p">)</span> <span class="o">&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">pos</span><span class="p">)):</span>
        <span class="n">distance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">pos</span><span class="p">))</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">distance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">distance_1</span><span class="p">,</span> <span class="n">distance_2</span><span class="p">)</span>
        
    <span class="k">return</span> <span class="n">d1d2</span><span class="p">[</span><span class="n">distance</span><span class="p">]</span></div>


<div class="viewcode-block" id="get_density">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.get_density">[docs]</a>
<span class="k">def</span> <span class="nf">get_density</span><span class="p">(</span><span class="n">read_forward</span> <span class="p">:</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignedSegment</span><span class="p">,</span> <span class="n">read_reverse</span> <span class="p">:</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignedSegment</span><span class="p">,</span> <span class="n">density_map</span> <span class="p">:</span> <span class="nb">dict</span><span class="p">[(</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">],</span> <span class="n">bin_size</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get density from density map dictionary.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    read_forward : pysam.AlignedSegment</span>
<span class="sd">        Forward read to get density from.</span>
<span class="sd">    read_reverse : pysam.AlignedSegment</span>
<span class="sd">        Reverse read to get density from.</span>
<span class="sd">    density_map : dict</span>
<span class="sd">        Dictionary containing density maps for each chromosome couple.</span>
<span class="sd">    bin_size : int, optional</span>
<span class="sd">        Resolution of the matrix on which density map has been computed, by default 2000</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float</span>
<span class="sd">        Density corresponding to the pair of reads.</span>

<span class="sd">    &quot;&quot;&quot;</span>    
    <span class="k">if</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">query_name</span> <span class="o">!=</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">query_name</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Reads are not coming from the same pair.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">(</span><span class="n">read_forward</span><span class="p">):</span>
        <span class="n">position_for</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_end</span> <span class="o">//</span> <span class="n">bin_size</span><span class="p">)</span>

    <span class="k">elif</span> <span class="ow">not</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">(</span><span class="n">read_forward</span><span class="p">):</span>
        <span class="n">position_for</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_start</span> <span class="o">//</span> <span class="n">bin_size</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">(</span><span class="n">read_reverse</span><span class="p">):</span>
        <span class="n">position_rev</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_end</span> <span class="o">//</span> <span class="n">bin_size</span><span class="p">)</span>

    <span class="k">elif</span> <span class="ow">not</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_reverse</span><span class="p">(</span><span class="n">read_reverse</span><span class="p">):</span>

        <span class="n">position_rev</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_start</span> <span class="o">//</span> <span class="n">bin_size</span><span class="p">)</span>

    <span class="n">couple_density</span> <span class="o">=</span> <span class="n">density_map</span><span class="o">.</span><span class="n">get</span><span class="p">((</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_name</span><span class="p">,</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_name</span><span class="p">))[</span><span class="n">position_for</span><span class="p">,</span> <span class="n">position_rev</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">couple_density</span></div>



<div class="viewcode-block" id="compute_propensity">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.compute_propensity">[docs]</a>
<span class="k">def</span> <span class="nf">compute_propensity</span><span class="p">(</span><span class="n">read_forward</span> <span class="p">:</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignedSegment</span><span class="p">,</span> <span class="n">read_reverse</span> <span class="p">:</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignedSegment</span><span class="p">,</span> <span class="n">restriction_map</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">xs</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">weirds</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">uncuts</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">loops</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">circular</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">trans_ps</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>  <span class="n">coverage</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">bins</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span> <span class="n">d1d2</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">density_map</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>  <span class="n">mode</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;full&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute propensity for read pair to be selected among all plausible pairs related to multi-mapping reads.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    read_forward : pysam.AlignedSegment</span>
<span class="sd">        Forward read to compare with the reverse read.</span>
<span class="sd">    read_reverse : pysam.AlignedSegment</span>
<span class="sd">        Reverse read to compare with the forward read.</span>
<span class="sd">    restriction_map : dict, optional</span>
<span class="sd">        Restriction map saved as a dictionary like chrom_name : list of restriction sites&#39; position, by default None</span>
<span class="sd">    xs : dict</span>
<span class="sd">        Dictionary containing log binning values for each chromosome.</span>
<span class="sd">    weirds : dict</span>
<span class="sd">        Dictionary containing number of weird events considering distance for each chromosome.</span>
<span class="sd">    uncuts : dict</span>
<span class="sd">        Dictionary containing number of uncuts events considering distance for each chromosome.</span>
<span class="sd">    loops : dict</span>
<span class="sd">        Dictionary containing number of loops events considering distance for each chromosome.</span>
<span class="sd">    circular : str, optional</span>
<span class="sd">        Name of the chromosomes to consider as circular, by default None, by default &quot;&quot;.</span>
<span class="sd">    trans_ps : dict</span>
<span class="sd">        Dictionary of trans-chromosomal P(s)</span>
<span class="sd">    coverage : dict</span>
<span class="sd">        Dictionary containing the coverage of each chromosome.</span>
<span class="sd">    bins : int</span>
<span class="sd">        Size of the desired bin, by default 2000</span>
<span class="sd">    d1d2 : np.array, optional</span>
<span class="sd">        Distribution of d1d2 values, by default None</span>
<span class="sd">    density : dict</span>
<span class="sd">        Dictionary of contact density by chromosome couple.</span>
<span class="sd">    mode : str, optional</span>
<span class="sd">        Mode to use to compute propensity among, by default &quot;full&quot;</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float</span>
<span class="sd">        Propensity to use for read couple drawing</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">read_forward</span><span class="o">.</span><span class="n">query_name</span> <span class="o">!=</span> <span class="n">read_reverse</span><span class="o">.</span><span class="n">query_name</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Reads are not coming from the same pair.&quot;</span><span class="p">)</span>
    

    <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;random&quot;</span><span class="p">:</span>

        <span class="k">return</span> <span class="mi">1</span>

    <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;full&quot;</span><span class="p">:</span>

        <span class="k">if</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_intra_chromosome</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">):</span>
        
            <span class="n">ps</span> <span class="o">=</span> <span class="n">get_pair_ps</span><span class="p">(</span>
                <span class="n">read_forward</span><span class="p">,</span>
                <span class="n">read_reverse</span><span class="p">,</span>
                <span class="n">xs</span><span class="p">,</span>
                <span class="n">weirds</span><span class="p">,</span>
                <span class="n">uncuts</span><span class="p">,</span>
                <span class="n">loops</span><span class="p">,</span>
                <span class="n">circular</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            
            <span class="n">ps</span> <span class="o">=</span> <span class="n">get_trans_ps</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">,</span> <span class="n">trans_ps</span><span class="p">)</span>


            <span class="c1"># Avoid ps = 0 making the read unselectable. Value of 1 make the propensity unsensitive to P(s).</span>
            <span class="k">if</span> <span class="n">ps</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>

                <span class="n">ps</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="n">cover</span> <span class="o">=</span> <span class="n">get_pair_cover</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">,</span> <span class="n">coverage</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">)</span>

        <span class="c1"># Avoid cover = 0 making the read unselectable. Value of 1 make the propensity unsensitive to coverage.</span>
        <span class="k">if</span> <span class="n">cover</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">cover</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">d1d2</span> <span class="o">=</span> <span class="n">get_d1d2</span><span class="p">(</span>
            <span class="n">read_forward</span><span class="p">,</span>
            <span class="n">read_reverse</span><span class="p">,</span>
            <span class="n">restriction_map</span><span class="p">,</span>
            <span class="n">d1d2</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">except</span><span class="p">:</span>

            <span class="n">d1d2</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="n">density</span> <span class="o">=</span> <span class="n">get_density</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">,</span> <span class="n">density_map</span> <span class="o">=</span> <span class="n">density_map</span><span class="p">)</span>
    
        <span class="k">return</span> <span class="n">ps</span> <span class="o">*</span> <span class="n">d1d2</span> <span class="o">*</span> <span class="n">cover</span> <span class="o">*</span> <span class="n">density</span>
    
    <span class="k">elif</span> <span class="n">mode</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;standard&quot;</span><span class="p">,</span> <span class="s2">&quot;omics&quot;</span><span class="p">]</span> <span class="p">:</span>
    
        <span class="k">if</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_intra_chromosome</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">):</span>
        
            <span class="n">ps</span> <span class="o">=</span> <span class="n">get_pair_ps</span><span class="p">(</span>
                <span class="n">read_forward</span><span class="p">,</span>
                <span class="n">read_reverse</span><span class="p">,</span>
                <span class="n">xs</span><span class="p">,</span>
                <span class="n">weirds</span><span class="p">,</span>
                <span class="n">uncuts</span><span class="p">,</span>
                <span class="n">loops</span><span class="p">,</span>
                <span class="n">circular</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            
            <span class="n">ps</span> <span class="o">=</span> <span class="n">get_trans_ps</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">,</span> <span class="n">trans_ps</span><span class="p">)</span>


            <span class="c1"># Avoid ps = 0 making the read unselectable. Value of 1 make the propensity unsensitive to P(s).</span>
            <span class="k">if</span> <span class="n">ps</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>

                <span class="n">ps</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="n">cover</span> <span class="o">=</span> <span class="n">get_pair_cover</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">,</span> <span class="n">coverage</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">)</span>

        <span class="c1"># Avoid cover = 0 making the read unselectable. Value of 1 make the propensity unsensitive to coverage.</span>
        <span class="k">if</span> <span class="n">cover</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">cover</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="k">return</span> <span class="n">ps</span> <span class="o">*</span> <span class="n">cover</span>

    <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;one_enzyme&quot;</span><span class="p">:</span>
    
        <span class="k">if</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_intra_chromosome</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">):</span>
        
            <span class="n">ps</span> <span class="o">=</span> <span class="n">get_pair_ps</span><span class="p">(</span>
                <span class="n">read_forward</span><span class="p">,</span>
                <span class="n">read_reverse</span><span class="p">,</span>
                <span class="n">xs</span><span class="p">,</span>
                <span class="n">weirds</span><span class="p">,</span>
                <span class="n">uncuts</span><span class="p">,</span>
                <span class="n">loops</span><span class="p">,</span>
                <span class="n">circular</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            
            <span class="n">ps</span> <span class="o">=</span> <span class="n">get_trans_ps</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">,</span> <span class="n">trans_ps</span><span class="p">)</span>


            <span class="c1"># Avoid ps = 0 making the read unselectable. Value of 1 make the propensity unsensitive to P(s).</span>
            <span class="k">if</span> <span class="n">ps</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>

                <span class="n">ps</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="n">cover</span> <span class="o">=</span> <span class="n">get_pair_cover</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">,</span> <span class="n">coverage</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">)</span>

        <span class="c1"># Avoid cover = 0 making the read unselectable. Value of 1 make the propensity unsensitive to coverage.</span>
        <span class="k">if</span> <span class="n">cover</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">cover</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">d1d2</span> <span class="o">=</span> <span class="n">get_d1d2</span><span class="p">(</span>
            <span class="n">read_forward</span><span class="p">,</span>
            <span class="n">read_reverse</span><span class="p">,</span>
            <span class="n">restriction_map</span><span class="p">,</span>
            <span class="n">d1d2</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">except</span><span class="p">:</span>

            <span class="n">d1d2</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">ps</span><span class="p">)</span> <span class="ow">or</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">cover</span><span class="p">)</span> <span class="ow">or</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">d1d2</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ps : </span><span class="si">{</span><span class="n">ps</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;cover : </span><span class="si">{</span><span class="n">cover</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;d1d2 : </span><span class="si">{</span><span class="n">d1d2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;read_forward : </span><span class="si">{</span><span class="n">read_forward</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;read_forward reference name : </span><span class="si">{</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;read_reverse : </span><span class="si">{</span><span class="n">read_reverse</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;read_reverse reference name : </span><span class="si">{</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ps : </span><span class="si">{</span><span class="n">ps</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;cover : </span><span class="si">{</span><span class="n">cover</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;d1d2 : </span><span class="si">{</span><span class="n">d1d2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;read_forward : </span><span class="si">{</span><span class="n">read_forward</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;read_forward reference name : </span><span class="si">{</span><span class="n">read_forward</span><span class="o">.</span><span class="n">reference_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;read_reverse : </span><span class="si">{</span><span class="n">read_reverse</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;read_reverse reference name : </span><span class="si">{</span><span class="n">read_reverse</span><span class="o">.</span><span class="n">reference_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            
            <span class="c1"># sys.exit()</span>

        <span class="k">return</span> <span class="n">ps</span> <span class="o">*</span> <span class="n">d1d2</span> <span class="o">*</span> <span class="n">cover</span>

    
    <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;ps&quot;</span><span class="p">:</span>
    
        <span class="k">if</span> <span class="n">hut</span><span class="o">.</span><span class="n">is_intra_chromosome</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">):</span>
        
            <span class="n">ps</span> <span class="o">=</span> <span class="n">get_pair_ps</span><span class="p">(</span>
                <span class="n">read_forward</span><span class="p">,</span>
                <span class="n">read_reverse</span><span class="p">,</span>
                <span class="n">xs</span><span class="p">,</span>
                <span class="n">weirds</span><span class="p">,</span>
                <span class="n">uncuts</span><span class="p">,</span>
                <span class="n">loops</span><span class="p">,</span>
                <span class="n">circular</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            
            <span class="n">ps</span> <span class="o">=</span> <span class="n">get_trans_ps</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">,</span> <span class="n">trans_ps</span><span class="p">)</span>


            <span class="c1"># Avoid ps = 0 making the read unselectable. Value of 1 make the propensity unsensitive to P(s).</span>
            <span class="k">if</span> <span class="n">ps</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>

                <span class="n">ps</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="k">return</span> <span class="n">ps</span>

    <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;cov&quot;</span><span class="p">:</span>
        
        <span class="n">cover</span> <span class="o">=</span> <span class="n">get_pair_cover</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">,</span> <span class="n">coverage</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">)</span>

        <span class="c1"># Avoid cover = 0 making the read unselectable. Value of 1 make the propensity unsensitive to coverage.</span>
        <span class="k">if</span> <span class="n">cover</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">cover</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="k">return</span> <span class="n">cover</span>


    <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;d1d2&quot;</span><span class="p">:</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">d1d2</span> <span class="o">=</span> <span class="n">get_d1d2</span><span class="p">(</span>
            <span class="n">read_forward</span><span class="p">,</span>
            <span class="n">read_reverse</span><span class="p">,</span>
            <span class="n">restriction_map</span><span class="p">,</span>
            <span class="n">d1d2</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">except</span><span class="p">:</span>

            <span class="n">d1d2</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="k">return</span> <span class="n">d1d2</span>


    <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;density&quot;</span><span class="p">:</span>
        
        <span class="n">density</span> <span class="o">=</span> <span class="n">get_density</span><span class="p">(</span><span class="n">read_forward</span><span class="p">,</span> <span class="n">read_reverse</span><span class="p">,</span> <span class="n">density_map</span> <span class="o">=</span> <span class="n">density_map</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">density</span></div>


    
<div class="viewcode-block" id="draw_read_couple">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.draw_read_couple">[docs]</a>
<span class="k">def</span> <span class="nf">draw_read_couple</span><span class="p">(</span><span class="n">propensities</span> <span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Draw an index respecting distribution of propensities. This function is used to draw a couple of reads considering the propensity of each couple.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    propensities : np.array</span>
<span class="sd">        Array containing all the propensities of each couple of reads.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    int</span>
<span class="sd">        Index of the couple of reads drawn.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">xk</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">propensities</span><span class="p">))</span>

    <span class="k">if</span>  <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">propensities</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span> 
        <span class="k">try</span><span class="p">:</span>
            <span class="n">pk</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">propensities</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">propensities</span><span class="p">))</span>
        
        <span class="k">except</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;pk : </span><span class="si">{</span><span class="n">pk</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;propensities : </span><span class="si">{</span><span class="n">propensities</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">propensities</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">try</span> <span class="p">:</span> 
            <span class="n">pk</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="n">xk</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">propensities</span><span class="p">)))</span>

        <span class="k">except</span> <span class="p">:</span>    
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;pk : </span><span class="si">{</span><span class="n">pk</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;propensities : </span><span class="si">{</span><span class="n">propensities</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">else</span> <span class="p">:</span> 
        <span class="n">pk</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="n">xk</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">propensities</span><span class="p">)))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Propensities : </span><span class="si">{</span><span class="n">propensities</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1">#TODO: Added 08/11/2024</span>
    <span class="c1">## Correct propensities to avoid nan values</span>
    <span class="c1">## If propensity contains nan values, replace them with 0</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">pk</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">pk</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">pk</span><span class="p">)</span>

    <span class="n">index</span> <span class="o">=</span> <span class="n">choice</span><span class="p">(</span><span class="n">xk</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="n">pk</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">index</span></div>


<div class="viewcode-block" id="reattribute_reads">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.reattribute_reads">[docs]</a>
<span class="k">def</span> <span class="nf">reattribute_reads</span><span class="p">(</span><span class="n">reads_couple</span> <span class="p">:</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;group2.1.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;group2.2.bam&quot;</span><span class="p">),</span> <span class="n">restriction_map</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">xs</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="s2">&quot;xs.npy&quot;</span><span class="p">,</span> <span class="n">weirds</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="s2">&quot;weirds.npy&quot;</span><span class="p">,</span> <span class="n">uncuts</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="s2">&quot;uncuts.npy&quot;</span><span class="p">,</span> <span class="n">loops</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="s2">&quot;loops.npy&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">trans_ps</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="s2">&quot;trans_ps.npy&quot;</span><span class="p">,</span>  <span class="n">coverage</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="s2">&quot;coverage.npy&quot;</span><span class="p">,</span> <span class="n">bins</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span> <span class="n">d1d2</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="s2">&quot;d1d2.npy&quot;</span><span class="p">,</span> <span class="n">density_map</span> <span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="s2">&quot;density_map.npy&quot;</span><span class="p">,</span>  <span class="n">mode</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;full&quot;</span><span class="p">,</span> <span class="n">output_dir</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Re-attribute multi-mapping (ambiguous) reads considering sets of statistical laws.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    reads_couple : tuple[str, str], optional</span>
<span class="sd">        Paths to ambiguous reads alignment files (.bam), by default (&quot;group2.1.bam&quot;, &quot;group2.2.bam&quot;)</span>
<span class="sd">    restriction_map : dict, optional</span>
<span class="sd">        Restriction map saved as a dictionary like chrom_name : list of restriction sites&#39; position, by default None</span>
<span class="sd">    xs : dict</span>
<span class="sd">        Dictionary containing log binning values for each chromosome.</span>
<span class="sd">    weirds : dict</span>
<span class="sd">        Dictionary containing number of weird events considering distance for each chromosome.</span>
<span class="sd">    uncuts : dict</span>
<span class="sd">        Dictionary containing number of uncuts events considering distance for each chromosome.</span>
<span class="sd">    loops : dict</span>
<span class="sd">        Dictionary containing number of loops events considering distance for each chromosome.</span>
<span class="sd">    circular : str, optional</span>
<span class="sd">        Name of the chromosomes to consider as circular, by default None, by default &quot;&quot;.</span>
<span class="sd">    trans_ps : dict</span>
<span class="sd">        Dictionary of trans-chromosomal P(s)</span>
<span class="sd">    coverage : dict</span>
<span class="sd">        Dictionary containing the coverage of each chromosome.</span>
<span class="sd">    bins : int</span>
<span class="sd">        Size of the desired bin, by default 2000</span>
<span class="sd">    d1d2 : np.array, optional</span>
<span class="sd">        Distribution of d1d2 values, by default None</span>
<span class="sd">    density : np.array, optional</span>
<span class="sd">        Dictionary containing density maps per chromosome couples as, by default None</span>
<span class="sd">    mode : str, optional</span>
<span class="sd">        Mode to use to compute propensity among, by default &quot;full&quot;</span>
<span class="sd">    output_dir : str, optional</span>
<span class="sd">        Path to the re-attributed ambiguous reads alignment files are saved, by default None</span>
<span class="sd">    &quot;&quot;&quot;</span> 

    <span class="k">if</span> <span class="n">output_dir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            
        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">getcwd</span><span class="p">())</span>

    <span class="k">else</span><span class="p">:</span>
            
        <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>

    <span class="c1">#Reload dictionaries</span>
    <span class="n">xs</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_dictionary</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">xs</span><span class="p">)</span>
    <span class="n">weirds</span>  <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_dictionary</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">weirds</span><span class="p">)</span>
    <span class="n">uncuts</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_dictionary</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">uncuts</span><span class="p">)</span>
    <span class="n">loops</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_dictionary</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">loops</span><span class="p">)</span>
    <span class="n">trans_ps</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_dictionary</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">trans_ps</span><span class="p">)</span>
    <span class="n">coverage</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_dictionary</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="n">coverage</span><span class="p">)</span>
    <span class="n">d1d2</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">density</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;full&quot;</span> <span class="p">:</span>

        <span class="n">d1d2</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_dictionary</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;d1d2.npy&quot;</span><span class="p">)</span>
        <span class="n">density</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_dictionary</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;density_map.npy&quot;</span><span class="p">)</span>
    
    <span class="k">elif</span> <span class="n">mode</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;d1d2&quot;</span><span class="p">,</span> <span class="s2">&quot;one_enzyme&quot;</span><span class="p">]:</span> <span class="c1"># TODO : to be completed</span>

        <span class="n">d1d2</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_dictionary</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;d1d2.npy&quot;</span><span class="p">)</span>
        
    <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;density&quot;</span><span class="p">:</span>
            
        <span class="n">density</span> <span class="o">=</span> <span class="n">hio</span><span class="o">.</span><span class="n">load_dictionary</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;density_map.npy&quot;</span><span class="p">)</span>

    <span class="n">forward_bam_path</span><span class="p">,</span> <span class="n">reverse_bam_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">reads_couple</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">Path</span><span class="p">(</span><span class="n">reads_couple</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">file_id</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="n">id_for</span> <span class="o">=</span> <span class="n">uuid</span><span class="o">.</span><span class="n">uuid4</span><span class="p">()</span>
    <span class="n">id_rev</span> <span class="o">=</span> <span class="n">uuid</span><span class="o">.</span><span class="n">uuid4</span><span class="p">()</span>

    <span class="n">forward_bam_handler</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">forward_bam_path</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
    <span class="n">reverse_bam_handler</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">reverse_bam_path</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>

    <span class="n">forward_out_bam_handler</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;forward_</span><span class="si">{</span><span class="n">id_for</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">file_id</span><span class="si">}</span><span class="s2">_predicted.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">,</span> <span class="n">template</span><span class="o">=</span><span class="n">forward_bam_handler</span><span class="p">)</span>
    <span class="n">reverse_out_bam_handler</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;reverse_</span><span class="si">{</span><span class="n">id_rev</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">file_id</span><span class="si">}</span><span class="s2">_predicted.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">,</span> <span class="n">template</span><span class="o">=</span><span class="n">reverse_bam_handler</span><span class="p">)</span>

    <span class="c1"># Instanciate generators</span>
    <span class="n">forward_generator</span> <span class="o">=</span> <span class="n">hut</span><span class="o">.</span><span class="n">bam_iterator</span><span class="p">(</span><span class="n">forward_bam_path</span><span class="p">)</span>
    <span class="n">reverse_generator</span> <span class="o">=</span> <span class="n">hut</span><span class="o">.</span><span class="n">bam_iterator</span><span class="p">(</span><span class="n">reverse_bam_path</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">forward_block</span><span class="p">,</span> <span class="n">reverse_block</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">forward_generator</span><span class="p">,</span> <span class="n">reverse_generator</span><span class="p">):</span>

        <span class="n">propensities</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">combinations</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="n">forward_block</span><span class="p">),</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">reverse_block</span><span class="p">)))</span>

        <span class="c1"># print(f&quot;combinations : {combinations}&quot;)</span>

        <span class="k">for</span> <span class="n">combination</span> <span class="ow">in</span> <span class="n">combinations</span><span class="p">:</span>

            <span class="n">propensities</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">compute_propensity</span><span class="p">(</span><span class="n">read_forward</span> <span class="o">=</span> <span class="n">combination</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">read_reverse</span> <span class="o">=</span> <span class="n">combination</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">restriction_map</span> <span class="o">=</span> <span class="n">restriction_map</span><span class="p">,</span> <span class="n">xs</span> <span class="o">=</span> <span class="n">xs</span><span class="p">,</span> <span class="n">weirds</span> <span class="o">=</span> <span class="n">weirds</span><span class="p">,</span> <span class="n">uncuts</span> <span class="o">=</span> <span class="n">uncuts</span><span class="p">,</span> <span class="n">loops</span> <span class="o">=</span> <span class="n">loops</span><span class="p">,</span> <span class="n">trans_ps</span> <span class="o">=</span> <span class="n">trans_ps</span><span class="p">,</span> <span class="n">coverage</span> <span class="o">=</span> <span class="n">coverage</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="n">bins</span><span class="p">,</span> <span class="n">d1d2</span> <span class="o">=</span> <span class="n">d1d2</span><span class="p">,</span> <span class="n">density_map</span> <span class="o">=</span> <span class="n">density</span><span class="p">,</span> <span class="n">mode</span> <span class="o">=</span> <span class="n">mode</span><span class="p">))</span>

        <span class="n">selected_couple_index</span> <span class="o">=</span> <span class="n">draw_read_couple</span><span class="p">(</span><span class="n">propensities</span><span class="p">)</span>

        <span class="n">selected_read_forward</span><span class="p">,</span> <span class="n">selected_read_reverse</span> <span class="o">=</span> <span class="n">combinations</span><span class="p">[</span><span class="n">selected_couple_index</span><span class="p">]</span>
        <span class="n">selected_read_forward</span><span class="o">.</span><span class="n">set_tag</span><span class="p">(</span><span class="s2">&quot;XL&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">forward_block</span><span class="p">))</span>
        <span class="n">selected_read_reverse</span><span class="o">.</span><span class="n">set_tag</span><span class="p">(</span><span class="s2">&quot;XL&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">reverse_block</span><span class="p">))</span>

        <span class="n">forward_out_bam_handler</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">selected_read_forward</span><span class="p">)</span>
        <span class="n">reverse_out_bam_handler</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">selected_read_reverse</span><span class="p">)</span>
        
    
    <span class="n">forward_bam_handler</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">reverse_bam_handler</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Predictions written in </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="pearson_score">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.pearson_score">[docs]</a>
<span class="k">def</span> <span class="nf">pearson_score</span><span class="p">(</span><span class="n">original_matrix</span> <span class="p">:</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">,</span> <span class="n">rescued_matrix</span> <span class="p">:</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span> <span class="p">,</span> <span class="n">markers</span> <span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">int</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute Pearson correlation between concatenated matrix bins which have been deleted and reconstructed.</span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    original_matrix : cooler.Cooler</span>
<span class="sd">        Cooler object containing the original matrix.</span>
<span class="sd">    rescued_matrix : cooler.Cooler</span>
<span class="sd">        Cooler object containing the reconstructed matrix.</span>
<span class="sd">    markers : list[int]</span>
<span class="sd">        List of markers to consider. Markers are the bins which have been deleted.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float</span>
<span class="sd">        Pearson correlation between original and reconstructed matrix.</span>
<span class="sd">    &quot;&quot;&quot;</span>    
    <span class="n">ori_matrix</span> <span class="o">=</span> <span class="n">original_matrix</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">balance</span><span class="o">=</span><span class="kc">False</span><span class="p">)[:]</span>
    <span class="n">reco_matrix</span> <span class="o">=</span> <span class="n">rescued_matrix</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">balance</span><span class="o">=</span><span class="kc">False</span><span class="p">)[:]</span>

    <span class="n">ori_vector</span> <span class="o">=</span> <span class="n">ori_matrix</span><span class="p">[</span><span class="n">markers</span><span class="p">]</span>
    <span class="n">reco_vector</span> <span class="o">=</span> <span class="n">reco_matrix</span><span class="p">[</span><span class="n">markers</span><span class="p">]</span>

    <span class="n">pearson_score</span> <span class="o">=</span> <span class="n">pearsonr</span><span class="p">(</span><span class="n">ori_vector</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span> <span class="n">reco_vector</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span>

    <span class="k">return</span> <span class="n">pearson_score</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span></div>


<span class="c1"># Benchamrk analysis functions</span>
<div class="viewcode-block" id="get_top_pattern">
<a class="viewcode-back" href="../../hicberg.html#hicberg.statistics.get_top_pattern">[docs]</a>
<span class="k">def</span> <span class="nf">get_top_pattern</span><span class="p">(</span><span class="n">file</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">top</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">chromosome</span> <span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get top patterns from a dataframe</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    df : pd.DataFrame, optional</span>
<span class="sd">        Dataframe containing patterns given by Chromosight, by default None</span>
<span class="sd">    top : int, optional</span>
<span class="sd">        Percentage of top patterns to get, by default 10</span>
<span class="sd">    chromosome : str, optional</span>
<span class="sd">        Chromosome to consider, by default None</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pd.DataFrame</span>
<span class="sd">        Dataframe containing top percentage patterns.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
    <span class="n">top_factor</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">top</span><span class="p">)</span> <span class="o">//</span> <span class="mi">100</span>

    <span class="k">if</span> <span class="n">chromosome</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;chrom1 == &#39;</span><span class="si">{</span><span class="n">chromosome</span><span class="si">}</span><span class="s2">&#39; and chrom2 == &#39;</span><span class="si">{</span><span class="n">chromosome</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>
    <span class="n">df_top</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;score&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">top_factor</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">df_top</span></div>


</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Sébastien Gradit.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>